```{r}

source("~/data2/rstudio/birds/utils/seq_analysis.R")
source("~/data2/rstudio/birds/utils/colors.R")
source("~/data2/rstudio/birds/utils/network.R")
source("~/data2/rstudio/birds/utils/db.R")
source("~/data2/rstudio/birds/utils/stats.R")
#source("~/data2/rstudio/birds/utils/scde.R")
source("~/data2/rstudio/birds/utils/common_aesthetics.R")
source("~/data2/rstudio/birds/utils/scRNA.R")
library(tidyverse)
library(reshape2)
library(viridis)
library(RColorBrewer)

library(limma)
library(edgeR)
library(sva)
library(proxy)
library(ggrepel)
library(cowplot)
library(egg)
library(Biostrings)
library(BiocParallel)
library(SummarizedExperiment)
library(ComplexHeatmap)
library(future)
library(circlize)
library(scales)
library(furrr)
library(flextable)
library(ggrastr)

filter = dplyr::filter
rename = dplyr::rename
select = dplyr::select
```

## Set up dir

```{r}
#prefix_song = "~/data2/rstudio/birds/singing/song_analysis"
base_dir = "~/data2/rstudio/birds" ## to change
prefix = file.path(base_dir, "deaf_gex/seq_analysis")
prefix_data = file.path(prefix, "data", "deaf_slcr-seq")
script_name = "deaf_voom"
prefix_cur = file.path(prefix, script_name)
dir.create(prefix_cur, recursive = T)
```

## Function def

```{r}
run_fit = function(se, design_char, blocking) {
  
  info_df1 = se@meta.data
  info_df1 = droplevels(info_df1)
  
  dat_cur = GetAssayData(se, "counts", assay="RNA")
  dge = DGEList(dat_cur)
  
  ### Filter genes
  to_keep = filterByExpr(dge, group = info_df1[,blocking])
  sprintf("Num genes to keep: %s", sum(to_keep))
  dge = dge[to_keep,]
  
  dge = calcNormFactors(dge,method =c("TMMwsp"))
  
  ### FIT --------------------------------------------------

  design = model.matrix(as.formula(design_char), info_df1)
  colnames(design) = make.names(colnames(design))
  print(colnames(design))
  v1 = voom(dge,design,plot=F, normalize.method=normalize.method)
  corfit = duplicateCorrelation(v1, design, block = info_df1[,blocking])
  fit = lmFit(v1, design, block = info_df1[,blocking], correlation =
                corfit$consensus.correlation)
  return(fit)
}
```

## Load object
```{r}
se_fname = file.path(prefix_data, "seurat_object.qs")
se1 = qread(se_fname)
```

## Load coefs

```{r}
model_fname = file.path(prefix_data, "coefs.rds")
se_coefs = readRDS(model_fname)
```

```{r}
loc_trans = read_delim(file.path(prefix_data, "GCF_005870125.1_lonStrDom2_genomic_loc_synonyms.txt"), delim="\t")
```

## Gene filter
```{r}

redo = F
gene_var_fname = file.path(prefix_data, "gene_variance_stats.rds")
if (redo) {
  dat = GetAssayData(se1, assay = "RNA", slot="data")
  #dat = AverageExpression(se1)
  
  md = FetchData(se1, c("set", "duration.of.experiment", "procedure2", "tags", "position", "index2", "random", "nCount_RNA")) %>%
    rownames_to_column("id")
  
  dat_df_full = dat %>% as.data.frame %>% rownames_to_column(var="gene_id") %>% 
    pivot_longer(-gene_id, names_to="id", values_to="value") %>% 
    left_join(md) %>% 
    mutate(set= factor(set))
  
  dat_df_avg = dat_df_full %>% group_by(gene_id, position, tags) %>% 
    summarize(value=mean(value)) %>%
    ungroup() 
  
  dat_df_var = dat_df_avg %>%
    group_by(tags, gene_id) %>%
    summarize(value_var_pos = LogVMR(value))  %>%
    group_by(gene_id) %>%
    summarize(value_var_pos_mean = mean(value_var_pos))
  
  
  dat_df_var2 = dat_df_avg %>%
    group_by(position, gene_id) %>%
    summarize(value_var_tags = LogVMR(value)) %>%
    ungroup() %>%
    group_by(gene_id) %>%
    summarize(value_var_tags_mean = mean(value_var_tags))
  
  dat_df_stat = dat_df_var %>% left_join(dat_df_var2)
  # dat_df_stat = dat_df_stat %>%
  #   mutate(top_diff = gene_id %in% dat_df$gene_id)
  
  mod = lm(value_var_pos_mean ~ value_var_tags_mean, dat_df_stat)
  dat_df_stat$resid = residuals(mod)
  saveRDS(dat_df_stat, gene_var_fname)
} else {
  dat_df_stat = readRDS(gene_var_fname)
}
```

```{r}
thresh_pos = .3
thresh_tags = 1
thresh_resid = -1
gg = ggplot(dat_df_stat, aes(value_var_tags_mean, resid)) + 
  geom_point() + 
#  geom_abline(slope=mod$coefficients[2], intercept=mod$coefficients[1]-mod$coefficients[1]*6)
  #geom_vline(xintercept=thresh_tags) + 
  geom_hline(yintercept=thresh_resid)
gg

dat_df_stat_thresh = dat_df_stat %>%
  filter(resid > thresh_resid)
```

```{r}
se1 = se1[dat_df_stat_thresh$gene_id,]
```

## Transformations
```{r}
md = se1@meta.data %>% 
  rownames_to_column() %>%
  select(-nsongs_per_day_pre_scale) # from earlier calculation, replacing here

md_tags = md %>%
  #filter(!is.na(nsongs_per_day_pre)) %>% 
  distinct(tags, num_songs_on_euth_date, kl_mean, nsongs_per_day_pre, procedure2, duration.of.experiment) %>%
  mutate(num_songs_on_euth_date_log = log1p(num_songs_on_euth_date),
         num_songs_on_euth_date_log_mean = mean(num_songs_on_euth_date_log),
         num_songs_on_euth_date_log_sd = sd(num_songs_on_euth_date_log),
         num_songs_on_euth_date_log_scale = (num_songs_on_euth_date_log - num_songs_on_euth_date_log_mean) / num_songs_on_euth_date_log_sd,
         num_songs_on_euth_date_log_scale_trunc = case_when(num_songs_on_euth_date_log_scale < -2 ~ -2,
                                                            num_songs_on_euth_date_log_scale > 2 ~ 2,
                                                            TRUE ~ num_songs_on_euth_date_log_scale),
         num_songs_on_euth_date_log_scale_cut2 = factor(as.numeric(num_songs_on_euth_date_log_scale_trunc > 0)),
         num_songs_on_euth_date_log_scale_cut3 = factor(cut(num_songs_on_euth_date_log_scale_trunc, breaks=3, labels=F)),
         num_songs_on_euth_date_log_scale_cut4 = factor(cut(num_songs_on_euth_date_log_scale_trunc, breaks=c(-2,-1,0,1,2), labels=F)),
         
         num_songs_on_euth_date_mean = mean(num_songs_on_euth_date),
         num_songs_on_euth_date_sd = sd(num_songs_on_euth_date),
         num_songs_on_euth_date_scale = (num_songs_on_euth_date - num_songs_on_euth_date_mean) / num_songs_on_euth_date_sd,
         nsongs_per_day_pre = if_else(is.na(nsongs_per_day_pre), mean(nsongs_per_day_pre, na.rm=T), nsongs_per_day_pre),
         nsongs_per_day_pre_scale = (nsongs_per_day_pre - mean(nsongs_per_day_pre)) / sd(nsongs_per_day_pre),
         nsongs_per_day_pre_log = log1p(nsongs_per_day_pre),
         nsongs_per_day_pre_log_sd = sd(nsongs_per_day_pre_log),
         # nsongs_per_day_pre_log = if_else(is.na(nsongs_per_day_pre_log), mean(nsongs_per_day_pre_log, na.rm=T), nsongs_per_day_pre_log),
         nsongs_per_day_pre_log_scale = (nsongs_per_day_pre_log - mean(nsongs_per_day_pre_log)) / nsongs_per_day_pre_log_sd ,
         nsongs_per_day_pre_log_scale_trunc = case_when(nsongs_per_day_pre_log_scale < -2 ~ -2,
                                                        nsongs_per_day_pre_log_scale > 2 ~ 2,
                                                        TRUE ~ nsongs_per_day_pre_log_scale),
         nsongs_per_day_pre_log_scale_cut2 = factor(as.numeric(nsongs_per_day_pre_log_scale_trunc > 0)),
         nsongs_per_day_pre_log_scale_cut3 = factor(cut(nsongs_per_day_pre_log_scale_trunc, breaks=3, labels=F)),
         nsongs_per_day_pre_log_scale_cut4 = factor(cut(nsongs_per_day_pre_log_scale_trunc, breaks=c(-2,-1,0,1,2), labels=F)),
         
         kl_mean_log = log(kl_mean + .1),
         kl_mean_log_mean = mean(kl_mean_log),
         kl_mean_log_sd = sd(kl_mean_log),
         kl_mean_log_scale = (kl_mean_log - kl_mean_log_mean) / kl_mean_log_sd,
         
         #kl_mean_log_scale_cut3 = cut(kl_mean_log_scale, breaks=3, include.lowest = T, labels = F),
         kl_mean_log_scale_trunc = case_when(kl_mean_log_scale < -2 ~ -2,
                                             kl_mean_log_scale > 2 ~ 2,
                                             TRUE ~ kl_mean_log_scale),
         kl_mean_log_scale_cut3 = factor(cut(kl_mean_log_scale_trunc, breaks=3, labels=F)),
         kl_mean_log_scale_cut4 = factor(cut(kl_mean_log_scale_trunc, breaks=c(-2,-1,0,1,2), labels=F)),
         kl_mean_log_scale_cut3_proc2 = paste(procedure2, kl_mean_log_scale_cut3, sep="-"),
         kl_mean_log_scale_cut3_proc2 = factor(kl_mean_log_scale_cut3_proc2, levels=c("intact-1", "intact-2", "deaf-2", "deaf-3"))
  )
```

```{r}
gg= ggplot(md_tags, aes(procedure2, kl_mean_log, color=factor(duration.of.experiment))) + 
  geom_point(position=position_jitter(width = .2)) + 
  geom_hline(yintercept=-.65) + 
  geom_hline(yintercept=-1.3)
gg

gg= ggplot(md_tags, aes(procedure2, kl_mean, color=factor(duration.of.experiment))) + 
  geom_point(position=position_jitter(width = .2))
gg
```

```{r}

md_tags = md_tags %>%
  group_by(procedure2) %>%
  mutate(kl_mean_log_scale_cut2_proc2 = cut(kl_mean_log, breaks=c(-3, -1.3, -.65, 2), labels=F)) %>%
  ungroup() %>%
  mutate(kl_mean_log_scale_cut2_proc2 = paste(procedure2, kl_mean_log_scale_cut2_proc2, sep="-")) %>%
  mutate(kl_mean_log_scale_cut2_proc2 = factor(kl_mean_log_scale_cut2_proc2, levels=c("intact-1", "intact-2", "deaf-2", "deaf-3"))) 
  

table(md_tags %>% select(procedure2, kl_mean_log_scale_cut2_proc2))

md = md %>% 
  select(-num_songs_on_euth_date, -kl_mean, -nsongs_per_day_pre) %>%
  left_join(md_tags) %>%
  mutate(frac_mito_scale = scale(frac_mito),
         tags_position = paste(tags, position, sep="-")) %>%
  column_to_rownames()

se1 = AddMetaData(se1, md)

se1$index2_pool = paste(se1$index2, se1$pool, sep="-")
```

```{r}
md %>% select(position, tags) %>% table()
```

```{r}
md = se1@meta.data
md_tags = md %>% distinct(tags, .keep_all=T)

gg = ggplot(md_tags, aes(num_songs_on_euth_date_log_scale)) + 
  geom_density()
gg

gg = ggplot(md_tags, aes(nsongs_per_day_pre_log_scale)) + 
  geom_density()
gg

gg = ggplot(md_tags, aes(procedure2, kl_mean_log)) + 
  geom_point(position=position_jitter())
gg



gg = ggplot(md_tags, aes(procedure2, kl_mean_log_scale)) + 
  geom_point(position=position_jitter())
gg

gg = ggplot(md_tags, aes(num_songs_on_euth_date_log_scale, nsongs_per_day_pre_scale)) + 
  geom_point()
gg

gg = ggplot(md_tags, aes(num_songs_on_euth_date_log_scale, nsongs_per_day_pre_log_scale)) + 
  geom_point()
gg


cor(md_tags[,c("num_songs_on_euth_date_scale", "nsongs_per_day_pre_scale")])
cor(md_tags[,c("num_songs_on_euth_date_log_scale", "nsongs_per_day_pre_log_scale")])
```

## SVA
```{r}
sv_dir = file.path(prefix_cur, "sva")
dir.create(sv_dir)
sv_fname = file.path(sv_dir, "sva.rds")
vp_fname = file.path(sv_dir, "vp.rds")

dat = as.matrix(GetAssayData(se1, slot = "counts"))
  
redo = F
if (redo) {
  

  
  des_char ="~0 + position +
position:num_songs_on_euth_date_log_scale +
position:kl_mean_log_scale_cut2_proc2 +
position:kl_mean_log_scale_cut2_proc2:num_songs_on_euth_date_log_scale + 
position:nsongs_per_day_pre_log_scale + 
cdr_scaled + frac_mito_scale"
  
  des_char0 = "~0 + position +
cdr_scaled + frac_mito_scale"
  
  mod = model.matrix(as.formula(des_char), se1@meta.data)
  mod0 = model.matrix(as.formula(des_char0), se1@meta.data)
  n.sv = num.sv(dat,mod,method="leek")
  n.sv
  
  svobj = svaseq(dat, mod, mod0)
  sv = svobj$sv
  rownames(sv) = colnames(se1)
  colnames(sv) = paste0("sv", 1:ncol(sv))
  saveRDS(sv, sv_fname)
  
} else {
  sv = readRDS(sv_fname)
}

redo_vp = F
if (redo_vp) {
  ## Variance partition
  library(variancePartition)
  dat_cur = dat
  to_use = rowSums(cpm(dat_cur)>1) >=0.5 * ncol(dat_cur)
  dat_cur1 = dat_cur[to_use,]
  
  info = as.data.frame(cbind(se1@meta.data, sv))
  # info = as.data.frame(se1@meta.data)
  info = info %>%
    mutate(lib_column = factor(lib_column),
           set = factor(set))
  gExpr = DGEList(counts=dat_cur1)
  gExpr = calcNormFactors(gExpr, method = "TMMwsp")
  
  design = model.matrix( ~ 1, info)
  vobjGenes = voom(gExpr, design )
  
  param = SnowParam(workers = 10, type = "SOCK")
  
  form = ~ cdr_scaled + frac_mito_scale + sv1 + sv2 + sv3 + sv4 + sv5 + sv6 + sv7 + sv8 + sv9 + sv10 + (1|lib_column) + (1|set) + (1|index2_pool) +
    (1|position)
  varPart = fitExtractVarPartModel(vobjGenes, form, info, BPPARAM = param )
  
  saveRDS(varPart, vp_fname)
} else {
  varPart = readRDS(vp_fname)
}

vp <- sortCols( varPart )
plotPercentBars( vp[c("CRHBP", "SST", "EGR1", "SLC17A6"),] )
gg = plotVarPart( vp )
gg
save_plot(file.path(sv_dir, "variancePartition_variance.pdf"), gg)

vp %>% as.data.frame() %>% rownames_to_column() %>%
  pivot_longer(-rowname) %>%
  group_by(name) %>%
  summarize(value_median = median(value),
            value_mean = mean(value),
            value_n = sum(value>.1)) %>%
  arrange(desc(value_n))



se1 = AddMetaData(se1, as.data.frame(sv))
```


## Model

```{r}
redo = F

max_sv = 2

outdir1 = file.path(prefix_cur, sprintf("sv%s", max_sv))
dir.create(outdir1)

model_fname = file.path(outdir1, "fit.rds")
formula_fname = file.path(outdir1, "formula.txt")         
normalize.method="none"
blocking="tags_position"

se1_cur = se1

des_char = "~0 + position +
position:num_songs_on_euth_date_log_scale +
position:kl_mean_log_scale_cut2_proc2 +
position:kl_mean_log_scale_cut2_proc2:num_songs_on_euth_date_log_scale + 
position:nsongs_per_day_pre_log_scale + 
cdr_scaled + frac_mito_scale"
if (max_sv > 0) {
  svs = paste(paste0("sv", 1:max_sv), collapse=" + ")
  des_char =paste(des_char, svs, sep=" + ")
}





if (redo) {
  fit = run_fit(se1_cur, des_char, blocking)
  saveRDS(fit, model_fname)
  write_lines(des_char, file = formula_fname)
} else {
  fit = readRDS(model_fname)
}

```
### Standard errors
```{r}
fit1 = eBayes(fit)
coefs = fit1$coefficients[,-1]
std_errors = sqrt(fit1$s2.post) * fit1$stdev.unscaled
std_errors = std_errors[,-1]
cd = data.frame(cn = colnames(coefs)) %>%
  filter(grepl("position", cn)) %>% 
  mutate(cn_s = map(cn, ~unlist(str_split(.x, "\\."))),
         cn_s_len = map_int(cn_s, length)) %>%
  filter(cn_s_len > 1) %>%
  mutate(
    position = map_chr(cn_s, 1),
         position = sub("position", "", position),
    coef_orig = sub("position[a-z]+\\.", "", cn))

coefs = coefs[,cd$cn]
std_errors = std_errors[,cd$cn]
se_coefs = SummarizedExperiment(assays=list(coef=coefs, std_errors = std_errors), colData = cd)
```

## Reduced model
```{r}
# redo = T
# 
# outdir1 = file.path(prefix_cur, sprintf("sv%s", max_sv))
# dir.create(outdir1)
# 
# model_fname = file.path(outdir1, "fit_drop-kl_mean.rds")
# residuals_fname = file.path(outdir1, "seurat_drop-kl_mean_residuals.rds")
# formula_fname = file.path(outdir1, "formula_drop-kl_mean.txt")
# normalize.method="none"
# blocking="tags_position"
# 
# se1_cur = se1
# 
# des_char = paste("~0 + position +
# position:num_songs_on_euth_date_log_scale +
# position:nsongs_per_day_pre_log_scale +
# cdr_scaled + frac_mito_scale",  svs, sep=" + ")
# 
# 
# if (redo) {
#   fit_red = run_fit(se1_cur, des_char, blocking)
#   saveRDS(fit_red, model_fname)
#   write_lines(des_char, path = formula_fname)
# } else {
#   fit_red = readRDS(model_fname)
# }
# 
# dat = GetAssayData(se1_cur)
# dat = dat[rownames(dat) %in% rownames(fit_red$coefficients),]
# res_mat = residuals(fit_red, dat)
# 
# res_ass = CreateAssayObject(data=res_mat)
# se1_cur[["residuals"]] = res_ass
# saveRDS(se1_cur, residuals_fname)
```

```{r}
# 
# to_plot = c("CRHBP", "CORT", "SST", "EPB41L1", "PBX1", "MAN1C1")
# 
# dat_m = melt(res_mat[to_plot,])
# colnames(dat_m) = c("gene_id", "id", "value")
# md = FetchData(se1_cur, c("kl_mean_log_scale_cut2_proc", "position", "procedure2", "tags")) %>%
#   rownames_to_column("id") %>%
#   mutate(id = as.numeric(id))
# dat_m = dat_m %>% left_join(md)
```

```{r}
# dat_m_sum = dat_m %>% group_by(kl_mean_log_scale_cut2_proc, position, tags, gene_id) %>%
#   summarize(value_mean = mean(value),
#             value_sem = sem(value))
# gg = ggplot(dat_m_sum %>% filter(position=="ra"), aes(kl_mean_log_scale_cut2_proc, value_mean)) + 
#   geom_point() +
#   facet_wrap(~gene_id)
# gg
```

## Contrasts

### By position

```{r}
outdir2 = file.path(outdir1, "by_position")
dir.create(outdir2)
results_fname = file.path(outdir2, "model_results.rds")
redo = F

positions = unique(se1_cur$position)
if (redo) {
  
  design = fit$design
  coefs_names = list(
    kl_mean_2_hearing = c(
      "position%s.kl_mean_log_scale_cut2_proc2intact.2"
    ),
    kl_mean_2_deaf = c(
      "position%s.kl_mean_log_scale_cut2_proc2deaf.2"
    ),
    kl_mean_3 = c(
      "position%s.kl_mean_log_scale_cut2_proc2deaf.3"
    ),
    # kl_mean_aov = c(
    #   "position%s.kl_mean_log_scale_cut2_proc2",
    #   "position%s.kl_mean_log_scale_cut2_proc3"
    #   
    # ),
    num_songs_euth_date = c(
      "position%s.num_songs_on_euth_date_log_scale"
      
    ),
    num_songs_pre = c(
      "position%s.nsongs_per_day_pre_log_scale"
      
    ),
    inter = c(
      #  "position%s.num_songs_on_euth_date_log_scale.kl_mean_log_scale_cut2_proc2",
      "position%s.num_songs_on_euth_date_log_scale.kl_mean_log_scale_cut2_proc2deaf.3"
      
    )
  )
  coefs_names_short = map(coefs_names, ~sub("position%s\\.", "", .x))
  coefs_names_short = coefs_names_short[!grepl("aov", names(coefs_names_short))]
  coefs_names_short_df = data.frame(coef_orig = unlist(coefs_names_short), coef = names(coefs_names_short))
  res = map(positions, function(p) {
    print(p)
    map(coefs_names, function(co) {
      print(co)
      co_cur = sprintf(co, p)
      fit2 = contrasts.fit(fit, coefficients = co_cur)   
      fit2 = eBayes(fit2)
      tt = topTable(fit2, p.value=1, number=nrow(se1), confint=T)
      tt = as.data.frame(tt) %>% rownames_to_column(var="gene_id") %>%
        select(gene_id, AveExpr, P.Value, adj.P.Val, one_of("F", "t", "logFC"))
      tt
    }) %>% set_names(names(coefs_names)) %>% bind_rows(.id="coef")
  }) %>% set_names(positions) %>% bind_rows(.id="position")
  
 std_errors = assay(se_coefs, "std_errors")
  std_errors_df = melt(std_errors)
  colnames(std_errors_df) = c("gene_id", "cn", "std_error")
  se_coefs_md = colData(se_coefs) %>% as.data.frame()
  std_errors_df = std_errors_df %>% left_join(se_coefs_md) %>%
    left_join(coefs_names_short_df)
  res = res %>% left_join(std_errors_df %>% select(gene_id, position, coef, std_error))
res = res %>% mutate(position = factor(position, levels=position_levels))
saveRDS(res, results_fname)
} else{
  res = readRDS(results_fname)
}
```
##### Write out flat file
```{r}
flat_fname = file.path(outdir2, "voom_results.csv")
flat_fname_xls = file.path(outdir2, "voom_results.xls")

library(xlsx)
write_csv(res, flat_fname)
write_xl
```


```{r}
tab = res %>% filter(adj.P.Val<.1) %>% group_by(coef, position) %>%
  summarize(n=n()) %>%
  pivot_wider(names_from="coef", values_from="n") %>% arrange(position)

ft = flextable(tab)
ft

save_as_html(ft, path = file.path(outdir2, "deg_table.html"))


tab = res %>% filter(adj.P.Val<.1) %>% 
  mutate(sign = sign(logFC)) %>%
  group_by(coef, position, sign) %>%
  summarize(n=n()) %>%
  pivot_wider(names_from="coef", values_from="n") %>% arrange(position)

ft = flextable(tab)
ft

save_as_html(ft, path = file.path(outdir2, "deg_table_signed.html"))

```

```{r}
# res = res %>% mutate(effect_size = logFC / std_error)
# gg = ggplot(res %>% filter(coef=="kl_mean_3"), aes(effect_size, color=position)) + 
#   geom_density() 
# gg
```

```{r}
# md_tags_stats = md_tags %>%
#   distinct(num_songs_on_euth_date_log_sd, num_songs_on_euth_date_sd, kl_mean_log_sd, nsongs_per_day_pre_log_sd) %>%
#   pivot_longer(everything(), names_to="coef2", values_to="std.dev") %>%
#   mutate(coef2 = sub("_sd", "", coef2))
# 
# res = res %>% mutate(coef2 = sub("position%s.", "", coef)) %>%
#   mutate(coef2 = sub("_scale", "", coef2)) %>% 
#   left_join(md_tags_stats)
# 
# res = res %>% mutate(logFC_rescale = logFC * std.dev)
# 
# res = res %>% left_join(loc_trans) 
# res = res %>% mutate(gene_id = if_else(is.na(synonym), gene_id, synonym))
```

##### MA

```{r}
co ="kl_mean_3"
p = "ra"
res_cur = res %>%
  filter(coef==co, position==p)

gg = ggplot(res_cur, aes(AveExpr, logFC)) + 
  geom_point()+ 
  geom_hline(yintercept=0, color=2)
gg
```

##### Volcano

```{r}
volcano_dir = file.path(outdir2, "volcano")
dir.create(volcano_dir)
```

num_songs_on_euth_date_log_scale

```{r}
# co = "position%s.num_songs_on_euth_date_log_scale"
# co_label = sub("position%s.", "", co)
# res_cur = res %>% filter(coef==co)
# res_cur = res_cur %>% mutate(padjust_signed = -1 * log10(adj.P.Val))
# 
# 
# info2 = se1@meta.data
# pos_pos2_all = info2 %>% distinct(position, position2, position_pretty) %>%
#   mutate(songsystem = as.integer(position %in% c("ra", "hvc", "lman", "x"))) 
# positions = intersect(position_levels, unique(res_cur$position))
# 
# 
# ## Set y max
# res_cur = res_cur %>% left_join(pos_pos2_all %>% select(position, position2))
# y_max = res_cur %>% 
#   group_by(position2) %>%
#   summarize(y_max = max(padjust_signed))
# res_cur = res_cur %>%
#   left_join(y_max)
# 
# ggs = map(positions, function(p) {
#   pos_pos2_all = pos_pos2_all %>% mutate(col = position_colors[match(position, names(position_colors))])
#   pos_pos2_all_cur = pos_pos2_all %>% filter(position==p)
# 
#   tmp = res_cur %>% filter(position==p) %>%
#     mutate(sig = adj.P.Val < .1) %>% 
#     mutate( class = "none",
#                     class = if_else(logFC < 0 & sig, "down" , class),
#                     class = if_else(logFC > 0 & sig, "up", class))
#   
#   colors_cur = pos_pos2_all_cur$col
#   colors_cur = c(up = colors_cur[1], down = colors_cur[1], none="grey50")
# 
#   sig_label = tmp %>% 
#     filter(sig) %>%
#     top_n(-15, wt = adj.P.Val)
#   tmp = tmp %>%
#     mutate(gene_id_sig = ifelse(gene_id %in% sig_label$gene_id, gene_id, NA))
#   if (nrow(tmp) == 0)
#     next
#   gg = ggplot(tmp, aes(logFC_rescale, padjust_signed, color=class)) +
#     geom_hline(yintercept=1, linetype=1, size=.25, color="grey") + 
#     geom_vline(xintercept=0, linetype=1, size=.25, color="grey") + 
#     geom_point(size = .25) +
#     geom_text_repel(aes(label=gene_id_sig), color="black", alpha=1, size=6 / 2.83, segment.size = 0.25) + 
#     scale_color_manual("", values=colors_cur) +
#     scale_y_continuous(limits=c(0, res_cur$y_max[1])) + 
#     labs(x="log fold-change # songs on euthanasia date",
#                 y="-log10(adjusted p-value)") + 
#     theme_bw() + 
#     theme(legend.position="none",
#           axis.title = element_text(size=6),
#           axis.text = element_text(size=5),
#           axis.line = element_line(size=.1),
#           axis.ticks  = element_line(size=.1),
#           panel.grid = element_blank(),
#           ) 
#   gg
#   save_plot(file.path(volcano_dir, sprintf("%s_%s.pdf", co_label, p)), gg, base_height=2, base_aspect_ratio = 1.1)
#   save_plot(file.path(volcano_dir, sprintf("%s_%s.png", co_label, p)), gg, base_height=2, base_aspect_ratio = 1.1)
#   return(gg)
# })
# 
# pg = plot_grid(plotlist=ggs, ncol=2, nrow=4, align="hv")
# save_plot(file.path(volcano_dir, sprintf("combined_%s.pdf", co_label)), pg, ncol=2, nrow=4, base_height=1.5, base_aspect_ratio = 1.1)
# save_plot(file.path(volcano_dir, sprintf("combined_%s.png", co_label)), pg, ncol=2, nrow=4, base_height=1.5, base_aspect_ratio = 1.1)
# pg

```

```{r}
params = list(list(co = "num_songs_euth_date", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
              list(co = "num_songs_pre", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
              list(co = "kl_mean_3", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
              list(co = "kl_mean_2_deaf", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
              list(co = "kl_mean_2_hearing", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
              list(co=  "inter", xvar = "logFC", ymax=5, xlab="log fold-change singing x song destabilization")
             )

walk(params, function(param) {
  co = param[["co"]]
  xvar = param[["xvar"]]
  ymax = param[["ymax"]]
  xlabel = param[["xlab"]]
  co_label = sub("position%s.", "", co)
  res_cur = res %>% filter(coef==co)
  
  
  res_cur = res_cur %>% mutate(padjust_signed = -1 * log10(adj.P.Val))  %>%
    mutate(padjust_signed = if_else(padjust_signed>ymax, ymax, padjust_signed))
  
  
  info2 = se1@meta.data
  pos_pos2_all = info2 %>% distinct(position, position2, position_pretty) %>%
    mutate(songsystem = as.integer(position %in% c("ra", "hvc", "lman", "x"))) %>%
    mutate(col = position_colors[match(position, names(position_colors))])
  positions = intersect(position_levels, unique(res_cur$position))
  
  
  ## Set y max
  res_cur = res_cur %>% left_join(pos_pos2_all %>% select(position, position2))
  y_max = res_cur %>% 
    group_by(position2) %>%
    summarize(y_max = max(padjust_signed))
  res_cur = res_cur %>%
    left_join(y_max)
  
  res_cur = res_cur %>% left_join(pos_pos2_all) %>%
    mutate(color_sig = if_else(padjust_signed>1, col, "grey50")) %>%
    mutate(position = factor(position, levels=position_levels),
           position_pretty = factor(position_pretty, levels=position_pretty_levels))
  
  gg = ggplot(res_cur, aes_string(xvar, "padjust_signed", color="color_sig")) +
    scale_color_identity() + 
    geom_hline(yintercept=1, linetype=1, size=.25, color="grey") + 
    geom_vline(xintercept=0, linetype=1, size=.25, color="grey") + 
    geom_point(size = .25) +
    scale_y_continuous(limits=c(0, ymax)) + 
    facet_wrap(~position_pretty, ncol=2) +
    labs(x=xlabel,
         y="-log10(adjusted p-value)") + 
    theme_bw() + 
    theme(legend.position="none",
          axis.title = element_text(size=6),
          axis.text = element_text(size=5),
          axis.line = element_line(size=.1),
          axis.ticks  = element_line(size=.1),
          strip.background = element_rect(size=1),
          strip.text = element_text(size=6, face="bold", margin = margin(1,0,1,0)),
          panel.grid = element_blank()
    ) 
  gg
  save_plot(file.path(volcano_dir, sprintf("combined_grid_%s_no_text.pdf", co_label)), gg, ncol=2, nrow=4, base_height=1.2, base_aspect_ratio = .8)
})
```

```{r}
params = list(
  list(co = "num_songs_euth_date", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co = "num_songs_pre", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co = "kl_mean_3", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
    list(co = "kl_mean_2", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
  list(co=  "inter", xvar = "logFC", ymax=5, xlab="log fold-change singing x song destabilization")
)

walk(params, function(param) {
  co = param[["co"]]
  xvar = param[["xvar"]]
  ymax = param[["ymax"]]
  xlabel = param[["xlab"]]
  co_label = sub("position%s.", "", co)
  res_cur = res %>% filter(coef==co)
  
  
  res_cur = res_cur %>% mutate(padjust_signed = -1 * log10(adj.P.Val))  %>%
    mutate(padjust_signed = if_else(padjust_signed>ymax, ymax, padjust_signed)) %>%
    mutate(sig = padjust_signed>1)
  
  
  info2 = se1@meta.data
  pos_pos2_all = info2 %>% distinct(position, position2, position_pretty) %>%
    mutate(songsystem = as.integer(position %in% c("ra", "hvc", "lman", "x"))) %>%
    mutate(col = position_colors[match(position, names(position_colors))])
  positions = intersect(position_levels, unique(res_cur$position))
  
  
  ## Set y max
  res_cur = res_cur %>% left_join(pos_pos2_all %>% select(position, position2))
  y_max = res_cur %>% 
    group_by(position2) %>%
    summarize(y_max = max(padjust_signed))
  res_cur = res_cur %>%
    left_join(y_max)
  
  res_cur = res_cur %>% left_join(pos_pos2_all) %>%
    mutate(color_sig = if_else(padjust_signed>1, col, "grey50")) %>%
    mutate(position = factor(position, levels=position_levels),
           position_pretty = factor(position_pretty, levels=position_pretty_levels))
  
  
  sig_label = res_cur %>% 
    filter(sig) %>%
    group_by(position) %>% 
    top_n(-10, wt = adj.P.Val) %>%
      mutate(gene_id_sig = gene_id) %>%
    select(position, gene_id_sig, gene_id)
  res_cur = res_cur %>% left_join(sig_label) %>%
    mutate(point_color = if_else(is.na(gene_id_sig), color_sig, "black"))

  
  gg = ggplot(res_cur, aes_string(xvar, "padjust_signed", fill="color_sig", color="point_color")) +
    scale_color_identity() + 
    scale_fill_identity() + 
    geom_hline(yintercept=1, linetype=1, size=.25, color="grey") + 
    geom_vline(xintercept=0, linetype=1, size=.25, color="grey") + 
    geom_point(size = .25, shape=21) +
    geom_text_repel(aes(label=gene_id_sig), size=5/2.83, segment.size = .25) + 
    scale_y_continuous(limits=c(0, ymax)) + 
    facet_wrap(~position_pretty, ncol=2) +
    labs(x=xlabel,
         y="-log10(adjusted p-value)") + 
    theme_bw() + 
    theme(legend.position="none",
          axis.title = element_text(size=6),
          axis.text = element_text(size=5),
          axis.line = element_line(size=.1),
          axis.ticks  = element_line(size=.1),
          strip.background = element_rect(size=1, linetype = "blank"),
          strip.text = element_text(size=6, face="bold", margin = margin(1,0,1,0)),
          panel.grid = element_blank()
    ) 
  gg
  save_plot(file.path(volcano_dir, sprintf("combined_grid_%s_text.pdf", co_label)), gg, ncol=2, nrow=4, base_height=1.2, base_aspect_ratio = .8)
})
```

```{r}
params = list(
  list(co = "num_songs_euth_date", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co = "num_songs_pre", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co = "kl_mean_3", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
    list(co = "kl_mean_2", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
  list(co=  "inter", xvar = "logFC", ymax=5, xlab="log fold-change singing x song destabilization")
)

positions = as.character(positions)
walk(positions, function(p) {
walk(params, function(param) {
  co = param[["co"]]
  xvar = param[["xvar"]]
  ymax = param[["ymax"]]
  xlabel = param[["xlab"]]
  co_label = sub("position%s.", "", co)
  res_cur = res %>% filter(coef==co)
  
  
  res_cur = res_cur %>% 
    filter(position==p) %>%
    mutate(padjust_signed = -1 * log10(adj.P.Val))  %>%
    mutate(padjust_signed = if_else(padjust_signed>ymax, ymax, padjust_signed)) %>%
    mutate(sig = padjust_signed>1)
  
  
  info2 = se1@meta.data
  pos_pos2_all = info2 %>%
    distinct(position, position2, position_pretty) %>%
    mutate(songsystem = as.integer(position %in% c("ra", "hvc", "lman", "x"))) %>%
    mutate(col = position_colors[match(position, names(position_colors))])
  positions = intersect(position_levels, unique(res_cur$position))
  
  
  ## Set y max
  res_cur = res_cur %>% 
    left_join(pos_pos2_all %>% select(position, position2))
  y_max = res_cur %>% 
    group_by(position2) %>%
    summarize(y_max = max(padjust_signed))
  res_cur = res_cur %>%
    left_join(y_max)
  
  res_cur = res_cur %>% left_join(pos_pos2_all) %>%
    mutate(color_sig = if_else(padjust_signed>1, col, "grey50")) %>%
    mutate(position = factor(position, levels=position_levels),
           position_pretty = factor(position_pretty, levels=position_pretty_levels))
  
  
  sig_label = res_cur %>% 
    filter(sig) %>%
    group_by(position) %>% 
    top_n(-10, wt = adj.P.Val) %>%
      mutate(gene_id_sig = gene_id) %>%
    select(position, gene_id_sig, gene_id)
  res_cur = res_cur %>% left_join(sig_label) %>%
    mutate(point_color = if_else(is.na(gene_id_sig), color_sig, "black"))

  
  gg = ggplot(res_cur, aes_string(xvar, "padjust_signed", fill="color_sig", color="point_color")) +
    scale_color_identity() + 
    scale_fill_identity() + 
    geom_hline(yintercept=1, linetype=1, size=.25, color="grey") + 
    geom_vline(xintercept=0, linetype=1, size=.25, color="grey") + 
    geom_point(size = .25, shape=21) +
    geom_text_repel(aes(label=gene_id_sig), size=5/2.83, segment.size = .25) + 
    scale_y_continuous(limits=c(0, ymax)) + 
    #facet_wrap(~position_pretty, ncol=2) +
    labs(x=xlabel,
         y="-log10(adjusted p-value)") + 
    theme_bw() + 
    theme(legend.position="none",
          axis.title = element_text(size=6),
          axis.text = element_text(size=5),
          axis.line = element_line(size=0),
          axis.ticks  = element_line(size=.1),
          strip.background = element_rect(size=1, linetype = "blank"),
          strip.text = element_text(size=6, face="bold", margin = margin(1,0,1,0)),
          panel.grid = element_blank()
    ) 
  gg
  save_plot(file.path(volcano_dir, sprintf("%s_%s_text.pdf", p, co_label)), gg, ncol=1, nrow=1, base_height=1.2, base_aspect_ratio = .8)
})
})
```

##### DEG score
```{r}
deg_dir = file.path(outdir2, "deg_scores")
dir.create(deg_dir)
```

```{r}
params = list(
  list(co = "kl_mean_3", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance")#,
 # list(co = "kl_mean_2_deaf", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance")#,
#       list(co = "kl_mean_2_hearing", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance")
  )
  
  res_sum = res %>%
  mutate(padjust_signed = -1 * log10(adj.P.Val)) %>%
  filter(padjust_signed>1) %>%
  mutate(sign = sign(logFC)) %>%
  ungroup() %>% 
  group_by(coef, position, sign) %>%
  summarize(score = sum(padjust_signed)) %>%
  ungroup() %>% 
  mutate(score_sign = score * sign) %>%
  mutate(position_sign = ifelse(sign>0, as.character(position), NA))

coefs_to_plot = map_chr(params, "co")
res_sum_to_plot = res_sum %>% filter(coef %in% coefs_to_plot )
min_y = min(res_sum_to_plot$score_sign)
max_y = max(res_sum_to_plot$score_sign)

walk(params, function(param) {
  co = param[["co"]]
  
co_label = sub("position%s.", "", co)
res_cur = res_sum %>% 
  filter(coef==co) %>%
  mutate(position = factor(position, levels=c("ra", "arco", "hvc", "ncl", "lman", "nido", "x", "stri"))) %>%
  complete(position, fill=list(score = 0, score_sign = 0))

gg = ggplot(res_cur, aes(position, score_sign)) + 
  geom_bar(stat="identity", aes(color=position, fill=position_sign), size=.25) + 
  scale_fill_manual(name = "",values=position_colors, guide=F, na.value=NA) + 
  scale_color_manual(values=position_colors, guide=F) + 
  ylim(min_y, max_y) + 
  labs(y="DEG score", title=co) + 
  theme_bw() + 
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid.major.x=element_blank(),  
        panel.grid.minor.y = element_blank())
print(gg)
save_plot(file.path(deg_dir, sprintf("%s.pdf" ,co_label)), gg, base_height = 1, base_width=1.2)
})
```
Effect size (not a good representation)
```{r}
params = list(
  list(co = "kl_mean_3", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance")#,
 # list(co = "kl_mean_2_deaf", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance")#,
#       list(co = "kl_mean_2_hearing", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance")
  )
  
nselect = 100
res_sum = res %>%
  group_by(position, coef) %>% 
  top_n(-1 * nselect, P.Value) %>%
  #mutate(padjust_signed = -1 * log10(adj.P.Val)) %>%
  #filter(padjust_signed>1) %>%
  mutate(sign = sign(effect_size)) %>%
  ungroup() %>% 
  group_by(coef, position, sign) %>%
  summarize(score = sum(effect_size)) %>%
  ungroup() %>% 
  mutate(score_sign = score * sign) %>%
  mutate(position_sign = ifelse(sign>0, as.character(position), NA))

coefs_to_plot = map_chr(params, "co")
res_sum_to_plot = res_sum %>% filter(coef %in% coefs_to_plot )
min_y = min(res_sum_to_plot$score_sign)
max_y = max(res_sum_to_plot$score_sign)

walk(params, function(param) {
  co = param[["co"]]
  
  co_label = sub("position%s.", "", co)
  res_cur = res_sum %>% 
    filter(coef==co) %>%
    mutate(position = factor(position, levels=c("ra", "arco", "hvc", "ncl", "lman", "nido", "x", "stri"))) %>%
  complete(position, fill=list(score = 0, score_sign = 0))

gg = ggplot(res_cur, aes(position, score_sign)) + 
  geom_bar(stat="identity", aes(color=position, fill=position_sign), size=.25) + 
  scale_fill_manual(name = "",values=position_colors, guide=F, na.value=NA) + 
  scale_color_manual(values=position_colors, guide=F) + 
  ylim(min_y, max_y) + 
  labs(y="DEG score", title=co) + 
  theme_bw() + 
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid.major.x=element_blank(),  
        panel.grid.minor.y = element_blank())
print(gg)
save_plot(file.path(deg_dir, sprintf("effect-size_%s.pdf" ,co_label)), gg, base_height = 1, base_width=1.2)
})
```

```{r}
params = list(
  list(co = "num_songs_euth_date", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co = "num_songs_pre", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co=  "inter", xvar = "logFC", ymax=5, xlab="log fold-change singing x song destabilization")
)

res_sum = res %>%
  mutate(padjust_signed = -1 * log10(adj.P.Val)) %>%
  filter(padjust_signed>1) %>%
  mutate(sign = sign(logFC)) %>%
  ungroup() %>% 
  group_by(coef, position, sign) %>%
  summarize(score = sum(padjust_signed)) %>%
  ungroup() %>% 
  mutate(score_sign = score * sign) %>%
  mutate(position_sign = ifelse(sign>0, as.character(position), NA))

coefs_to_plot = map_chr(params, "co")
res_sum_to_plot = res_sum %>% filter(coef %in% coefs_to_plot )
min_y = min(res_sum_to_plot$score_sign)
max_y = max(res_sum_to_plot$score_sign)

walk(params, function(param) {
  co = param[["co"]]
  
co_label = sub("position%s.", "", co)
res_cur = res_sum %>% 
  filter(coef==co) %>%
  mutate(position = factor(position, levels=c("ra", "arco", "hvc", "ncl", "lman", "nido", "x", "stri"))) %>%
  complete(position, fill=list(score = 0, score_sign = 0))

gg = ggplot(res_cur, aes(position, score_sign)) + 
  geom_bar(stat="identity", aes(color=position, fill=position_sign), size=.25) + 
  scale_fill_manual(name = "",values=position_colors, guide=F, na.value=NA) + 
  scale_color_manual(values=position_colors, guide=F) + 
  ylim(min_y, max_y) + 
  labs(y="DEG score", title=co) + 
  theme_bw() + 
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid.major.x=element_blank(), 
        panel.grid.minor.y = element_blank())
print(gg)
save_plot(file.path(deg_dir, sprintf("%s.pdf" ,co_label)), gg, base_height = 1, base_width=1.2)
}) 
```

##### Heatmap

```{r}
heat_dir = file.path(outdir2, "heatmaps")
dir.create(heat_dir)
```

```{r}

coefs_to_use = unique(res$coef)
coefs_to_use = c("kl_mean_3", "num_songs_euth_date", "inter", "num_songs_pre")
cols = list("RdBu", "PuOr", "BrBG", "RdBu")
names(cols) = coefs_to_use

res_mean = res %>% group_by(coef, position, gene_id) %>% 
  summarize()
walk(coefs_to_use, function(co) {
  print(co)
  res_sig = res %>%
    filter(coef==co) %>% 
    #group_by(position) %>%
    filter(adj.P.Val<.1) #%>%
    #top_n(50, abs(logFC))
    #top_n(-50, adj.P.Val)
  
  
  #value_var = ifelse(co %in% c("num_songs_euth_date", "position%s.kl_mean_log_scale" ), "logFC_rescale", "logFC")
  value_var = "logFC"
  res_cur = res %>% filter(gene_id %in% res_sig$gene_id) %>% filter(coef==co) %>%
    group_by(position, gene_id) %>%
    summarize(logFC = mean(logFC),
              logFC_rescale = mean(logFC_rescale))
  mat = acast(res_cur, gene_id~position, value.var=value_var)
  colnames(mat) = toupper(colnames(mat))
  
  mat_var = apply(mat, 1, var)
  #mat = mat[mat_var>quantile(mat_var, .25),]
  width = .75
  height = width * nrow(mat) / ncol(mat)
  
  max_val = max(c(abs(min(mat)), max(mat)))
  max_val = 2
  length_out = 9
  cols_cur = colorRamp2(breaks=seq(-1*max_val, max_val, length.out=length_out), colors = rev(brewer_pal(palette=cols[[co]])(length_out)))
  hm = Heatmap(mat, col=cols_cur,
               clustering_method_rows = "ward.D", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("%s_top_fc_heatmap.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  #hm
  
  hm_t = Heatmap(t(mat), col=cols_cur,
               clustering_method_columns  = "average", 
               cluster_rows = F,
               height = unit(width, "in"),
               width =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
  
    pdf(file.path(heat_dir, sprintf("%s_top_fc_heatmap_horiz.pdf", co)), width=height+2, height=width+2)
  print(hm_t)
  dev.off()
  #hm
})
```

```{r}

coefs_to_use = unique(res$coef)
coefs_to_use = c("kl_mean_3", "num_songs_euth_date", "inter", "num_songs_pre")
cols = list("RdBu", "PuOr", "BrBG", "RdBu")
names(cols) = coefs_to_use

res_mean = res %>% group_by(coef, position, gene_id) %>% 
  summarize()
walk(coefs_to_use, function(co) {
  print(co)
  res_sig = res %>%
    filter(coef==co) %>% 
    filter(position %in% c("ra", "hvc", "lman", "x")) %>% 
    #group_by(position) %>%
    filter(adj.P.Val<.1) #%>%
    #top_n(50, abs(logFC))
    #top_n(-50, adj.P.Val)
  
  

  value_var = "logFC"
  res_cur = res %>% filter(gene_id %in% res_sig$gene_id) %>% filter(coef==co) %>%
    group_by(position, gene_id) %>%
    summarize(logFC = mean(logFC),
              logFC_rescale = mean(logFC_rescale))
  mat = acast(res_cur, gene_id~position, value.var=value_var)
  mat = mat[,c("ra", "hvc", "lman", "x")]
  colnames(mat) = toupper(colnames(mat))
  
  mat_var = apply(mat, 1, var)
  #mat = mat[mat_var>quantile(mat_var, .25),]
  width = .25
  height = width * nrow(mat) / ncol(mat)
  
  max_val = max(c(abs(min(mat)), max(mat)))
  max_val = 2
  #length_out = 9
 # cols_cur = colorRamp2(breaks=seq(-1*max_val, max_val, length.out=length_out), colors = rev(brewer_pal(palette=cols[[co]])(length_out)))
  cols_cur = colorRamp2(breaks=c(-1*max_val, 0, max_val), colors = c(muted("blue"), "white", muted("red")))
  hm = Heatmap(mat, col=cols_cur,
               clustering_method_rows = "ward.D", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=5), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=5)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("heatmap_%s_top_fc_song-only.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  #hm
  
  hm_t = Heatmap(t(mat), col=cols_cur,
               clustering_method_columns  = "average", 
               cluster_rows = F,
               height = unit(width, "in"),
               width =unit( height, "in"),
               column_names_gp = gpar(fontsize=5), column_names_side = "top",
               row_names_gp = gpar(fontsize=5)
               #  clustering_method_columns = "ward.D"
  )
  
    pdf(file.path(heat_dir, sprintf("heatmap_%s_top_fc_song-only_horiz.pdf", co)), width=height+2, height=width+2)
  print(hm_t)
  dev.off()
  #hm
})
```


```{r}
coefs_to_use = unique(res$coef)
coefs_to_use = c("kl_mean_3", "num_songs_euth_date", "inter", "num_songs_pre")
cols = list("RdBu", "PuOr", "BrBG", "RdBu")
names(cols) = coefs_to_use

#res_mean = res %>% group_by(coef, position, gene_id) %>% 
#  summarize()


res = res %>%
  mutate(songsystem = case_when(position %in% c("ra", "hvc", "lman", "x") ~ TRUE,
                                TRUE ~ FALSE),
         position2 = case_when(position %in% c("ra", "arco") ~ "ra",
                               position %in% c("hvc", "ncl") ~ "hvc",
                               position %in% c("lman", "nido") ~ "lman",
                               position %in% c("x", "stri") ~ "x"))

walk(coefs_to_use, function(co) {
  print(co)
  res_sig = res %>%
    filter(coef==co) %>% 
    filter(position %in% c("ra", "hvc", "lman", "x")) %>% 
    #group_by(position) %>%
    filter(adj.P.Val<.1) #%>%
    #top_n(50, abs(logFC))
    #top_n(-50, adj.P.Val)
  
  
  
  #value_var = ifelse(co %in% c("num_songs_euth_date", "position%s.kl_mean_log_scale" ), "logFC_rescale", "logFC")
  value_var = "logFC_rel"
  res_cur = res %>% filter(gene_id %in% res_sig$gene_id) %>%
    filter(coef==co) %>%
    group_by(position, gene_id, position2, songsystem) %>%
    summarize(logFC = mean(logFC),
              logFC_rescale = mean(logFC_rescale)) %>%
    group_by(position2, gene_id) %>%
    mutate(logFC_rel = logFC - logFC[!songsystem])
  
  mat = acast(res_cur, gene_id~position, value.var=value_var)
    mat = mat[,c("ra", "hvc", "lman", "x")]
  colnames(mat) = toupper(colnames(mat))

  #mat_var = apply(mat, 1, var)
  #mat = mat[mat_var>quantile(mat_var, .25),]
  width = .25
  height = width * nrow(mat) / ncol(mat)
  
  max_val = max(c(abs(min(mat)), max(mat)))
  max_val = 2
  length_out = 9
 #cols_cur = colorRamp2(breaks=seq(-1*max_val, max_val, length.out=length_out), colors = rev(brewer_pal(palette=cols[[co]])(length_out)))
    cols_cur = colorRamp2(breaks=c(-1*max_val, 0, max_val), colors = c(muted("blue"), "white", muted("red")))
  hc_row = hclust(dist(mat, method="euclidean"), method = "ward.D2")
  
  hm = Heatmap(mat, col=cols_cur,
               cluster_rows = hc_row,
               cluster_columns = F,
               width = unit(width, "in"),
               height = unit( height, "in"),
               column_names_gp = gpar(fontsize=5), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=5)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("heatmap_%s_top_fc_norm-surround.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  #hm
  
  hm_t = Heatmap(t(mat), col=cols_cur,
               clustering_method_columns  = "average", 
               cluster_rows = F,
               height = unit(width, "in"),
               width =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
  
    pdf(file.path(heat_dir, sprintf("heatmap_%s_top_fc_norm-surround_horiz.pdf", co)), width=height+2, height=width+2)
  print(hm_t)
  dev.off()
  #hm
})
```

```{r}

walk(unique(res$coef), function(co) {
  print(co)
  res_sig = res %>%
    filter(coef==co) %>% 
    group_by(position) %>%
        filter(adj.P.Val<.1)  %>%
    top_n(-20, adj.P.Val)

  
  value_var = "t"
  mat = acast(res %>% filter(gene_id %in% res_sig$gene_id) %>% filter(coef==co), gene_id~position, value.var=value_var)
  colnames(mat) = toupper(colnames(mat))
  
  #mat_var = apply(mat, 1, var)
  #mat = mat[mat_var>quantile(mat_var, .5),]
  width = .75
  height = width * nrow(mat) / ncol(mat)
  
  
  hm = Heatmap(mat, 
               clustering_method_rows = "average", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("%s_top_t_heatmap.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  hm
})
```

```{r}
walk(unique(res$coef), function(co) {

  res_sig = res %>%
    filter(position %in% c("ra", "hvc", "lman", "x")) %>% 
    filter(coef==co) %>% 
    group_by(position) %>%
    filter(adj.P.Val<.1) %>%
    top_n(10, abs(logFC))

  res_cur = res %>% filter(gene_id %in% res_sig$gene_id) %>% filter(coef==co) %>%
     filter(position %in% c("ra", "hvc", "lman", "x"))
  mat = acast(res_cur, gene_id~position, value.var="logFC")
  colnames(mat) = toupper(colnames(mat))
  
  width = .75
  height = width * nrow(mat) / ncol(mat)
  
  
  hm = Heatmap(mat, clustering_method_rows = "ward.D", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("%s_top_fc_songsystem_heatmap.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  hm
})
```

```{r}
#mat = scale(mat, scale = T)
km = kmeans(mat, centers = 10)
res = res %>% mutate(k = km$cluster[match(gene_id, names(km$cluster))])
res_sum = res %>% 
  group_by(k, position) %>% 
  summarize(logFC_mean = mean(logFC)) %>%
  mutate(position = factor(position, levels=position_levels))
gg = ggplot(res_sum, aes(position, logFC_mean)) + 
  geom_point()+ 
  theme_cowplot() + 
  facet_wrap(~k)
gg

gg = ggplot(res %>% filter(gene_id %in% c("CRHBP", "PMP2", "TRAP1")), aes(position, logFC)) + 
  geom_point()+ 
  theme_cowplot() + 
  facet_wrap(~gene_id)
gg
```

##### Correlations

```{r}
cor_dir = file.path(outdir2, "correlations")
dir.create(cor_dir)
```

```{r}

coefs_to_use = c("kl_mean_2_hearing", "kl_mean_2_deaf", "kl_mean_3")

ngenes = 200
walk(coefs_to_use, function(coef_to_use) {
res_to_use = res %>%
   filter(coef==coef_to_use) %>%
  #filter(adj.P.Val<.1)
   top_n(-1 * ngenes, P.Value)

res_select = res %>% 
  filter(coef==coef_to_use) %>%
  filter(gene_id %in% res_to_use$gene_id)


res_df = res_select %>%
    select(gene_id, position, logFC) %>% 
  pivot_wider(names_from = "position", values_from="logFC") 

positions = as.character(positions)
walk(positions, function(p) {
  positions_sub = positions[!grepl(p, positions)]
  walk(positions_sub, function(p2) {
    # gg = ggplot(res_df, aes_string(p, p2)) + 
    #   geom_hline(yintercept=0, size=.5) + 
    #   geom_vline(xintercept=0, size=.5) +
    #   geom_point(size = .1, shape=20) +
    #   stat_smooth(method="lm", se=F) + 
    #   theme_bw() + 
    #   theme(
    #     legend.position="none",
    #     #strip.text = element_text(size=6, face="italic", hjust=0, vjust=1),
    #     #strip.background = element_blank(),
    #     axis.title.y = element_text(size=6),
    #     axis.title.x = element_blank(),
    #     axis.text.y = element_text(size=5),
    #     axis.text.x = element_blank(),
    #     axis.line = element_blank(),
    #     axis.ticks.y = element_line(size=.25),
    #     axis.ticks.x = element_blank(),
    #     panel.grid=element_blank()
    #   )
    # gg
    # save_plot(file.path(cor_dir, sprintf("scatter_%s_%s_%s.pdf", coef_to_use, p, p2)), gg, base_height=1, base_width=1)
    data.frame(p = p, p2 = p2, value = cor(res_df[,c(p, p2)], method=method)[1,2])
  })
})


res_mat = res_df %>%
  column_to_rownames(var="gene_id") %>%
  as.matrix()

res_cor = cor(res_mat)
res_cor[res_cor>.99] = NA

positions = intersect(position_levels, colnames(res_mat))
res_cor = res_cor[positions, positions]
cor_df = melt(res_cor)


#Shuffle

# 
# nshuf = 1000
# npositions = length(positions)
# 
# res_cur = res %>% filter(coef==coef_to_use)
# genes = unique(res$gene_id)
# num_genes = nrow(res_to_use)
# 
# cor_shuf = map_dfr(1:nshuf, function(k) {
#   genes_to_use = sample(genes, num_genes )
#   # top_n(-1000, P.Value)
#   
#   res_select = res_cur %>% 
#     filter(gene_id %in% genes_to_use)
#   
#   res_df = res_select %>%
#     select(gene_id, position, logFC) %>% 
#     pivot_wider(names_from = "position", values_from="logFC") 
#   
#   res_mat = res_df %>%
#     column_to_rownames(var="gene_id") %>%
#     as.matrix()
#   
#   res_cor = cor(res_mat)
#   melt(res_cor)
# })
# 
# cor_shuf_stat = cor_shuf %>% 
#   rename(cor = value) %>% 
#   group_by(Var1, Var2) %>%
#   summarize(cor_rand_mean = mean(cor),
#                 cor_rand_sd = sd(cor),
#                 cor_rand_q99 = quantile(cor, .99),
#                 cor_rand_q01 = quantile(cor, .01),
#                 cor_rand_q95 = quantile(cor, .95),
#                 cor_rand_q05 = quantile(cor, .05))
# 
# cor_df1 = cor_df %>% 
#   left_join(cor_shuf_stat) %>%
#   rename(cor = value) %>%
#   mutate(cor_z = (cor - cor_rand_mean ) / cor_rand_sd)
# 
# res_cor_z = cor_df1 %>% 
#   select(Var1, Var2, cor_z) %>%
#   pivot_wider(names_from = "Var2", values_from = "cor_z") %>%
#   column_to_rownames("Var1") %>%
#   as.matrix()

height = width = 1

cols = colorRamp2(breaks=seq(0, .7, length.out=11), colors=viridis_pal()(11))
hm = Heatmap(res_cor, col=cols,
             cluster_columns = F,
             cluster_rows = F,
              height = unit(height, "in"),
             width = unit(width, "in"))
hm
pdf(file.path(cor_dir, sprintf("position_correlations_%s.pdf", coef_to_use)), height = height +2, width = width+2)
print(hm)
dev.off()


res_cor_mean = mean(res_cor, na.rm=T)
res_cor_sd = sd(res_cor, na.rm=T)
res_cor_scale = (res_cor - res_cor_mean) #/ res_cor_sd
cols = colorRamp2(breaks=seq(-2, 2, length.out=11), colors=viridis_pal()(11))
cols = colorRamp2(breaks=c(min(res_cor_scale, na.rm=T), 0, max(res_cor_scale, na.rm=T)), colors=c(muted("blue"), "white", muted("red")))


hm = Heatmap(res_cor_scale, col=cols, na_col = "white",
             cluster_columns = F,
             cluster_rows = F,
             height = unit(height, "in"),
             width = unit(width, "in"),
              column_names_gp = gpar(fontsize=5), 
             column_names_side = "top",
               row_names_gp = gpar(fontsize=5)
             )
hm

pdf(file.path(cor_dir, sprintf("position_correlations_%s_mean-sub.pdf", coef_to_use)), height = height +2, width = width+2)
print(hm)
dev.off()
})

# res_cor_df = melt(res_cor_scale)
# walk(positions, function(p) {
#   res_cor_df_cur = res_cor_df %>% filter(Var1 == p)
#   
#   gg = ggplot(res_cor_df_cur, aes(Var2, value, fill=Var2)) + 
#     geom_bar(stat = "identity") + 
#     scale_fill_manual(values=position_colors) + 
#     theme_classic()
#   print(gg)
# })
```

Signed p-adjust
Plotting

```{r}
coefs_to_use = c("kl_mean_2_hearing", "kl_mean_2_deaf", "kl_mean_3")
features_to_use = c("logFC", "padjust_signed")
n_to_use = 200

res = res %>%
  mutate(padjust_signed = sign(logFC) * -1 * log10(adj.P.Val))

walk(features_to_use, function(feature_to_use) {
  walk(coefs_to_use , function(coef_to_use) {
    cor_dir_cur = file.path(cor_dir, coef_to_use)
    dir.create(cor_dir_cur)
  
    res_select = res %>%
      filter(coef==coef_to_use) %>%
      group_by(position) %>% 
    top_n(-1 * n_to_use, P.Value)

    positions = as.character(positions)
    res_cors = map(positions, function(p) {
      print(p)
      positions_sub = positions[!grepl(p, positions)]
      res_cor_df1 = map(positions_sub, function(p2) {
        
        res_select_cur = res_select %>% 
          filter(position %in% c(p, p2)) %>%
          distinct(gene_id)
        
        res_df = res_select %>%
          filter(gene_id %in% res_select_cur$gene_id) %>% 
          distinct(gene_id, position, .keep_all=T) %>%
          select(gene_id, position, !!as.symbol(feature_to_use)) %>% 
          pivot_wider(names_from = "position", values_from=feature_to_use)
        
        gg = ggplot(res_df, aes_string(p, p2)) +
          geom_hline(yintercept=0, size=.25) +
          geom_vline(xintercept=0, size=.25) +
          geom_point(size = .1, shape=20) +
          theme_bw() +
          theme(
            legend.position="none",
            axis.title = element_text(size=6),
            axis.text = element_text(size=5),
            axis.ticks = element_line(size=.25),
            panel.grid=element_blank()
          ) + 
          lims(x=c(-2.5, 2.5), y=c(-2.5, 2.5)) +
          coord_equal()
        gg
        save_plot(file.path(cor_dir_cur, sprintf("scatter_%s_%s_%s_%s.pdf", feature_to_use, coef_to_use, p, p2)), gg, base_height=1, base_width=1)
        gg = gg + stat_smooth(method="lm", se=F, size=.5)
        save_plot(file.path(cor_dir_cur, sprintf("scatter_%s_%s_%s_%s_line.pdf", feature_to_use,coef_to_use, p, p2)), gg, base_height=1, base_width=1)
      })
    }) 
  })
})
```

```{r}
coefs_to_use = c("kl_mean_2_hearing", "kl_mean_2_deaf", "kl_mean_3")
features_to_use = c("logFC", "padjust_signed")
n_to_use = 500
method = "spearman"

res = res %>%
  mutate(padjust_signed = sign(logFC) * -1 * log10(adj.P.Val))

walk(features_to_use, function(feature_to_use) {
  map(coefs_to_use , function(coef_to_use) {
    cor_dir_cur = file.path(cor_dir, coef_to_use)
    dir.create(cor_dir_cur)
    
    res_select = res %>%
      filter(coef==coef_to_use) #%>%
      #group_by(position) %>% 
      #top_n(-1 * n_to_use, P.Value)
    
     # res_df = res %>%
     #      filter(gene_id %in% res_select$gene_id) %>% 
     #      distinct(gene_id, position, .keep_all=T) %>%
     #      select(gene_id, position, !!as.symbol(feature_to_use)) %>% 
     #      pivot_wider(names_from = "position", values_from=feature_to_use)
     # 
    #   
    # res_select = res %>%
    #   filter(coef==coef_to_use) %>%
    #   mutate(padjust_signed = sign(logFC) * -1 * log10(adj.P.Val))
    # 
    #res_df = res %>%
    #    select(gene_id, position, logFC) %>%
    #  pivot_wider(names_from = "position", values_from="logFC")
    positions = as.character(positions)
    res_cors = map(positions, function(p) {
      print(p)
      positions_sub = positions[!grepl(p, positions)]
      res_cor_df1 = map(positions_sub, function(p2) {
        #print(p2)
        res_select_cur = res_select %>%
          filter(position %in% c(p, p2)) %>%
          group_by(position) %>%
         # top_n(n_to_use, abs(padjust_signed)) %>%
          top_n(-1 * n_to_use, P.Value) %>% 
          distinct(gene_id)
        
        # res_select_cur = res_select %>% 
        #   filter(position %in% c(p, p2)) %>%
        #   distinct(gene_id)
          
        res_df = res_select %>%
          filter(position %in% c(p, p2)) %>% 
          filter(gene_id %in% res_select_cur$gene_id) %>% 
          distinct(gene_id, position, .keep_all=T) %>%
          select(gene_id, position, !!as.symbol(feature_to_use)) %>% 
          pivot_wider(names_from = "position", values_from=feature_to_use)
        
        data.frame(p = p, p2 = p2, value = cor(res_df[,c(p, p2)], method=method)[1,2])
      }) %>% bind_rows()
    }) %>% bind_rows()
    
    position_levels_cur = intersect(position_levels, positions)
    res_cors = res_cors %>%
      mutate(p = factor(p, levels=position_levels_cur),
             p2 = factor(p2, levels=position_levels_cur)) %>%
      arrange(p, p2)
    
    res_cor = res_cors %>% 
      pivot_wider(names_from = "p2", values_from = "value") %>%
      column_to_rownames("p") %>%
      as.matrix()
    
    res_cor = res_cor[,position_levels_cur]
    cor_df = melt(res_cor)
    
    height = width = 1
    
    cols = colorRamp2(breaks=seq(0, .7, length.out=11), colors=viridis_pal()(11))
    hm = Heatmap(res_cor, col=cols, na_col = "white",
                 cluster_columns = F,
                 cluster_rows = F,
                 height = unit(height, "in"),
                 width = unit(width, "in"))
    hm
    pdf(file.path(cor_dir_cur, sprintf("position_correlations_%s_%s_%s.pdf", feature_to_use, coef_to_use, method)), height = height +2, width = width+2)
    print(hm)
    dev.off()
    
    
    res_cor_mean = mean(res_cor, na.rm=T)
    res_cor_sd = sd(res_cor, na.rm=T)
    res_cor_scale = (res_cor - res_cor_mean) #/ res_cor_sd
    cols = colorRamp2(breaks=seq(-2, 2, length.out=11), colors=viridis_pal()(11))
    cols = colorRamp2(breaks=c(min(res_cor_scale, na.rm=T), 0, max(res_cor_scale, na.rm=T)), colors=c(muted("blue"), "white", muted("red")))
    
    
    hm = Heatmap(res_cor_scale, col=cols, na_col = "white",
                 cluster_columns = F,
                 cluster_rows = F,
                 height = unit(height, "in"),
                 width = unit(width, "in"),
                 column_names_gp = gpar(fontsize=5),
                 column_names_side = "top",
                 row_names_gp = gpar(fontsize=5)
    )
    hm
    
    pdf(file.path(cor_dir_cur, sprintf("position_correlations_%s_%s_%s_mean-sub.pdf", feature_to_use, coef_to_use, method)), height = height +2, width = width+2)
    print(hm)
    dev.off()
    #})
    
    res_cor_df = melt(res_cor_scale)
    walk(positions, function(p) {
      res_cor_df_cur = res_cor_df %>% filter(Var1 == p)
      
      gg = ggplot(res_cor_df_cur, aes(Var2, value, fill=Var2)) +
        geom_bar(stat = "identity") +
        scale_fill_manual(values=position_colors) +
        theme_classic()
      print(gg)
    })
  })
})
```


##### Intersection

```{r}
intersect_dir = file.path(outdir2, "intersections")
dir.create(intersect_dir)
```

###### Hypergeometric p-values of gene list intersections
```{r}
np = length(positions)
Ngenes = length(unique(res$gene_id))
interp_mat = matrix(0, ncol=np, nrow=np)

n_genes = 250
coef_to_use = "kl_mean_3"
res_sig = res %>%
  filter(coef == coef_to_use) %>%
  group_by(position) %>% 
  top_n(-1 * n_genes, P.Value) %>%
  mutate(position = droplevels(position))
res_sig_split = split(res_sig, res_sig$position)
res_sig_split = map(res_sig_split, function(x) x$gene_id)

dimnames(interp_mat) = list(positions, positions)

for (i in positions) {
  for (j in positions) {
    ovlp = length(intersect(res_sig_split[[i]], res_sig_split[[j]]))
    length1 = length(res_sig_split[[i]])
    length2 = length(res_sig_split[[j]])
    if (length1 == 0 | length2 == 0) {
      interp_mat[i,j] = 0
    } else {
      interp_mat[i,j] = -1 * log10(phyper(q=ovlp, m=length1, n = Ngenes - length1, k = length2, log.p = F, lower.tail=F))
      #interp_mat[i,j] = -1 * log10(1- phyper(q=ovlp, m=length1, n = Ngenes - length1, k = length2))
    }
  }
}

# interp_mat_m = melt(interp_mat)
# interp_mat_m = interp_mat_m %>% mutate(Var1 = factor(Var1, levels=position_levels),
#                                      Var2 = factor(Var2, levels=rev(position_levels)))
# interp_mat_m$value[is.infinite(interp_mat_m$value)] = 0
# ncomparisons = (np^2 - np) / 2
# thresh = -1 * log10(.01 / ncomparisons)
# 
# gg = ggplot(interp_mat_m, aes(Var1, Var2, fill=value)) +
#   geom_raster() +
#   scale_fill_viridis("-1 * log10(p-value)", limits=c(0,max(interp_mat_m$value)), na.value="white")
# # gg = gg + theme(axis.text.x=element_text(angle=90, hjust=1, color=rep(position_colors1, times=3)),
# #                 axis.text.y=element_text(color=rep(rev(position_colors1), times=3)),
# #                 legend.position="bottom", legend.text=element_text(angle=0))
# #gg = gg + scale_fill_viridis("-1 * log10(p-value)", limits=c(thresh,max(interp_mat_m$value)), na.value="white")
# #gg = gg + labs(x="", y="")
# gg
# save_plot(file.path(intersect_dir, sprintf("phyper_intersect_heatmap_%s.pdf", coef_to_use)), gg, base_width = 5, base_height=5.5)

positions_ordered = intersect( position_levels, positions)
interp_mat = interp_mat[positions_ordered, positions_ordered]
interp_mat[is.infinite(interp_mat)]  = NA

cols = colorRamp2(breaks=seq(0, max(interp_mat, na.rm=T), length.out=11), colors=viridis_pal()(11))
width = 1
height = width
hm = Heatmap(interp_mat, col=cols,
             cluster_rows = F,
             cluster_columns = F,
             width = unit(width, "in"),
             height = unit(height, "in"))
hm
pdf(file.path(intersect_dir, sprintf("phyper_intersect_heatmap_%s.pdf", coef_to_use)), height = height+2, width = width + 2)
print(hm)
dev.off()

positions_song = c("ra", "hvc", "lman", "x")
lt = which(lower.tri(interp_mat), arr.ind=T)
interp_mat_m = data.frame(Var1 = rownames(interp_mat)[lt[,1]], Var2 = colnames(interp_mat)[lt[,2]], value = interp_mat[lt])
interp_mat_m = interp_mat_m %>% 
  mutate(class = case_when(Var1 %in% positions_song & Var2 %in% positions_song ~ "song-song",
                           !(Var1 %in% positions_song) & !(Var2 %in% positions_song) ~ "nsong-nsong",
                           TRUE ~ "song-nsong")) %>%
 # mutate(pair = paste(Var1, Var2, sep="-"),
#         pair2 = paste(Var2, Var1, sep="-")) %>%
  filter(Var1 != Var2) %>%
  #distinct(Var1, Var2, round(value), .keep_all=T) %>% 
  mutate(class = factor(class, levels=c("song-song", "song-nsong", "nsong-nsong")))

gg = ggplot(interp_mat_m, aes(class, value)) + 
  geom_boxplot(aes(fill=class)) + 
  scale_fill_npg(guide="none") + 
  theme_bw() +
  theme(axis.text = element_text(size=5)) +
  labs(x="", y="")
gg
save_plot(file.path(intersect_dir, sprintf("phyper_intersect_distribution_%s.pdf", coef_to_use)), gg, base_height = 2, base_asp = .5)
```

Signed
```{r}
np = length(positions)
Ngenes = length(unique(res$gene_id))
#interp_mat = matrix(0, ncol=np, nrow=np)

n_genes = 250
coef_to_use = "kl_mean_3"
res_sig = res %>%
  filter(coef == coef_to_use) %>%
  mutate(sign = sign(logFC)) %>%
  group_by(position, sign) %>% 
  top_n(-1 * n_genes, P.Value) %>%
  mutate(position = droplevels(position)) %>%
  mutate(position_sign = paste(position, sign, sep="_"))

#res_sig_split = split(res_sig, res_sig$position_sign)
#res_sig_split = map(res_sig_split, function(x) x$gene_id)

res_sig_split = map(c(1, -1), function(si) {
  res_sig1 = res_sig %>% filter(sign == si) 
  map(positions, function(p) {
    res_sig2 = res_sig1 %>%
      filter(position == p)
    res_sig2$gene_id
  }) %>% set_names(positions)
}) %>% set_names(c("pos", "neg"))



mats = map(c("pos", "neg"), function(si) {
  res_sig_split_cur = res_sig_split[[si]]
  interp_mat = matrix(0, ncol=np, nrow=np)
  dimnames(interp_mat) = list(positions, positions)
  for (i in positions) {
    for (j in positions) {
      ovlp = length(intersect(res_sig_split_cur[[i]], res_sig_split_cur[[j]]))
      length1 = length(res_sig_split_cur[[i]])
      length2 = length(res_sig_split_cur[[j]])
      if (length1 == 0 | length2 == 0) {
        interp_mat[i,j] = 0
      } else {
        interp_mat[i,j] = -1 * log10(phyper(q=ovlp, m=length1, n = Ngenes - length1, k = length2, log.p = F, lower.tail=F))
        #interp_mat[i,j] = -1 * log10(1- phyper(q=ovlp, m=length1, n = Ngenes - length1, k = length2))
      }
    }
  }
  interp_mat[is.infinite(interp_mat)]  = NA
  interp_mat
}) %>% set_names(names(res_sig_split))
# interp_mat_m = melt(interp_mat)
# interp_mat_m = interp_mat_m %>% mutate(Var1 = factor(Var1, levels=position_levels),
#                                      Var2 = factor(Var2, levels=rev(position_levels)))
# interp_mat_m$value[is.infinite(interp_mat_m$value)] = 0
# ncomparisons = (np^2 - np) / 2
# thresh = -1 * log10(.01 / ncomparisons)
# 
# gg = ggplot(interp_mat_m, aes(Var1, Var2, fill=value)) +
#   geom_raster() +
#   scale_fill_viridis("-1 * log10(p-value)", limits=c(0,max(interp_mat_m$value)), na.value="white")
# # gg = gg + theme(axis.text.x=element_text(angle=90, hjust=1, color=rep(position_colors1, times=3)),
# #                 axis.text.y=element_text(color=rep(rev(position_colors1), times=3)),
# #                 legend.position="bottom", legend.text=element_text(angle=0))
# #gg = gg + scale_fill_viridis("-1 * log10(p-value)", limits=c(thresh,max(interp_mat_m$value)), na.value="white")
# #gg = gg + labs(x="", y="")
# gg
# save_plot(file.path(intersect_dir, sprintf("phyper_intersect_heatmap_%s.pdf", coef_to_use)), gg, base_width = 5, base_height=5.5)

positions_ordered = intersect( position_levels, positions)

interp_mat_max = max(map_dbl(mats, max, na.rm=T), na.rm=T)
map(names(res_sig_split), function(si) {
  interp_mat = mats[[si]]
  interp_mat = interp_mat[positions_ordered, positions_ordered]

  
  cols = colorRamp2(breaks=seq(0, interp_mat_max, length.out=11), colors=viridis_pal()(11))
  width = 1
  height = width
  hm = Heatmap(interp_mat, col=cols,
               cluster_rows = F,
               cluster_columns = F,
               width = unit(width, "in"),
               height = unit(height, "in"))
  hm
  pdf(file.path(intersect_dir, sprintf("phyper_intersect_heatmap_%s_%s.pdf", coef_to_use, si)), height = height+2, width = width + 2)
  print(hm)
  dev.off()
})

positions_song = c("ra", "hvc", "lman", "x")
interp_mat_m = map(mats, function(interp_mat) {
lt = which(lower.tri(interp_mat), arr.ind=T)
interp_mat_m = data.frame(Var1 = rownames(interp_mat)[lt[,1]], Var2 = colnames(interp_mat)[lt[,2]], value = interp_mat[lt])
interp_mat_m = interp_mat_m %>% 
  mutate(class = case_when(Var1 %in% positions_song & Var2 %in% positions_song ~ "song-song",
                           !(Var1 %in% positions_song) & !(Var2 %in% positions_song) ~ "nsong-nsong",
                           TRUE ~ "song-nsong")) %>%
 # mutate(pair = paste(Var1, Var2, sep="-"),
#         pair2 = paste(Var2, Var1, sep="-")) %>%
  filter(Var1 != Var2) %>%
  #distinct(Var1, Var2, round(value), .keep_all=T) %>% 
  mutate(class = factor(class, levels=c("song-song", "song-nsong", "nsong-nsong")))

interp_mat_m
}) %>% set_names(names(mats)) %>%  bind_rows(.id="sign")

gg = ggplot(interp_mat_m, aes(class, value)) + 
  geom_boxplot(aes(fill=class), size=.25, outlier.size=.25) + 
  scale_fill_npg(guide="none") + 
  theme_bw() +
  theme(axis.text = element_text(size=5),
        panel.grid=element_blank()) +
  labs(x="", y="") + 
  facet_grid(.~sign)
gg
save_plot(file.path(intersect_dir, sprintf("phyper_intersect_distribution_signed_%s.pdf", coef_to_use)), gg, base_height = 1.8, base_asp = .4, ncol=2)


```




##### Specific genes

```{r}
genes_dir = file.path(outdir2, "specific_genes")
dir.create(genes_dir)
```


###### Estimated coefs

```{r}
se_coefs = readRDS(file.path(prefix_data, "coefs.rds"))
md = se1@meta.data
md_tags = md %>% distinct(tags, procedure2)
```

```{r}
ra_genes = c("CRHBP", "CREM", "SIX2", "SLC8A2", "BDNF", "HOMER1", "ARC", "EGR1", "NPY", "PROM1", "EML6", "SST", "PPP1R1C","FABP7", "AMN1", "CHCHD2", "IQUB", "PCSK2", "SCG2", "ATF3", "PRKCB", "CD24", "STMN1", "LOC110476506", "LPL", "RPL18", "CD99", "ETV1", "ATF4", "IGF2", "CHGB","CORT", "NTS" )
lman_genes = c("SERPINI1", "NRN1", "AR", "CALCB", "PAPPA", "LGALS1", "RIMS4", "HTR1B", "PVALB", "BSN", "LRTM2", "RASL11B", "ATP1A1")
genes = unique(c(ra_genes, lman_genes))
se_coefs_cur = se_coefs[genes,]

coefs = assay(se_coefs_cur, "coef") %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  pivot_longer(-rowname, names_to = "condition", values_to="estimate")

std_errors = assay(se_coefs_cur, "std_errors") %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  pivot_longer(-rowname, names_to = "condition", values_to="std_error")

coefs = coefs %>% left_join(std_errors) %>%
  rename(gene = rowname) %>% 
  mutate(condition_s = map(condition, ~unlist(str_split(.x, "\\."))),
         position = sub("position", "", map_chr(condition_s, 1)),
         tags = sub("tags", "", map_chr(condition_s, 2))) %>%
  left_join(md_tags)

coefs_stat = coefs %>%
    filter(procedure2=="intact") %>% 
  group_by(gene, position) %>%
  summarize(estimate_mean = mean(estimate),
            estimate_sd = sd(estimate)) 

coefs = coefs %>% left_join(coefs_stat) %>%
  mutate(estimate_z = (estimate - estimate_mean) / estimate_sd,
         std_error_z = (std_error / estimate_sd))
md_tags = se1@meta.data %>% distinct(tags, .keep_all=T) %>%
  select(-position, -condition)

coefs = coefs %>% left_join(md_tags) %>%
  mutate(duration_sum_end_log = log1p(duration_sum_end))
```

```{r}
ncols = 2
positions_to_use = c("ra", "hvc", "lman")

coefs_lims = coefs %>% 
  filter(position %in% positions_to_use) %>% 
  group_by(gene) %>%
  summarize(ymin = min(estimate_z, na.rm=T),
            ymax = max(estimate_z, na.rm=T))

x_coefs = c("num_songs_on_euth_date_log", "duration_sum_end_log", "duration_sum_end", "kl_mean_log", "kl_mean")

group_vars = c("procedure2")
walk(positions_to_use, function(p) {
  coefs_cur = coefs %>% filter(position==p)
  
  walk(x_coefs, function(x_coef) {
    walk(group_vars, function(group_var) {
      gg = ggplot(coefs_cur, aes_string(x_coef, "estimate_z", color=group_var)) + 
        geom_hline(yintercept=0, linetype=2) +
        geom_pointrange(aes(ymin=(estimate_z - std_error_z), ymax=(estimate_z + std_error_z)), size=.2) + 
        stat_smooth(se = F, method="loess", span=1, width=1) + 
        
        facet_wrap(~gene, ncol=ncols) + 
        scale_color_startrek() + 
        theme_cowplot()
      print(gg)
      save_plot(file.path(genes_dir, sprintf("gene_by_%s_split-by-%s_%s_z.pdf", x_coef, group_var, p)), gg, base_height = 2, ncol=ncols, nrow = ceiling(length(genes) / ncols))
    })
  })
})

walk(positions_to_use, function(p) {
  coefs_cur = coefs %>% filter(position==p)
  walk(x_coefs, function(x_coef) {
    walk(group_vars, function(group_var) {
      gg = ggplot(coefs_cur, aes_string(x_coef, "estimate", color=group_var)) + 
        geom_pointrange(aes(ymin=(estimate - std_error), ymax=(estimate + std_error)), size=.2) + 
        stat_smooth(se = F, method="loess", span=2, width=1) + 
        facet_wrap(~gene, ncol=ncols, scales="free") +
        scale_y_continuous(expand = expand_scale(add=1)) + 
        scale_color_startrek() + 
        theme_minimal()
      print(gg)
      save_plot(file.path(genes_dir, sprintf("gene_by_%s_split-by-%s_%s.pdf", x_coef, group_var, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(genes) / ncols))
    })
  })
})
```

```{r}
ncols = 2
positions_to_use = c("ra", "hvc", "lman")
x_coefs = c("kl_mean_log_scale_cut2_proc2")
#cols = c("grey90", "grey50", "grey10")
x_coef_levels =  levels(se1$kl_mean_log_scale_cut2_proc2)
cols = brewer_pal(palette="Reds")(9)[seq(3, 9, length.out=length(x_coef_levels))]
names(cols) = x_coef_levels

coefs_sum = coefs %>%
  group_by(position, gene, kl_mean_log_scale_cut2_proc2) %>%
  summarize(estimate_mean = mean(estimate, na.rm=T))

walk(positions_to_use, function(p) {
  coefs_cur = coefs %>% filter(position==p)
  coefs_sum_cur = coefs_sum %>% filter(position==p)
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur, aes_string(x_coef, "estimate")) + 
      geom_point(position=position_jitter(width=.2), color="grey50") + 
      geom_point(data=coefs_sum_cur, aes_string(x_coef, y="estimate_mean", color=x_coef), size=2) + 
      facet_wrap(~gene, ncol=ncols, scales="free") +
      scale_y_continuous(expand = expand_scale(add=1)) + 
      scale_color_manual(name="", values=cols) + 
      theme_bw()
    #print(gg)
    save_plot(file.path(genes_dir, sprintf("gene_by_%s_%s.pdf", x_coef, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(genes) / ncols))
  })
})

colorby="procedure2"
walk(positions_to_use, function(p) {
  coefs_cur = coefs %>% filter(position==p)  #%>% filter(gene=="CRHBP")
  coefs_sum_cur = coefs_sum %>% filter(position==p) #%>% filter(gene=="CRHBP")
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur, aes_string(x_coef, "estimate")) + 
      geom_point(position=position_jitter(width=.2), aes_string(color=colorby)) + 
      geom_point(data=coefs_sum_cur, aes_string(x_coef, y="estimate_mean"), color="black", size=2) + 
      geom_line(data=coefs_sum_cur, aes_string(x_coef, y="estimate_mean", group="gene"), color="black") + 

      facet_wrap(~gene, ncol=ncols, scales="free") +
      scale_y_continuous(expand = expand_scale(add=1)) + 
      scale_color_manual(name="", values=deaf_colors) + 
      theme_bw()
    #print(gg)
    save_plot(file.path(genes_dir, sprintf("gene_by_%s_colorby-%s_%s.pdf", x_coef, colorby, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(genes) / ncols))
  })
})

colorby="procedure2"
walk(positions_to_use, function(p) {
  coefs_cur = coefs %>% filter(position==p)
  coefs_sum_cur = coefs_sum %>% filter(position==p)
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur, aes_string(x_coef, "estimate")) + 
      #geom_pointrange(aes(ymin=(estimate - std_error), ymax=(estimate + std_error)), size=.2) + 
      geom_bar(data=coefs_sum_cur, aes_string(x_coef, y="estimate_mean"), fill="grey", color=NA, stat="identity") + 
      geom_point(position=position_jitter(width=.2), aes_string(color=colorby)) + 

      #stat_summary(fun.y = "mean", color="blue") + 
     # stat_smooth(se = F, method="loess", span=2, width=1) + 
      facet_wrap(~gene, ncol=ncols, scales="free") +
      scale_y_continuous(expand = expand_scale(add=1)) + 
      scale_color_manual(name="", values=deaf_colors) + 
      theme_bw()
    #print(gg)
    save_plot(file.path(genes_dir, sprintf("bar_gene_by_%s_colorby-%s_%s.pdf", x_coef, colorby, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(genes) / ncols))
  })
})


```

```{r}
ncols = 2
positions_to_use = c("ra", "hvc", "lman")
x_coefs = c("kl_mean_log_scale_cut2_proc")
#cols = c("grey90", "grey50", "grey10")
cols = brewer_pal(palette="Reds")(9)[c(3,6,9)]
names(cols) = c(1,2,3)

coefs_cur = coefs %>% 
  filter(gene %in% genes)
coefs_sum = coefs_cur %>%
  
  group_by(position, gene, kl_mean_log_scale_cut2_proc) %>%
  summarize(estimate_mean = mean(estimate, na.rm=T))

walk(positions_to_use, function(p) {
  coefs_cur1 = coefs_cur %>% filter(position==p)
  coefs_sum_cur = coefs_sum %>% filter(position==p)
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur1, aes_string(x_coef, "estimate")) + 
      #geom_pointrange(aes(ymin=(estimate - std_error), ymax=(estimate + std_error)), size=.2) + 
      geom_point(position=position_jitter(width=.2), color="grey50") + 
      geom_point(data=coefs_sum_cur, aes_string(x_coef, y="estimate_mean", color=x_coef), size=2) + 
      #stat_summary(fun.y = "mean", color="blue") + 
     # stat_smooth(se = F, method="loess", span=2, width=1) + 
      facet_wrap(~gene, ncol=ncols, scales="free") +
      scale_y_continuous(expand = expand_scale(add=1)) + 
      scale_color_manual(name="", values=cols) + 
      theme_bw()
    #print(gg)
    save_plot(file.path(genes_dir, sprintf("lman-genes_by_%s_%s.pdf", x_coef, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(lman_genes) / ncols))
  })
})

colorby="procedure2"
walk(positions_to_use, function(p) {
  coefs_cur1 = coefs_cur %>% filter(position==p)
  coefs_sum_cur = coefs_sum %>% filter(position==p)
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur1, aes_string(x_coef, "estimate")) + 
      #geom_pointrange(aes(ymin=(estimate - std_error), ymax=(estimate + std_error)), size=.2) + 
      geom_point(position=position_jitter(width=.2), aes_string(color=colorby)) + 
      geom_point(data=coefs_sum_cur, aes_string(x_coef, y="estimate_mean"), color="black", size=2) + 
      geom_line(data=coefs_sum_cur, aes_string(x_coef, y="estimate_mean", group="gene"), color="black") + 
      #stat_summary(fun.y = "mean", color="blue") + 
     # stat_smooth(se = F, method="loess", span=2, width=1) + 
      facet_wrap(~gene, ncol=ncols, scales="free") +
      scale_y_continuous(expand = expand_scale(add=1)) + 
      scale_color_manual(name="", values=deaf_colors) + 
      theme_bw()
    #print(gg)
    save_plot(file.path(genes_dir, sprintf("lman-genes_by_%s_colorby-%s_%s.pdf", x_coef, colorby, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(lman_genes) / ncols))
  })
})

``` 

###### DEG across kl_mean_3
```{r}
coefs_to_use = list(c("kl_mean_2", "kl_mean_3"))

nrows = 5
walk(positions, function(p) {
  walk(coefs_to_use, function(coef_to_use) {
    res_cur = res %>% 
      #filter(pathway==pathway_to_use) %>%
      filter(coef %in% coef_to_use) %>%
      filter(position==p)  %>%
      filter(adj.P.Val<.1) %>%
      group_by(gene_id) %>%
      summarize(logFC = mean(logFC)) %>% 
      mutate(logFC_sign = sign(logFC))
    
    genes_split = split(res_cur$gene_id, res_cur$logFC_sign)
    
    #walk(names(genes_split), function(logFC_sign) {
      genes = unique(res_cur$gene_id)
      
      if(length(genes) <= 2) {
        return()
      }
      
      line_color = position_colors[p]
      df = FetchData(se1, c(genes), slot="data") %>%
        as.data.frame() %>%
        rownames_to_column(var="id") %>%
        pivot_longer(cols=-id, names_to="gene_id", values_to="value") %>%
        mutate(id = as.numeric(id)) %>% 
        left_join(info2 %>% select(id, position, procedure2)) %>%
        mutate(position = factor(position, levels=position_levels),
               gene_id = factor(gene_id, levels=genes)) #%>%
      
      df_id = df %>% 
        distinct(id, position) %>%
        arrange(position) 
      id_levels = df_id$id
      
      df = df %>%
      filter(position == p) %>% 
      left_join(md %>% select(id, position, kl_mean_log_scale_cut2_proc))
    
    df = df %>%
      group_by(gene_id, logFC_sign) %>% 
      mutate(
        baseline = kl_mean_log_scale_cut2_proc==1,
        value_rel = value - mean(value[baseline]),
        value_z = (value - mean(value[baseline])) / sd(value[baseline]))
    
    df_sum = df %>% 
      group_by(gene_id, kl_mean_log_scale_cut2_proc) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean))
    
    df_group_sum = df  %>%
      group_by(kl_mean_log_scale_cut2_proc, logFC_sign) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean)) %>% 
      ungroup()
    
    gg = ggplot() + 
      #geom_point(size = .5, alpha=.5, shape=20, position=position_jitter(width=.1)) + 
      #geom_point(data=df_sum, color=line_color, size=.5, aes(kl_mean_log_scale_cut2_proc, value_mean)) + 
            geom_hline(yintercept=0, linetype=1, size=.25) + 
      geom_line(data=df_sum, color="grey", alpha=.5, aes(kl_mean_log_scale_cut2_proc, value_z_mean,  group=gene_id), size=.25) +
      geom_line(data=df_group_sum, color=line_color, aes(kl_mean_log_scale_cut2_proc, value_z_mean, group=logFC_sign)) +

      #scale_color_manual(values=position_colors) + 
      #facet_wrap(~gene_id, scales="free", nrow=nrows) + 
      labs(x = "", y = "expression\nz-scored rel. low Song DKL") + 
      theme_bw() + 
      theme(
        legend.position="none",
        strip.text = element_text(size=6, face="italic", hjust=0, vjust=1),
        strip.background = element_blank(),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.text.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid=element_blank()
      )
    print(gg)
    width = .5
    height = .6
    gg = set_panel_size(gg, width = unit(width, "in"), height = unit(height, "in"))
    save_plot(file.path(genes_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc_%s_%s_%s_grouped.pdf", p, paste(coef_to_use, collapse="-"), logFC_sign)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
    
    # gg = gg + geom_text_repel(data=df_sum %>% filter(kl_mean_log_scale_cut2_proc==3), aes(x = 3, value_z_mean, label=gene_id), size=4)
    # 
    #     save_plot(file.path(gsea_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc_%s_%s_grouped_with-labels.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
 # })
})
})
```


###### Residuals

```{r}
genes_dir_resid = file.path(genes_dir, "residuals")
dir.create(genes_dir_resid)
```

Model reduced

```{r}
residuals_fname = file.path(outdir1, "seurat_drop-kl_mean_residuals.rds")
se1_resid = readRDS(residuals_fname)
resids =  GetAssayData(se1_resid, "counts", assay="residuals")
```

```{r}
# dat_cur = GetAssayData(se1_cur, "counts", assay="RNA")
#   dge = DGEList(dat_cur)
#   dge = calcNormFactors(dge,method =c("TMMwsp"))
# resids = residuals(fit, dge)
```

```{r}

```

```{r}
genes = c("CRHBP", "CREM", "SIX2", "SLC8A2", "BDNF", "HOMER1", "ARC", "EGR1", "PROCA1", "NPY", "PROM1", "EML6", "NTN4", "SST" ,"FBXW4", "PPP1R1C", "LRRC4B", "SYT16", "SOD1", "FABP7", "AMN1", "CHCHD2", "IQUB", "PCSK2", "SCG2", "ATF3", "PRKCB", "CD24")
resids_cur = resids[genes,]

resids_df = resids_cur %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  pivot_longer(-rowname, names_to = "id", values_to="estimate") %>%
  mutate(estimate = log1p(estimate))

# std_errors = assay(se_resids_df_cur, "std_errors") %>%
#   as.data.frame() %>%
#   rownames_to_column() %>%
#   pivot_longer(-rowname, names_to = "condition", values_to="std_error")

md = FetchData(se1_cur, c("position", "procedure2", "tags", "kl_mean_log_scale", "kl_mean_log_scale_cut2_proc", "num_songs_on_euth_date_log_scale")) %>%
  rownames_to_column(var="id")
resids_df = resids_df %>% 
  left_join(md) %>%
  #left_join(std_errors) %>%
  rename(gene = rowname) #%>% 
 # mutate(condition_s = map(condition, ~unlist(str_split(.x, "\\."))),
#         position = sub("position", "", map_chr(condition_s, 1)),
#         tags = sub("tags", "", map_chr(condition_s, 2))) %>%
 # left_join(md_tags)

# resids_df_stat = resids_df %>%
#     filter(procedure2=="intact") %>% 
#   group_by(gene, position) %>%
#   summarize(estimate_mean = mean(estimate),
#             estimate_sd = sd(estimate)) 
# 
# resids_df = resids_df %>% left_join(resids_df_stat) %>%
#   mutate(estimate_z = (estimate - estimate_mean) / estimate_sd,
#          std_error_z = (std_error / estimate_sd))
# md_tags = se1@meta.data %>% distinct(tags, .keep_all=T) %>%
#   select(-position, -condition)
# 
# resids_df = resids_df %>% left_join(md_tags) %>%
#   mutate(duration_sum_end_log = log1p(duration_sum_end))
```

```{r}
x_coefs = c("kl_mean_log_scale_cut2_proc")
walk(positions_to_use, function(p) {
  resids_df_cur = resids_df %>% filter(position==p)

  walk(x_coefs, function(x_coef) {
    gg = ggplot(resids_df_cur, aes_string(x_coef, "estimate", color="num_songs_on_euth_date_log_scale")) + 
      #geom_pointrange(aes(ymin=(estimate - std_error), ymax=(estimate + std_error)), size=.2) + 
      geom_point(position=position_jitter(width=.2)) + 
      stat_summary() + 
     # stat_smooth(se = F, method="loess", span=2, width=1) + 
      facet_wrap(~gene, ncol=ncols, scales="free") +
      scale_y_continuous(expand = expand_scale(add=1)) + 
     # scale_color_startrek(name="") + 
      scale_color_continuous(name="") + 
      theme_minimal()
    print(gg)
    save_plot(file.path(genes_dir_resid, sprintf("gene_by_%s_%s.pdf", x_coef, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(genes) / ncols))
  })
})
```

###### Batch corrected

```{r}
genes_dir_batch = file.path(genes_dir, "batch_corrected")
dir.create(genes_dir_batch)
```

```{r}
#se_coefs = readRDS(file.path(prefix_data, "coefs.rds"))
se1$position_tags = paste(se1$position, se1$tags, sep=".")

cdr_thresh = .5
plot(density(se1$cdr))
se1_filt = se1[,se1$cdr >= cdr_thresh]
md = se1_filt@meta.data
md_tags = md %>% distinct(tags, procedure2)
se_dat = GetAssayData(se1_filt, assay="combat_id2")
se_dat = se_dat[,match(md$id, colnames(se_dat))]

Idents(se1_filt) = se1_filt$position_tags
se_coefs = AverageExpression(se1_filt, assays="combat_id2")
se_coefs = log1p(se_coefs[[1]])

#se_coefs = t(apply(se_dat, 1, function(x) tapply()))
se_sem = t(apply(se_dat, 1, function(x) tapply(x, md$position_tags, sem)))

```

```{r}
ra_genes = c("CRHBP", "CREM", "SIX2", "SLC8A2", "BDNF", "HOMER1", "ARC", "EGR1", "PROCA1", "NPY", "PROM1", "EML6", "NTN4", "SST" ,"FBXW4", "PPP1R1C", "LRRC4B", "SYT16", "SOD1", "FABP7", "AMN1", "CHCHD2", "IQUB", "PCSK2", "SCG2", "ATF3", "PRKCB", "CD24", "STMN1", "LOC110476506", "LPL", "RPL18", "CD99")
lman_genes = c("SERPINI1", "NRN1", "AR", "CALCB", "PAPPA", "LGALS1", "RIMS4", "HTR1B", "PVALB", "BSN")
genes = unique(c(ra_genes, lman_genes))
se_coefs_cur = se_coefs[genes,]

coefs = se_coefs_cur %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  pivot_longer(-rowname, names_to = "condition", values_to="estimate")

std_errors = se_sem %>% 
  as.data.frame() %>%
  rownames_to_column() %>%
  pivot_longer(-rowname, names_to = "condition", values_to="std_error")

coefs = coefs %>% left_join(std_errors) %>%
  rename(gene = rowname) %>% 
  mutate(condition_s = map(condition, ~unlist(str_split(.x, "\\."))),
         position = sub("position", "", map_chr(condition_s, 1)),
         tags = sub("tags", "", map_chr(condition_s, 2))) %>%
  left_join(md_tags)

coefs_stat = coefs %>%
    filter(procedure2=="intact") %>% 
  group_by(gene, position) %>%
  summarize(estimate_mean = mean(estimate),
            estimate_sd = sd(estimate)) 

coefs = coefs %>% left_join(coefs_stat) %>%
  mutate(estimate_z = (estimate - estimate_mean) / estimate_sd,
         std_error_z = (std_error / estimate_sd))
md_tags = se1@meta.data %>% distinct(tags, .keep_all=T) %>%
  dplyr::select(-position, -condition)

coefs = coefs %>% left_join(md_tags) %>%
  mutate(duration_sum_end_log = log1p(duration_sum_end))
```

```{r}
ncols = 2
positions_to_use = c("ra", "hvc", "lman")

coefs_lims = coefs %>% 
  filter(position %in% positions_to_use) %>% 
  group_by(gene) %>%
  summarize(ymin = min(estimate_z, na.rm=T),
            ymax = max(estimate_z, na.rm=T))
walk(positions_to_use, function(p) {
  coefs_cur = coefs %>% filter(position==p) 
  x_coefs = c("num_songs_on_euth_date_log", "duration_sum_end_log", "duration_sum_end")
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur, aes_string(x_coef, "estimate_z", color="procedure2")) + 
      geom_hline(yintercept=0, linetype=2) +
      geom_pointrange(aes(ymin=(estimate_z - std_error_z), ymax=(estimate_z + std_error_z)), size=.2) + 
      stat_smooth(se = F, method="loess", span=1, width=1) + 

      facet_wrap(~gene, ncol=ncols) + 
      scale_color_startrek() + 
      theme_cowplot()
   # print(gg)
    save_plot(file.path(genes_dir, sprintf("mean_gene_by_%s_%s_z.pdf", x_coef, p)), gg, base_height = 2, ncol=ncols, nrow = ceiling(length(genes) / ncols))
  })
})

walk(positions_to_use, function(p) {
  coefs_cur = coefs %>% filter(position==p)
  x_coefs = c("num_songs_on_euth_date_log", "duration_sum_end_log", "duration_sum_end")
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur, aes_string(x_coef, "estimate", color="procedure2")) + 
      geom_pointrange(aes(ymin=(estimate - std_error), ymax=(estimate + std_error)), size=.2) + 
      stat_smooth(se = F, method="loess", span=2, width=1) + 
      facet_wrap(~gene, ncol=ncols, scales="free") +
      scale_y_continuous(expand = expand_scale(add=1)) + 
      scale_color_startrek() + 
      theme_minimal()
    #print(gg)
    save_plot(file.path(genes_dir_batch, sprintf("mean_gene_by_%s_%s.pdf", x_coef, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(genes) / ncols))
  })
})
```

```{r}
positions_to_use = c("ra", "hvc", "lman")
#walk(positions_to_use, function(p) {
  coefs_cur = coefs %>% filter(position %in% positions_to_use) %>%
    mutate(position = factor(position, levels=positions_to_use))
  x_coefs = c("num_songs_on_euth_date_log", "duration_sum_end_log", "duration_sum_end")
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur, aes_string(x_coef, "estimate_z", color="procedure2")) + 
      geom_hline(yintercept=0, linetype=2) +
      geom_pointrange(aes(ymin=(estimate_z - std_error_z), ymax=(estimate_z + std_error_z)), size=.2) + 
      stat_smooth(se = F, method="loess", span=1, size=1) + 

      facet_grid(gene~position, scales="free") + 
      scale_color_startrek() + 
      theme_cowplot()
    #print(gg)
    save_plot(file.path(genes_dir_batch, sprintf("gene_by_%s_facet_z.pdf", x_coef)), gg, base_height = 1, base_asp=2, ncol=length(positions_to_use), nrow = length(genes))
  })

```

```{r}
ncols = 2
positions_to_use = c("ra", "hvc", "lman")
x_coefs = c("kl_mean_log_scale_cut2_proc")
#cols = c("grey90", "grey50", "grey10")
cols = brewer_pal(palette="Reds")(9)[c(3,6,9)]
names(cols) = c(1,2,3)

coefs_sum = coefs %>%
  group_by(position, gene, kl_mean_log_scale_cut2_proc) %>%
  summarize(estimate_mean = mean(estimate, na.rm=T))

walk(positions_to_use, function(p) {
  coefs_cur = coefs %>% filter(position==p)
  coefs_sum_cur = coefs_sum %>% filter(position==p)
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur, aes_string(x_coef, "estimate")) + 
      #geom_pointrange(aes(ymin=(estimate - std_error), ymax=(estimate + std_error)), size=.2) + 
      geom_point(position=position_jitter(width=.2), color="grey50") + 
      geom_point(data=coefs_sum_cur, aes_string(x_coef, y="estimate_mean", color=x_coef), size=2) + 
      #stat_summary(fun.y = "mean", color="blue") + 
     # stat_smooth(se = F, method="loess", span=2, width=1) + 
      facet_wrap(~gene, ncol=ncols, scales="free") +
      scale_y_continuous(expand = expand_scale(add=1)) + 
      scale_color_manual(name="", values=cols) + 
      theme_bw()
    #print(gg)
    save_plot(file.path(genes_dir_batch, sprintf("ra-genes_by_%s_%s.pdf", x_coef, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(genes) / ncols))
  })
})

colorby="procedure2"
walk(positions_to_use, function(p) {
  coefs_cur = coefs %>% filter(position==p)
  coefs_sum_cur = coefs_sum %>% filter(position==p)
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur, aes_string(x_coef, "estimate")) + 
      #geom_pointrange(aes(ymin=(estimate - std_error), ymax=(estimate + std_error)), size=.2) + 
      geom_point(position=position_jitter(width=.2), aes_string(color=colorby)) + 
      geom_point(data=coefs_sum_cur, aes_string(x_coef, y="estimate_mean"), color="black", size=2) + 
      #stat_summary(fun.y = "mean", color="blue") + 
     # stat_smooth(se = F, method="loess", span=2, width=1) + 
      facet_wrap(~gene, ncol=ncols, scales="free") +
      scale_y_continuous(expand = expand_scale(add=1)) + 
      scale_color_manual(name="", values=deaf_colors) + 
      theme_bw()
    #print(gg)
    save_plot(file.path(genes_dir_batch, sprintf("ra-genes_by_%s_colorby-%s_%s.pdf", x_coef, colorby, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(genes) / ncols))
  })
})

```

```{r}
ncols = 2
positions_to_use = c("ra", "hvc", "lman")
x_coefs = c("kl_mean_log_scale_cut2_proc")
#cols = c("grey90", "grey50", "grey10")
cols = brewer_pal(palette="Reds")(9)[c(3,6,9)]
names(cols) = c(1,2,3)

coefs_cur = coefs %>% 
  filter(gene %in% lman_genes)
coefs_sum = coefs_cur %>%
  
  group_by(position, gene, kl_mean_log_scale_cut2_proc) %>%
  summarize(estimate_mean = mean(estimate, na.rm=T))

walk(positions_to_use, function(p) {
  coefs_cur1 = coefs_cur %>% filter(position==p)
  coefs_sum_cur = coefs_sum %>% filter(position==p)
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur1, aes_string(x_coef, "estimate")) + 
      #geom_pointrange(aes(ymin=(estimate - std_error), ymax=(estimate + std_error)), size=.2) + 
      geom_point(position=position_jitter(width=.2), color="grey50") + 
      geom_point(data=coefs_sum_cur, aes_string(x_coef, y="estimate_mean", color=x_coef), size=2) + 
      #stat_summary(fun.y = "mean", color="blue") + 
     # stat_smooth(se = F, method="loess", span=2, width=1) + 
      facet_wrap(~gene, ncol=ncols, scales="free") +
      scale_y_continuous(expand = expand_scale(add=1)) + 
      scale_color_manual(name="", values=cols) + 
      theme_bw()
    #print(gg)
    save_plot(file.path(genes_dir_batch, sprintf("lman-genes_by_%s_%s.pdf", x_coef, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(lman_genes) / ncols))
  })
})

colorby="procedure2"
walk(positions_to_use, function(p) {
  coefs_cur1 = coefs_cur %>% filter(position==p)
  coefs_sum_cur = coefs_sum %>% filter(position==p)
  walk(x_coefs, function(x_coef) {
    gg = ggplot(coefs_cur1, aes_string(x_coef, "estimate")) + 
      #geom_pointrange(aes(ymin=(estimate - std_error), ymax=(estimate + std_error)), size=.2) + 
      geom_point(position=position_jitter(width=.2), aes_string(color=colorby)) + 
      geom_point(data=coefs_sum_cur, aes_string(x_coef, y="estimate_mean"), color="black", size=2) + 
      #stat_summary(fun.y = "mean", color="blue") + 
     # stat_smooth(se = F, method="loess", span=2, width=1) + 
      facet_wrap(~gene, ncol=ncols, scales="free") +
      scale_y_continuous(expand = expand_scale(add=1)) + 
      scale_color_manual(name="", values=deaf_colors) + 
      theme_bw()
    #print(gg)
    save_plot(file.path(genes_dir_batch, sprintf("lman-genes_by_%s_colorby-%s_%s.pdf", x_coef, colorby, p)), gg, base_height = 1.5, ncol=ncols, nrow = ceiling(length(lman_genes) / ncols))
  })
})

```

###### Across ARGs

```{r}
library(fgsea)
gray_gmt_fname = "~/data/gene_lists/gray_2018_arg.xls.gmt"
gray_gmt = gmtPathways(gray_gmt_fname)
gray_gmt_df = data.frame(gene = unlist(gray_gmt), class = rep(names(gray_gmt), times = map_int(gray_gmt, length)))
```

Mean logFC
```{r}
gray_gmt_df_cur = gray_gmt_df %>% filter(class=="rapid PRGs")
res_cur = res %>% filter(gene_id %in% gray_gmt_df_cur$gene)
res_cur_mean = res_cur %>%
  group_by(position, coef) %>%
  summarize(logFC_mean = mean(logFC),
            logFC_sd = sd(logFC))

gg = ggplot(res_cur %>% filter(coef=="num_songs_euth_date"), aes(gene_id, logFC)) + 
  geom_point() + 
  geom_hline(yintercept=0) + 
  facet_wrap(~position) + 
  coord_flip()
gg
```
Whiteny singing genes
```{r}
## Singing
sing_fname = "~/data/gene_lists/whitney_science_2014/Supplementary Table S8 - Singing Regulated Transcripts v53.xlsx"
sing = readxl::read_xlsx(sing_fname, skip=2)
colnames(sing) = make.names(colnames(sing))
sing_hvc = sing %>% filter(HVC.p<.1) %>% 
  filter(grepl("Up|Increase", HVC.Cluster)) %>% 
  filter(!is.na(Symbol)) 

sing_ra = sing %>% filter(RA.p<.1) %>% 
  filter(grepl("Up|Increase", RA.Cluster)) %>% 
  filter(!is.na(Symbol)) 


res_cur = res %>% filter(gene_id %in% sing_ra$Symbol)
res_cur_mean = res_cur %>%
  group_by(position, coef) %>%
  summarize(logFC_mean = mean(logFC),
            logFC_sd = sd(logFC))

gg = ggplot(res_cur %>% filter(coef=="num_songs_euth_date"), aes(gene_id, logFC)) + 
  geom_point() + 
  geom_hline(yintercept=0) + 
  facet_wrap(~position) + 
  coord_flip()
gg
```

```{r}

se_coefs_cur = se_coefs[genes,]

coefs = assay(se_coefs_cur, "coef") %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  pivot_longer(-rowname, names_to = "condition", values_to="estimate")

std_errors = assay(se_coefs_cur, "std_errors") %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  pivot_longer(-rowname, names_to = "condition", values_to="std_error")

coefs = coefs %>% left_join(std_errors) %>%
  rename(gene = rowname) %>% 
  mutate(condition_s = map(condition, ~unlist(str_split(.x, "\\."))),
         position = sub("position", "", map_chr(condition_s, 1)),
         tags = sub("tags", "", map_chr(condition_s, 2))) %>%
  left_join(md_tags)

coefs_stat = coefs %>%
    filter(procedure2=="intact") %>% 
  group_by(gene, position) %>%
  summarize(estimate_mean = mean(estimate),
            estimate_sd = sd(estimate)) 

coefs = coefs %>% left_join(coefs_stat) %>%
  mutate(estimate_z = (estimate - estimate_mean) / estimate_sd,
         std_error_z = (std_error / estimate_sd))
md_tags = se1@meta.data %>% distinct(tags, .keep_all=T) %>%
  select(-position, -condition)
coefs = coefs %>% left_join(md_tags) %>%
  mutate(duration_sum_end_log = log1p(duration_sum_end))

coefs = coefs %>% left_join(gray_gmt_df)
coefs_mean = coefs %>%
  group_by(position, class, tags, procedure2, num_songs_on_euth_date_log) %>%
  summarize(estimate_z = mean(estimate_z)) 

```

```{r}
gg = ggplot(coefs_mean %>% filter(position %in% c("ra", "hvc", "lman")), aes(num_songs_on_euth_date_log, estimate_z, color=procedure2)) + 
  geom_point() + 
  stat_smooth()  + 
  facet_grid(class~position, scales="free_y")
gg
```

##### By singing rate

```{r}
sing_dir = file.path(outdir2, "by_singing")
dir.create(sing_dir)
```

```{r}
co ="position%s.kl_mean_log_scale.num_songs_on_euth_date_log_scale"
p = "ra"
res_to_use = res %>% filter(coef==co) %>%
  filter(position == p) %>%
  filter(adj.P.Val < .05)
genes = res_to_use %>% select(gene_id) %>% unlist()

se_coefs_rnames = data.frame(gene_id = rownames(se_coefs)) %>%
  left_join(loc_trans) %>%
  mutate(gene_id = if_else(!is.na(synonym), synonym, gene_id))
rownames(se_coefs) = se_coefs_rnames$gene_id
se_coefs_cur = se_coefs[genes,]

coefs = assay(se_coefs_cur, "coef") %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  pivot_longer(-rowname, names_to = "condition", values_to="estimate")

std_errors = assay(se_coefs_cur, "std_errors") %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  pivot_longer(-rowname, names_to = "condition", values_to="std_error")

coefs = coefs %>% left_join(std_errors) %>%
  rename(gene = rowname) %>% 
  mutate(condition_s = map(condition, ~unlist(str_split(.x, "\\."))),
         position = sub("position", "", map_chr(condition_s, 1)),
         tags = sub("tags", "", map_chr(condition_s, 2)))

coefs_stat = coefs %>%
  group_by(gene, position) %>%
  summarize(estimate_mean = mean(estimate),
            estimate_sd = sd(estimate))
coefs = coefs %>% left_join(coefs_stat) %>%
  mutate(estimate_z = (estimate - estimate_mean) / estimate_sd,
         std_error_z = (std_error / estimate_sd))
md_tags = se1@meta.data %>% distinct(tags, .keep_all=T) %>%
  select(-position, -condition)
coefs = coefs %>% left_join(md_tags) %>%
  mutate(duration_sum_end_log = log1p(duration_sum_end))
```

```{r}
coefs_cur = coefs %>% filter(position == p)

minx = min(coefs_cur$num_songs_on_euth_date_log)
maxx = max(coefs_cur$num_songs_on_euth_date_log)

xout = seq(minx, maxx, length.out=100)
newdata = data.frame(num_songs_on_euth_date_log = xout)

coefs_ext = coefs_cur  %>% group_by(gene, procedure2) %>% do({
  d = .
  newdata$estimate_loess = predict(loess(estimate ~ num_songs_on_euth_date_log, data = d, span=2), newdata = newdata)
  newdata
})  %>% ungroup() %>%
  group_by(gene) %>%
  mutate(estimate_diff = estimate_loess - estimate_loess[procedure2=="intact"])
```

```{r}
mat = coefs_ext %>% 
  filter(procedure2=="deaf") %>% 
  filter(!is.na(estimate_diff)) %>% 
  select(gene, num_songs_on_euth_date_log, estimate_diff) %>%
  pivot_wider(names_from = "num_songs_on_euth_date_log", values_from = "estimate_diff") %>%
  column_to_rownames("gene") %>%
  as.matrix()

stretch_factor = 4

width = 2
height = width * nrow(mat) / ncol(mat) * stretch_factor

row_hc = hclust(proxy::dist(mat, method="euclidean"), method = "average")

max_val = max(abs(min(mat)), abs(max(mat)))
max_val = 2
 length_out = 9
cols_cur = colorRamp2(breaks=seq(-1*max_val, max_val, length.out=length_out), colors = rev(brewer_pal(palette="BrBG")(length_out)))
hm = Heatmap(mat, col = cols_cur, 
             cluster_columns = F,
             show_column_names = F,
             cluster_rows = row_hc,
             height = unit(height, "in"),
             width = unit(width, "in"))
hm

pdf(file.path(sing_dir, sprintf("heatmap_estimate_diff_by_num_songs_on_euth_date_log_%s.pdf", p)), width = width + 4, height = height + 2)
print(hm)
dev.off()
```

```{r}
coefs_cur = coefs %>% filter(position == p) %>%
  filter(!is.na(estimate))

minx = min(coefs_cur$num_songs_on_euth_date_log)
maxx = max(coefs_cur$num_songs_on_euth_date_log)

miny = min(coefs_cur$kl_mean_log)
maxy = max(coefs_cur$kl_mean_log)

n = 20
xout = seq(minx, maxx, length.out=n)
yout = seq(miny, maxy, length.out = n)

newdata = data.frame(num_songs_on_euth_date_log = xout)

coefs_ext = coefs_cur  %>%
  group_by(gene, procedure2) %>% do({
  d = .
  x = d$num_songs_on_euth_date_log
  y = d$kl_mean_log
  z = d$estimate
  #mat = acast(d, kl_mean_log ~ num_songs_on_euth_date_log, value.var="estimate", fill=0)
  #kd = MASS::kde2d(d$num_songs_on_euth_date_log, y = d$kl_mean_log, n=n, lims=c(c(minx, maxx), c(miny, maxy)))
  in2 = akima::interp(x, y, z,  xo = xout, yo = yout )
  dat = melt(in2$z)
  dat = dat %>% mutate(value = if_else(is.na(value), 0, value))
  colnames(dat) = c("num_songs_on_euth_date_log_ind", "kl_mean_log_ind", "estimate")
  dat
  #newdata$estimate_loess = predict(loess(estimate ~ num_songs_on_euth_date_log, data = d, span=4), newdata = newdata)
 # newdata
})  %>% ungroup() %>%
  group_by(gene, num_songs_on_euth_date_log_ind, kl_mean_log_ind) %>%
  mutate(estimate_diff = estimate - estimate[procedure2=="intact"])
```

```{r}
gene_to_use = "CRHBP"
mat = coefs_ext %>%
  ungroup() %>% 
  filter(gene == gene_to_use) %>% 
  filter(procedure2=="intact") %>% 
  #filter(!is.na(estimate_diff)) %>% 
  select(num_songs_on_euth_date_log_ind, kl_mean_log_ind, estimate) %>%
  pivot_wider(names_from = "num_songs_on_euth_date_log_ind", values_from = "estimate") %>%
  column_to_rownames("kl_mean_log_ind") %>%
  as.matrix()
  
hm = Heatmap(mat, cluster_columns = F, cluster_rows = F, show_row_names = F, show_column_names = F)
hm
```

#### GSEA

```{r}
gsea_dir = file.path(outdir2, "gsea")
dir.create(gsea_dir)

library(fgsea)
```

##### Full
```{r}
gsea_sub_dir = file.path(gsea_dir, "full")
gsea_run_dir = file.path(gsea_sub_dir, "gsea_runs")
dir.create(gsea_run_dir, recursive = T)

gmt_dir = "~/data/gene_lists/gmt"
gmt_files = list.files(gmt_dir, full.names = T)

#dest_gmt_files = dest_gmt_files[!grepl("Drug|Disease", dest_gmt_files)]
gmt_files = gmt_files[grep("c5.all.v7", gmt_files)]

gmt_pathway = gmtPathways(gmt_files)
to_exclude = names(gmt_pathway)
#to_exclude = to_exclude[grep("ENDOPLAS|RIBOSOM|TRANSLATION|METABO|CATABOL|RESPIR|HYDROGEN|OXIDAT|OXIDO|MITOCH|PROTON|VIRAL|ELECTRON|ATP|RRNA|BIOSYNTHETIC|INORGANIC|INNER_MEMBRANE", to_exclude)]
#gmt_pathway = gmt_pathway[!(names(gmt_pathway) %in% to_exclude)]
```

```{r}

gray_gmt_fname = "~/data/gene_lists/gray_2018_arg.xls.gmt"
gray_gmt = gmtPathways(gray_gmt_fname)
names(gray_gmt) = make.names(names(gray_gmt))
gmt_gray_pathway = c(gmt_pathway, gray_gmt)
```

```{r}
positions = as.character(unique(res$position))
coefs = unique(res$coef)
plan(multisession, workers = length(positions))
options(future.globals.maxSize = 1024^2 * 5000)
```



```{r message=FALSE, warning=FALSE}

gsea_res_fname = file.path(gsea_sub_dir, "gsea_res.qs")
gsea_res_ind_fname = file.path(gsea_sub_dir, "gsea_res_ind.qs")

redo = T
if (redo) {
  positions = unique(res$position)
  
gsea_res = future_map(positions, function(p) {
  print(p)
  res_cur = res %>% filter(position == p)
  map(coefs, function(co) {
    print(p)
    res_cur1 = res_cur  %>% filter(coef==co)
    rnks = res_cur1$t
    names(rnks) = res_cur1$gene_id
    rnks = sort(rnks)
    gsea_res = fgseaMultilevel(pathways = gmt_gray_pathway,
                     stats = rnks,
                     minSize=20,
                     maxSize=200
                     )
    
    return(gsea_res)
 
  }) %>% set_names(coefs) %>% bind_rows(.id="coef")
})  %>% set_names(positions) %>% bind_rows(.id="position")

gsea_res_ind = future_map(positions, function(p) {
  print(p)
  res_cur = res %>% filter(position == p)
  gsea_res_cur = gsea_res %>% filter(position==p)
  map(coefs, function(co) {
    res_cur1 = res_cur  %>% filter(coef==co)
    gsea_res_cur1 = gsea_res_cur %>% filter(coef==co)
    rnks = res_cur1$t
    names(rnks) = res_cur1$gene_id
    rnks = sort(rnks)
    gsea_res_filt = gsea_res_cur1 %>% filter(padj<.2) %>% arrange(pval)
    collapsedPathways = collapsePathways(gsea_res_filt, pathways = gmt_gray_pathway, stats = rnks, pval.threshold = .05)
    gsea_res_ind = gsea_res_cur1 %>% filter(pathway %in% collapsedPathways$mainPathways) %>%
      arrange(-NES)
    gsea_res_ind
  }) %>% set_names(coefs) %>% bind_rows(.id="coef")
}) %>% bind_rows()

  qsave(gsea_res, gsea_res_fname)
  qsave(gsea_res_ind, gsea_res_ind_fname)
} else {
  gsea_res = qread(gsea_res_fname)
  gsea_res_ind = qread(gsea_res_ind_fname)
}
```

###### Plot individual positions
```{r}
padj_thresh = 1
positions = unique(gsea_res_ind$position)
coefs = unique(gsea_res_ind$coef)
walk(positions, function(p) {
  walk(coefs, function(co) {
    co_label = sub("position%s\\.", "", co)
    
    gsea_res_ind_filt = gsea_res %>% 
      filter(coef==co) %>% 
      filter(position==p) %>% 
      mutate(p.adjust_sign = -1 * log10(padj)) %>%
      mutate(NES_sign = sign(NES)) %>%
      filter(padj<.2) %>% 
      group_by(NES_sign) %>%
      top_n(10, abs(NES)) %>%
      ungroup()
    
    gsea_res_filt = gsea_res_ind %>% filter(padj<padj_thresh) %>% 
      filter(position==p, coef==co) %>%
      mutate(padj_log = -1 * log10(padj),
             p_log = -1 * log10(pval)) %>%
    #  filter(NES>0) %>% 
      mutate(size1 = size) %>%
      mutate(to_label = ifelse(pathway %in% gsea_res_ind_filt$pathway, pathway, NA)) %>%
      mutate(leadingEdge_len = map_int(leadingEdge, length)) %>%
      mutate(leading_frac = leadingEdge_len / size) %>% 
      
      mutate(pathway1 = sub("^GO", "", to_label ),
             pathway1 = gsub("_", " ", pathway1),
             pathway1 = str_to_title(pathway1))
    
    if (nrow(gsea_res_filt) < 10) 
      return()
    gg = ggplot(gsea_res_filt, aes(NES, p_log, size=leading_frac)) + 
      geom_point(position=position_jitter()) + 
      
      theme_classic() 
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_no_text.pdf", padj_thresh, p, co_label)), gg, base_width=6, base_height=6)
    
    gg = gg + theme(legend.position="none")
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_no_text_no_legend.pdf", padj_thresh, p, co_label)), gg, base_width=6, base_height=6)
    
    gg = gg +  
      geom_text_repel(aes(NES, p_log, label = pathway1), inherit.aes = F, max.overlaps=20) 
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_text.pdf", padj_thresh,  p, co_label)), gg, base_width=6, base_height=6)
  })
})
```

###### Heatmap
```{r}
coefs_to_use = c("kl_mean_2_hearing", "kl_mean_2_deaf", "kl_mean_3", "num_songs_euth_date", "num_songs_pre")
walk(coefs_to_use, function(co) {
  print(co)
  co_label = sub("position%s.", "", co)
 gsea_res_ind_filt = gsea_res %>% 
    filter(coef==co) %>% 
  #  distinct(position, .keep_all=T) %>%
    mutate(p.adjust_sign = -1 * log10(padj)) %>%
    mutate(NES_sign = sign(NES)) %>%
    filter(padj<.2) %>% 
    group_by(position, NES_sign) %>%
    top_n(10, abs(NES)) %>%
   ungroup()
  
  # gsea_res_ind_filt = gsea_res %>% 
  #   filter(coef==co) %>% 
  #   filter(pathway %in% gsea_res_ind_select$pathway)
  
  mat = gsea_res_ind_filt %>%
    filter(coef==co) %>%
    select(position, pathway, NES) %>%
    pivot_wider(names_from = "position", values_from = "NES", values_fill=list(NES=0)) %>%
    column_to_rownames("pathway") %>% 
    as.matrix() 
  
  mat = mat[,intersect(position_levels, colnames(mat))]
  
  
  height = 4
  width = height * ncol(mat) / nrow(mat)
  hm = Heatmap(mat, cluster_columns=F, 
               width = unit(width, "in"),
               height = unit(height, "in"))
  print(hm)
  
  pdf(file.path(gsea_sub_dir, sprintf("heatmap_%s.pdf", co_label)), width=width + 8, height=height + 2)
  print(hm)
  dev.off()
})
```

```{r}
coefs = c("kl_mean_2_hearing", "kl_mean_2_deaf", "kl_mean_3", "num_songs_euth_date", "num_songs_pre")
walk(coefs, function(co) {
  print(co)
co_label = sub("position%s.", "", co)
gsea_res_ind_filt = gsea_res_ind %>% 
  filter(padj<.1) %>%
  filter(coef==co) #%>%
 # group_by(position)# %>%
 # top_n(10, abs(NES))
  # gsea_res_ind_top = gsea_res_ind %>% filter(padj<.2) %>%
  #   top_n(10, NES)
##gsea_res_filt = gsea_res %>% filter(padj<.2)
gsea_res_filt = gsea_res %>% 
  #filter(padj<.2) %>%
  mutate(padj_signed = sign(NES) *  -1 * log10(padj)) %>% 
  filter(pathway %in% gsea_res_ind_filt$pathway)# %>%
 # filter(position %in% c("ra", "hvc", "lman", "x"))

mat = gsea_res_filt %>% filter(coef==co) %>%
  select(position, pathway, padj_signed) %>%
  pivot_wider(names_from = "position", values_from = "padj_signed", values_fill=list(NES=0)) %>%
  column_to_rownames("pathway") %>% 
  as.matrix() 

mat = mat[,intersect(position_levels, colnames(mat))]

max_val = 1.5
 cols_cur = colorRamp2(breaks=c(-1*max_val, 0, max_val), colors = c(muted("purple"), "white", muted("orange")))
 
width = .5
#height = width * nrow(mat) / ncol(mat)
height = 2.5
hm = Heatmap(mat, 
             cluster_columns=F, col = cols_cur,
             clustering_method_rows = "ward.D",
             width = unit(width, "in"),
             height = unit(height, "in"))
hm

pdf(file.path(gsea_sub_dir, sprintf("heatmap_pval_%s.pdf", co_label)), width=width + 8, height=height + 2)
print(hm)
dev.off()

width = 1
height = width * nrow(mat) / ncol(mat)
#height = 4
hm = Heatmap(mat, 
             cluster_columns=F, col = cols_cur,
             clustering_method_rows = "ward.D",
             width = unit(width, "in"),
             height = unit(height, "in"))
hm

pdf(file.path(gsea_sub_dir, sprintf("heatmap_pval_%s_big.pdf", co_label)), width=width + 8, height=height + 2)
print(hm)
dev.off()


mat_cor = cor(mat)
mat_cor[mat_cor>.99] = NA
max_val = .5
 cols_cur = colorRamp2(breaks=c(-1*max_val, 0, max_val), colors = c(muted("blue"), "white", muted("red")))
 
height = width = 1
#height = width * nrow(mat) / ncol(mat)

hm = Heatmap(mat_cor, 
             cluster_columns=F,
             col = cols_cur,
             cluster_rows = F,
             width = unit(width, "in"),
             height = unit(height, "in"))
print(hm)


pdf(file.path(gsea_sub_dir, sprintf("heatmap_pval_%s_cor.pdf", co_label)), width=width + 8, height=height + 2)
print(hm)
dev.off()


})
```

```{r}
coef_to_use = "kl_mean_3"
coefs_to_use = list(c("kl_mean_2", "kl_mean_3"), c("kl_mean_2"), c("kl_mean_3"))
walk(coefs_to_use, function(coef_to_use) { 
#walk(positions, function(p) {
  # modules_cur = node_infos[[p]] %>% select(contains(scale_to_use)) %>% unlist() %>% as.character() %>% c() %>% unique()
  # modules_cur = modules_cur[!is.na(modules_cur)]
  
  # res2_cur =  %>%
  #   filter(coef == coef_to_use) %>% 
  #   mutate(sig = case_when((logFC_mean < 0) & (logFC_mean < logFC_rand_q01) ~ "YES",
  #                          (logFC_mean > 0) & (logFC_mean > logFC_rand_q99) ~ "YES",
  #                          TRUE ~ "NO")) 
  # res2_cur_filt = res2_cur %>%
  #   filter(sig=="YES") %>%
  #   filter(module %in% modules_cur)
  
  gsea_res_ind_select = gsea_res_ind %>% 
    filter(coef %in% coef_to_use) %>% 
  #  distinct(position, .keep_all=T) %>%
    mutate(p.adjust_sign = -1 * log10(padj)) %>%
    mutate(NES_sign = sign(NES)) %>%
    filter(padj<.1) %>% 
    group_by(position, NES_sign) %>%
    top_n(5, abs(NES))
  
  gsea_res_ind_filt = gsea_res_ind %>% 
    filter(coef %in% coef_to_use) %>% 
    filter(pathway %in% gsea_res_ind_select$pathway) %>%
    distinct(position, pathway, .keep_all=T) %>%
     mutate(pathway = sub("^GO", "", pathway ),
             pathway = gsub("_", " ", pathway)) %>% 
    mutate(pathway = case_when(grepl("NONSENSE MEDIATED DECAY",pathway) ~ "NONSENSE MEDIATED DECAY",
                               grepl("SERINE THREONINE KINASE", pathway) ~ "SERINE THREONINE KINASE SIGNALING",
                               TRUE ~ pathway))
  #modules_to_plot = unique(res2_cur_filt$module) 
  
  go_res_mat = gsea_res_ind_filt %>% select(position, pathway, NES) %>%
    pivot_wider(names_from = "position", values_from = "NES", values_fill = list(NES= 0)) %>%
    column_to_rownames("pathway") %>% as.matrix()
  
  hc_row = hclust(dist(go_res_mat), method="complete")
  hc_row_order = hc_row$labels[hc_row$order]
  
  
  gsea_res_ind_filt = gsea_res_ind_filt %>% 
    mutate(leadingEdge_len = map_int(leadingEdge, length))
  
  gsea_res_ind_filt = gsea_res_ind_filt %>%
    mutate(pathway = factor(pathway, levels=hc_row_order)) %>%
    mutate(position = factor(position, levels=c("ra", "arco", "hvc", "ncl", "lman", "nido", "x", "stri"))) %>%
    complete(pathway, position, fill=list(NES=NA, size=NA, leadingEdge_len = NA))

  size_limits = seq(round(min(gsea_res_ind_filt$size, na.rm=T), -1), round(max(gsea_res_ind_filt$size, na.rm=T), -1), by=50)
    size_limits = seq(round(min(gsea_res_ind_filt$leadingEdge_len, na.rm=T), -1), round(max(gsea_res_ind_filt$leadingEdge_len, na.rm=T), -1), by=20)
  #size_limits = seq(-50, 250, 50)
  #print(size_limits)
  gg = ggplot(gsea_res_ind_filt, aes(position, pathway)) + 
    geom_point(aes(size=leadingEdge_len, color=NES)) + 
    #scale_color_viridis() + 
    scale_color_gradient2(low=muted("blue"), high=muted("red")) +
    scale_size_continuous(range=c(.5, 3), breaks=size_limits) + 
    labs(x = "", y ="") + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle=90, vjust=.5, hjust=1, size=6),
          axis.text.y = element_text(hjust=1, size=6),
          axis.ticks = element_line(size=.25),
          panel.border = element_blank()) + 
      coord_fixed()
  print(gg)
  ncols = ncol(go_res_mat)
  nrows = nrow(go_res_mat)
  
  width = ncols / 10
  height = nrows / 10
  gg = set_panel_size(gg, width = unit(width, "in"), height = unit(height, "in"))
  save_plot(file.path(gsea_sub_dir, sprintf("go_terms_by_module_%s.pdf", paste(coef_to_use, collapse="-"))), gg, base_height = (nrows / 2) + 4, base_width = (ncols / 2) + 4)
})
```

###### Plot leading edges
```{r}
pathways_to_use = c("GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY", "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE")
coef_to_use = c("kl_mean_3")
p = "ra"
positions_to_use = c("ra", "hvc")
walk(positions_to_use, function(p) {
  print(p)
  walk(pathways_to_use, function(pathway_to_use) {
    print(pathway_to_use)
    gsea_res_ind1 = gsea_res_ind %>% 
      filter(pathway==pathway_to_use) %>%
      filter(coef==coef_to_use) %>%
      filter(position==p) 
    genes = unique(gsea_res_ind1$leadingEdge[[1]])
    
    res_gsea = res %>% 
      
      filter(coef==coef_to_use)  %>%
      filter(position==p) %>%
      filter(gene_id %in% genes) %>%
      filter(logFC>0) %>% 
      arrange(logFC) %>%
      distinct(gene_id, .keep_all=T) %>% 
      mutate(gene_id = factor(gene_id, levels=gene_id))
    
    gg = ggplot(res_gsea, aes(gene_id, logFC)) + 
      geom_point() + 
      coord_flip() + 
      theme(axis.text=element_text(size=5)
      )
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("go_leadingEdge_%s_%s_%s.pdf", p, coef_to_use, pathway_to_use)), gg, base_height=4, base_width=1)
  })
})
```

Plot leading edge split by position
```{r}

p = "ra"
pathway_to_use = "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY"
coef_to_use = c("kl_mean_3")

gsea_res_ind1 = gsea_res_ind %>% 
  filter(pathway==pathway_to_use) %>%
  filter(coef==coef_to_use) %>%
  filter(position==p) 
genes = unique(gsea_res_ind1$leadingEdge[[1]])[19:24]
genes = c("JUN", "KLF15", "IRF2BP2", "PBX1", "LHX2", "NOTCH1", "PLAGL2")

df = FetchData(se1, c(genes), slot="data") %>%
  as.data.frame() %>%
  rownames_to_column(var="id") %>%
  pivot_longer(cols=-id, names_to="gene_id", values_to="value") %>%
  mutate(id = as.numeric(id)) %>% 
  left_join(info2 %>% select(id, position, procedure2)) %>%
  mutate(position = factor(position, levels=position_levels),
         gene_id = factor(gene_id, levels=genes)) #%>%

df_id = df %>% distinct(id, position) %>%
  arrange(position) 
id_levels = df_id$id

df = df %>%
#  filter(position == p) %>% 
  mutate(id = factor(id, levels=id_levels))
df_sum = df  %>%
  #filter(gene_id %in% genes) %>% 
  group_by(gene_id, position) %>%
  summarize(value_mean = mean(value))
gg = ggplot(df, aes(position, value, color=position)) + 
  geom_point(size = .5, alpha=.1, shape=20, position=position_jitter(width=.1)) + 
  geom_point(data=df_sum, color="black", size=.5, aes(position, value_mean)) + 
  scale_color_manual(values=position_colors) + 
  facet_wrap(~gene_id) + 
  labs(x = "", y = "log expression") + 
  theme_bw() + 
  theme(
    legend.position="none",
    #strip.text = element_blank(),
    strip.background = element_blank(),
    axis.title.y = element_text(size=6),
    axis.title.x = element_blank(),
    axis.text.y = element_text(size=5),
    axis.text.x = element_blank(),
    axis.line = element_blank(),
    axis.ticks.y = element_line(size=.25),
    axis.ticks.x = element_blank(),
    panel.grid=element_blank())
gg
#save_plot(file.path(out_dir, "points_by_position_crh.pdf"), gg, base_height=.75, base_width=1, ncol=length(to_plot_manual), nrow=1)
```

Plot leading edge across kl_mean, breakout
```{r}

p = "x"
pathway_to_use = "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY"
pathways_to_use = c(
#  "GO_GLIAL_CELL_DIFFENTIATION",
  "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY",
  "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE",
  "GO_RESPIRASOME",
  "GO_STRUCTURAL_CONSTITUENT_OF_RIBOSOME"
  )
coef_to_use = c("kl_mean_3")

nrows = 5
walk(positions, function(p) {
  walk(pathways_to_use, function(pathway_to_use) {
    print(pathway_to_use)
    gsea_res_ind1 = gsea_res %>% 
      filter(pathway==pathway_to_use) %>%
      filter(coef==coef_to_use) %>%
      filter(position==p) 
    genes = unique(gsea_res_ind1$leadingEdge[[1]])[1:20]
    #genes = c("JUN", "KLF15", "IRF2BP2", "PBX1", "LHX2", "NOTCH1", "PLAGL2")
    
    line_color = position_colors[p]
    df = FetchData(se1, c(genes), slot="data") %>%
      as.data.frame() %>%
      rownames_to_column(var="id") %>%
      pivot_longer(cols=-id, names_to="gene_id", values_to="value") %>%
      mutate(id = as.numeric(id)) %>% 
      left_join(info2 %>% select(id, position, procedure2)) %>%
      mutate(position = factor(position, levels=position_levels),
             gene_id = factor(gene_id, levels=genes)) #%>%
    
    df_id = df %>% 
      distinct(id, position) %>%
      arrange(position) 
    id_levels = df_id$id
    
    df = df %>%
      filter(position == p) %>% 
      #mutate(id = factor(id, levels=id_levels)) %>%
      left_join(md)
    df_sum = df  %>%
      group_by(gene_id, kl_mean_log_scale_cut2_proc2) %>%
      summarize(value_mean = mean(value))
    gg = ggplot(df, aes(kl_mean_log_scale_cut2_proc2, value)) + 
      geom_point(size = .5, alpha=.5, shape=20, position=position_jitter(width=.1)) + 
      geom_point(data=df_sum, color=line_color, size=.5, aes(kl_mean_log_scale_cut2_proc2, value_mean)) + 
      geom_line(data=df_sum, color=line_color, aes(kl_mean_log_scale_cut2_proc, value_mean,  group=gene_id)) +
      #scale_color_manual(values=position_colors) + 
      facet_wrap(~gene_id, scales="free", nrow=nrows) + 
      labs(x = "", y = "log expression") + 
      theme_bw() + 
      theme(
        legend.position="none",
        strip.text = element_text(size=6, face="italic", hjust=0, vjust=1),
        strip.background = element_blank(),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.text.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        #panel.grid=element_blank()
      )
    print(gg)
    save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2_%s_%s.pdf", p, pathway_to_use)), gg, base_height=.75, base_width=.75, ncol=length(genes)/nrows, nrow=nrows)
  })
})

```

Plot leading edge across kl_mean, grouped
```{r}

p = "x"
pathway_to_use = "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY"
pathways_to_use = c(
  "GO_GLIAL_CELL_DIFFERENTIATION",
  "GO_NEURON_SPINE",
  "GO_PRIMARY_ACTIVE_TRANSMEMBRANE_TRANSPORTER_ACTIVITY",
  "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY",
  "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE"#,
 # "GO_RESPIRASOME",
#  "GO_STRUCTURAL_CONSTITUENT_OF_RIBOSOME"
  )
coef_to_use = c("kl_mean_3")

nrows = 5
positions = as.character(positions)
walk(positions, function(p) {
  walk(pathways_to_use, function(pathway_to_use) {
    print(pathway_to_use)
    gsea_res_ind1 = gsea_res %>% 
      filter(pathway==pathway_to_use) %>%
      filter(coef==coef_to_use) %>%
      filter(position==p) 
    genes = unique(gsea_res_ind1$leadingEdge[[1]])[1:20]
    
    line_color = position_colors[p]
    df = FetchData(se1, c(genes), slot="data") %>%
      as.data.frame() %>%
      rownames_to_column(var="id") %>%
      pivot_longer(cols=-id, names_to="gene_id", values_to="value") %>%
      mutate(id = as.numeric(id)) %>% 
      left_join(md %>% select(id, position, procedure2)) %>%
      mutate(position = factor(position, levels=position_levels),
             gene_id = factor(gene_id, levels=genes)) #%>%
    
    df_id = df %>% 
      distinct(id, position) %>%
      arrange(position) 
    id_levels = df_id$id
    
    df = df %>%
      filter(position == p) %>% 
      left_join(md %>% select(id, position, kl_mean_log_scale_cut2_proc2))
    
    df = df %>%
      group_by(gene_id) %>% 
      mutate(
        baseline = kl_mean_log_scale_cut2_proc2=="intact-1",
        value_rel = value - mean(value[baseline]),
        value_z = (value - mean(value[baseline])) / sd(value[baseline]))
    
    df_sum = df %>% 
      group_by(gene_id, kl_mean_log_scale_cut2_proc2) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean))
    
    df_group_sum = df  %>%
      group_by(kl_mean_log_scale_cut2_proc2) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean)) %>% 
      ungroup()
    
    gg = ggplot() + 
      #geom_point(size = .5, alpha=.5, shape=20, position=position_jitter(width=.1)) + 
      #geom_point(data=df_sum, color=line_color, size=.5, aes(kl_mean_log_scale_cut2_proc, value_mean)) + 
            geom_hline(yintercept=0, linetype=1, size=.25) + 
      geom_line(data=df_sum, color="grey", alpha=.5, aes(kl_mean_log_scale_cut2_proc2, value_z_mean,  group=gene_id), size=.25) +
      geom_line(data=df_group_sum, color=line_color, aes(kl_mean_log_scale_cut2_proc2, value_z_mean, group=1)) +

      #scale_color_manual(values=position_colors) + 
      #facet_wrap(~gene_id, scales="free", nrow=nrows) + 
      labs(x = "", y = "expression\nz-scored rel. low Song DKL") + 
      theme_bw() + 
      theme(
        legend.position="none",
        strip.text = element_text(size=6, face="italic", hjust=0, vjust=1),
        strip.background = element_blank(),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.text.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid=element_blank()
      )
    print(gg)
    width = .5
    height = .6
    gg = set_panel_size(gg, width = unit(width, "in"), height = unit(height, "in"))
    save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2_%s_%s_grouped.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
    
    gg = gg + geom_text_repel(data=df_sum %>% filter(kl_mean_log_scale_cut2_proc2=="deaf-3"), aes(x = 3, value_z_mean, label=gene_id), size=4)
    
        save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2_%s_%s_grouped_with-labels.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
  })
})

```

Plot leading edge across kl_mean, grouped, just intact-1 and deaf-3
```{r}

p = "x"
pathway_to_use = "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY"
pathways_to_use = c(
  "GO_GLIAL_CELL_DIFFERENTIATION",
  "GO_NEURON_SPINE",
  "GO_PRIMARY_ACTIVE_TRANSMEMBRANE_TRANSPORTER_ACTIVITY",
  "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY",
  "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE"#,
 # "GO_RESPIRASOME",
#  "GO_STRUCTURAL_CONSTITUENT_OF_RIBOSOME"
  )
coef_to_use = c("kl_mean_3")

nrows = 5
positions = as.character(positions)
walk(positions, function(p) {
  walk(pathways_to_use, function(pathway_to_use) {
    print(pathway_to_use)
    gsea_res_ind1 = gsea_res %>% 
      filter(pathway==pathway_to_use) %>%
      filter(coef==coef_to_use) %>%
      filter(position==p) 
    genes = unique(gsea_res_ind1$leadingEdge[[1]])[1:20]
    
    line_color = position_colors[p]
    df = FetchData(se1, c(genes), slot="data") %>%
      as.data.frame() %>%
      rownames_to_column(var="id") %>%
      pivot_longer(cols=-id, names_to="gene_id", values_to="value") %>%
      mutate(id = as.numeric(id)) %>% 
      left_join(md %>% select(id, position, procedure2)) %>%
      mutate(position = factor(position, levels=position_levels),
             gene_id = factor(gene_id, levels=genes))
    

    
    df_id = df %>% 
      distinct(id, position) %>%
      arrange(position) 
    id_levels = df_id$id
    
    df = df %>%
      filter(position == p) %>% 
      left_join(md %>% select(id, position, kl_mean_log_scale_cut2_proc2))
    
    df = df %>% filter(kl_mean_log_scale_cut2_proc2 %in% c("intact-1", "deaf-3"))
    
    df = df %>%
      group_by(gene_id) %>% 
      mutate(
        baseline = kl_mean_log_scale_cut2_proc2=="intact-1",
        value_rel = value - mean(value[baseline]),
        value_z = (value - mean(value[baseline])) / sd(value[baseline]))
    
    df_sum = df %>% 
      group_by(gene_id, kl_mean_log_scale_cut2_proc2) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean))
    
    df_group_sum = df  %>%
      group_by(kl_mean_log_scale_cut2_proc2) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean)) %>% 
      ungroup()
    
    gg = ggplot() + 
      #geom_point(size = .5, alpha=.5, shape=20, position=position_jitter(width=.1)) + 
      #geom_point(data=df_sum, color=line_color, size=.5, aes(kl_mean_log_scale_cut2_proc, value_mean)) + 
            geom_hline(yintercept=0, linetype=1, size=.25) + 
      geom_line(data=df_sum, color="grey", alpha=.5, aes(kl_mean_log_scale_cut2_proc2, value_z_mean,  group=gene_id), size=.25) +
      geom_line(data=df_group_sum, color=line_color, aes(kl_mean_log_scale_cut2_proc2, value_z_mean, group=1)) +

      #scale_color_manual(values=position_colors) + 
      #facet_wrap(~gene_id, scales="free", nrow=nrows) + 
      labs(x = "", y = "expression\nz-scored rel. low Song DKL") + 
      theme_bw() + 
      theme(
        legend.position="none",
        strip.text = element_text(size=6, face="italic", hjust=0, vjust=1),
        strip.background = element_blank(),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.text.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid=element_blank()
      )
    print(gg)
    width = .5
    height = .6
    gg = set_panel_size(gg, width = unit(width, "in"), height = unit(height, "in"))
    save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2-low-hi_%s_%s_grouped.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
    
    gg = gg + geom_text_repel(data=df_sum %>% filter(kl_mean_log_scale_cut2_proc2=="deaf-3"), aes(x = 3, value_z_mean, label=gene_id), size=4)
    
        save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2-low-hi_%s_%s_grouped_with-labels.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
  })
})

```

###### Plot leading edges, heatmap across positions
```{r}
co = "kl_mean_3"

pathways_to_use = c(
  "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY",
  "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE",
  "GO_RESPIRASOME",
  "GO_STRUCTURAL_CONSTITUENT_OF_RIBOSOME"
  )
coef_to_use = c("kl_mean_3")

nrows = 5

positions_to_use = c("ra", "hvc", "lman", "x")
walk(pathways_to_use, function(pathway_to_use) {
  print(pathway_to_use)
  gsea_res_ind1 = gsea_res %>% 
    filter(pathway==pathway_to_use) %>%
    filter(coef==coef_to_use) 
  
  genes = map(positions_to_use, function(p) {
    gsea_res_ind2 = gsea_res_ind1 %>% filter(position==p)
    unique(gsea_res_ind2$leadingEdge[[1]])[1:20]
  }) %>% unlist() %>% unique()
  
  res_sig = res %>%
    filter(coef==co) %>% 
    filter(position %in% positions_to_use) %>%
    filter(gene_id %in% genes)
  

  value_var = "logFC"
  res_cur = res %>% 
    filter(gene_id %in% res_sig$gene_id) %>% 
    filter(coef==co) %>%
    group_by(position, gene_id) %>%
    summarize(logFC = mean(logFC),
              logFC_rescale = mean(logFC_rescale))
  mat = acast(res_cur, gene_id~position, value.var=value_var)
  mat = mat[,c("ra", "hvc", "lman", "x")]
  colnames(mat) = toupper(colnames(mat))
  
  mat_var = apply(mat, 1, var)
  #mat = mat[mat_var>quantile(mat_var, .25),]
  width = .25
  height = width * nrow(mat) / ncol(mat)
  
  max_val = max(c(abs(min(mat)), max(mat)))
  max_val = 1.5
  #length_out = 9
 # cols_cur = colorRamp2(breaks=seq(-1*max_val, max_val, length.out=length_out), colors = rev(brewer_pal(palette=cols[[co]])(length_out)))
  cols_cur = colorRamp2(breaks=c(-1*max_val, 0, max_val), colors = c(muted("purple"), "white", muted("orange")))
  hm = Heatmap(mat, col=cols_cur,
               clustering_method_rows = "ward.D", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=5), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=5)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(gsea_sub_dir, sprintf("heatmap_%s_song-only.pdf", pathway_to_use)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  #hm
  
  hm_t = Heatmap(t(mat), col=cols_cur,
               clustering_method_columns  = "average", 
               cluster_rows = F,
               height = unit(width, "in"),
               width =unit( height, "in"),
               column_names_gp = gpar(fontsize=5), column_names_side = "top",
               row_names_gp = gpar(fontsize=5)
               #  clustering_method_columns = "ward.D"
  )
  
    pdf(file.path(gsea_sub_dir, sprintf("heatmap_%s_song-only_horiz.pdf", pathway_to_use)), width=height+2, height=width+2)
  print(hm_t)
  dev.off()
  #hm
})
```

###### Plot top GO terms, points
```{r}

walk(positions, function(p) {
  print(p)
  walk(coefs, function(co) {
    print(co)
    co_label = sub("position%s.", "", co)
    gsea_res_ind_filt = gsea_res_ind %>% 
      filter(position==p) %>%
      filter(coef==co) %>%
      filter(padj<.2)
    
    gsea_res_ind_top = gsea_res_ind %>%
      filter(position==p) %>%
      filter(padj<.2) %>%
      filter(coef==co) %>% 
      # group_by(position) %>% 
      mutate(NES_sign = sign(NES)) %>% 
      group_by(NES_sign) %>% 
      top_n(10, abs(NES)) %>%
      ungroup() %>%
      distinct(pathway)
    
    gsea_res_filt = gsea_res %>% 
      filter(position==p) %>%
      filter(coef==co) %>% 
      filter(pathway %in% gsea_res_ind_top$pathway)# %>%
    #  filter(padj<.2)
    
    if (nrow(gsea_res_filt) < 2)
      return(NULL)
    tmp = gsea_res_filt %>% 
      
      mutate(class = case_when(grepl("HORMONE|NEUROTR|PEPTI", pathway) ~ "modulation",
                               grepl("OXID|NADH|MITO|RESPIR|PROTON|REDOX", pathway) ~ "respiration",
                               grepl("RIBO|PROTEIN_LOCALIZ|TRANSLATION|RRNA", pathway) ~ "translation",
                               TRUE ~ "none"),
      ) %>%
      mutate(pathway = sub("GO_", "", pathway),
             pathway = case_when(grepl("NONSENSE", pathway) ~ "NONSENSE_MEDIATED_DECAY",
                                 #  grepl("ESTABLISHMENT_OF_PROTEIN", pathway) ~ "PROTEIN_LOCALIZATION_TO_ER",
                                 TRUE ~ pathway),
             pathway = gsub("_", " ", pathway),
             pathway = gsub("\\.", " ", pathway), 
             pathway = str_to_title(pathway),
      )  %>%
      arrange(NES) %>%
      mutate(pathway = factor(pathway, levels=pathway)) 
    
    #cols =c("black",  pal_aaas()(3))
    #names(cols) = c("none", "respiration",  "modulation", "translation")
    gg = ggplot(tmp, aes(NES, pathway)) + 
      
      geom_segment(aes(x=min(NES),xend = NES, y=pathway, yend=pathway), size=.25, linetype=3) + 
      geom_segment(aes(x = 0, xend = NES, y=pathway, yend=pathway), size=.25) +
      geom_point() + 
      theme_light() + 
      theme(axis.text = element_text(size=5),
            axis.title = element_text(size=6),
            axis.line.x = element_line(size=.25),
            axis.line.y = element_blank(),
            axis.ticks.x = element_line(size=.25),
            axis.ticks.y = element_blank(),
            legend.position = "bottom",
            legend.direction="horizontal",
            legend.text = element_text(size=5)) + 
      labs(y="") + 
      scale_color_manual(values=cols, name="")
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("point_%s_%s.pdf", co_label, p)), gg, base_height=2.5, base_width=5)
  })
})
```

###### MSIGDBR

```{r}
library(msigdbr)
msigdbr_df = msigdbr() %>%
  filter(gs_subcat %in% c("CP:BIOCARTA", "GO:MF", "GO:BP", "GO:CC", "CP:KEGG", "CP", "CP:REACTOME", "CP:WIKIPATHWAYS", "TFT:GTRD"))
msigdbr_lists = split(msigdbr_df$gene_symbol, msigdbr_df$gs_name)
msigdbr_lists = c(msigdbr_lists, gray_gmt)
```

```{r}
# genes_avail = bitr(unique(res$gene_id), fromType = "SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db") %>%
#   mutate(gene_id = SYMBOL, entrezid = ENTREZID)
# res_cur = res %>% left_join(genes_avail)
```

```{r message=FALSE, warning=FALSE}

gsea_res_fname = file.path(gsea_sub_dir, "gsea_msigdbr_res.qs")
gsea_res_ind_fname = file.path(gsea_sub_dir, "gsea_msigdbr_res_ind.qs")

redo = F
if (redo) {
  
gsea_res = future_map(positions, function(p) {
  print(p)
  res_cur = res %>% filter(position == p)
  map(coefs, function(co) {
    res_cur1 = res_cur %>% filter(coef==co)
    rnks = res_cur1$t
    names(rnks) = res_cur1$gene_id
    rnks = sort(rnks)
    gsea_res = fgsea(pathways = msigdbr_lists,
                     stats = rnks,
                     minSize=20,
                     maxSize=200,
                     nperm=1000)
    
    return(gsea_res)
 
  }) %>% set_names(coefs) %>% bind_rows(.id="coef")
})  %>% set_names(positions) %>% bind_rows(.id="position")

gsea_res_ind = future_map(positions, function(p) {
  print(p)
  res_cur = res %>% filter(position == p)
  gsea_res_cur = gsea_res %>% filter(position==p)
  map(coefs, function(co) {
    res_cur1 = res_cur  %>% filter(coef==co)
    gsea_res_cur1 = gsea_res_cur %>% filter(coef==co)
    rnks = res_cur1$t
    names(rnks) = res_cur1$gene_id
    rnks = sort(rnks)
    gsea_res_filt = gsea_res_cur1 %>% filter(padj<.2) %>% arrange(pval)
    collapsedPathways = collapsePathways(gsea_res_filt, pathways = msigdbr_lists, stats = rnks, pval.threshold = .001)
    gsea_res_ind = gsea_res_cur1 %>% filter(pathway %in% collapsedPathways$mainPathways) %>%
      arrange(-NES)
    gsea_res_ind
  }) %>% set_names(coefs) %>% bind_rows(.id="coef")
}) %>% bind_rows()

  qsave(gsea_res, gsea_res_fname)
  qsave(gsea_res_ind, gsea_res_ind_fname)
} else {
  gsea_res = qread(gsea_res_fname)
  gsea_res_ind = qread(gsea_res_ind_fname)
}
```

```{r}
padj_thresh = 0.2
positions = unique(gsea_res_ind$position)
coefs = unique(gsea_res_ind$coef)
walk(positions, function(p) {
  walk(coefs, function(co) {
    co_label = sub("position%s\\.", "", co)
    gsea_res_filt = gsea_res_ind %>% filter(padj<padj_thresh) %>% 
      filter(position==p, coef==co) %>%
      mutate(padj_log = -1 * log10(padj),
             p_log = -1 * log10(pval)) %>%
    #  filter(NES>0) %>% 
      mutate(size1 = size) %>%
      # mutate(to_label = ifelse(pathway %in% gsea_res_ind_filt$pathway, pathway, NA)) %>%
      mutate(leadingEdge_len = map_int(leadingEdge, length)) %>%
      mutate(leading_frac = leadingEdge_len / size) %>% 
      mutate(pathway1 = sub("^GO", "", pathway ),
             pathway1 = gsub("_", " ", pathway1),
             pathway1 = str_to_title(pathway1))
    
    if (nrow(gsea_res_filt) < 10) 
      return()
    gg = ggplot(gsea_res_filt, aes(NES, p_log, size=leading_frac)) + 
      geom_point() + 
      
      theme_classic() 
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_no_text.pdf", padj_thresh, p, co_label)), gg, base_width=6, base_height=6)
    
    gg = gg + theme(legend.position="none")
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_no_text_no_legend.pdf", padj_thresh, p, co_label)), gg, base_width=6, base_height=6)
    
    gg = gg +  
      geom_text(aes(NES, p_log, label = pathway1), inherit.aes = F) 
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_text.pdf", padj_thresh,  p, co_label)), gg, base_width=6, base_height=6)
  })
})
```

```{r}
walk(coefs, function(co) {
co_label = sub("position%s.", "", co)
gsea_res_ind_filt = gsea_res_ind %>% 
  filter(padj<.2) %>%
  filter(coef==co) %>%
  group_by(position) %>%
  top_n(10, abs(NES))
  # gsea_res_ind_top = gsea_res_ind %>% filter(padj<.2) %>%
  #   top_n(10, NES)
##gsea_res_filt = gsea_res %>% filter(padj<.2)
gsea_res_filt = gsea_res %>% 
  #filter(padj<.2) %>%
  filter(pathway %in% gsea_res_ind_filt$pathway) %>%
  filter(position %in% c("ra", "hvc", "lman", "x"))

mat = gsea_res_filt %>% filter(coef==co) %>%
  select(position, pathway, NES) %>%
  pivot_wider(names_from = "position", values_from = "NES", values_fill=list(NES=0)) %>%
  column_to_rownames("pathway") %>% 
  as.matrix() 

mat = mat[,intersect(position_levels, colnames(mat))]

width = 1
height = 8
hm = Heatmap(mat, cluster_columns=F, width = unit(width, "in"),
             height = unit(height, "in"))
hm

pdf(file.path(gsea_sub_dir, sprintf("heatmap_%s.pdf", co_label)), width=width + 8, height=height + 2)
print(hm)
dev.off()
})
```

```{r}

walk(positions, function(p) {
  print(p)
  walk(coefs, function(co) {
    print(co)
    co_label = sub("position%s.", "", co)
    gsea_res_ind_filt = gsea_res_ind %>% 
      filter(position==p) %>%
      filter(coef==co) %>%
      filter(padj<.2)
    gsea_res_ind_top = gsea_res_ind %>%
      filter(position==p) %>%
      filter(padj<.2) %>%
      filter(coef==co) %>% 
      # group_by(position) %>% 
    # top_n(20, abs(NES)) %>%
      ungroup() %>%
      distinct(pathway)
    
    gsea_res_filt = gsea_res %>% 
      filter(position==p) %>%
      filter(coef==co) %>% 
      filter(pathway %in% gsea_res_ind_top$pathway)# %>%
    #  filter(padj<.2)
    
    if (nrow(gsea_res_filt) < 2)
      return(NULL)
    tmp = gsea_res_filt %>% 
      
      mutate(class = case_when(grepl("HORMONE|NEUROTR|PEPTI", pathway) ~ "modulation",
                               grepl("OXID|NADH|MITO|RESPIR|PROTON|REDOX", pathway) ~ "respiration",
                               grepl("RIBO|PROTEIN_LOCALIZ|TRANSLATION|RRNA", pathway) ~ "translation",
                               TRUE ~ "none"),
      ) %>%
      mutate(pathway = sub("GO_", "", pathway),
             pathway = case_when(grepl("NONSENSE", pathway) ~ "NONSENSE_MEDIATED_DECAY",
                                 #  grepl("ESTABLISHMENT_OF_PROTEIN", pathway) ~ "PROTEIN_LOCALIZATION_TO_ER",
                                 TRUE ~ pathway),
             pathway = gsub("_", " ", pathway),
             pathway = gsub("\\.", " ", pathway), 
             pathway = str_to_title(pathway),
      )  %>%
      arrange(NES) %>%
      mutate(pathway = factor(pathway, levels=pathway)) 
    
    cols =c("black",  pal_aaas()(3))
    names(cols) = c("none", "respiration",  "modulation", "translation")
    gg = ggplot(tmp, aes(NES, pathway)) + 
      
      geom_segment(aes(x=min(NES),xend = NES, y=pathway, yend=pathway), size=.25, linetype=3) + 
      geom_segment(aes(x = 0, xend = NES, y=pathway, yend=pathway), size=.25) +
      geom_point() + 
      theme_light() + 
      theme(axis.text = element_text(size=5),
            axis.title = element_text(size=6),
            axis.line.x = element_line(size=.25),
            axis.line.y = element_blank(),
            axis.ticks.x = element_line(size=.25),
            axis.ticks.y = element_blank(),
            legend.position = "bottom",
            legend.direction="horizontal",
            legend.text = element_text(size=5)) + 
      labs(y="") + 
      scale_color_manual(values=cols, name="")
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("point_%s_%s.pdf", co_label, p)), gg, base_height=2.5, base_width=3.5)
  })
})
```

Bar
```{r}

walk(positions, function(p) {
  print(p)
  walk(coefs, function(co) {
    print(co)
    co_label = sub("position%s.", "", co)
    gsea_res_ind_filt = gsea_res_ind %>% 
      filter(position==p) %>%
      filter(coef==co) %>%
      filter(padj<.2)
    gsea_res_ind_top = gsea_res_ind %>%
      filter(position==p) %>%
      filter(padj<.2) %>%
      filter(coef==co) %>% 
      # group_by(position) %>% 
      top_n(20, abs(NES)) %>%
      ungroup() %>%
      distinct(pathway)
    
    gsea_res_filt = gsea_res %>% 
      filter(position==p) %>%
      filter(coef==co) %>% 
      filter(pathway %in% gsea_res_ind_top$pathway)# %>%
    #  filter(padj<.2)
    
    if (nrow(gsea_res_filt) < 2)
      return(NULL)
    tmp = gsea_res_filt %>% 
      
      mutate(class = case_when(grepl("HORMONE|NEUROTR|PEPTI", pathway) ~ "modulation",
                               grepl("OXID|NADH|MITO|RESPIR|PROTON|REDOX", pathway) ~ "respiration",
                               grepl("RIBO|PROTEIN_LOCALIZ|TRANSLATION|RRNA", pathway) ~ "translation",
                               TRUE ~ "none"),
      ) %>%
      mutate(pathway = sub("GO_", "", pathway),
             pathway = case_when(grepl("NONSENSE", pathway) ~ "NONSENSE_MEDIATED_DECAY",
                                 #  grepl("ESTABLISHMENT_OF_PROTEIN", pathway) ~ "PROTEIN_LOCALIZATION_TO_ER",
                                 TRUE ~ pathway),
             pathway = gsub("_", " ", pathway),
             pathway = gsub("\\.", " ", pathway), 
             pathway = str_to_title(pathway),
      )  %>%
      distinct(pathway, .keep_all=T) %>% 
      arrange(NES) %>%
      mutate(pathway = factor(pathway, levels=pathway)) 
    
    cols =c("black",  pal_aaas()(3))
    names(cols) = c("none", "respiration",  "modulation", "translation")
    gg = ggplot(tmp, aes(NES, pathway)) + 
      geom_bar(stat="identity") + 
      #geom_segment(aes(x=min(NES),xend = NES, y=pathway, yend=pathway), size=.25, linetype=3) + 
      #geom_segment(aes(x = 0, xend = NES, y=pathway, yend=pathway), size=.25) +
      #geom_point() + 
      theme_light() + 
      theme(axis.text = element_text(size=5),
            axis.title = element_text(size=6),
            axis.line.x = element_line(size=.25),
            axis.line.y = element_blank(),
            axis.ticks.x = element_line(size=.25),
            axis.ticks.y = element_blank(),
            legend.position = "bottom",
            legend.direction="horizontal",
            legend.text = element_text(size=5)) + 
      labs(y="") + 
      scale_color_manual(values=cols, name="")
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("bar_%s_%s.pdf", co_label, p)), gg, base_height=2.5, base_width=10)
  })
})
```

##### With no metabolic/ribo
```{r}
gsea_sub_dir = file.path(gsea_dir, "no_metabolic")
gsea_run_dir = file.path(gsea_sub_dir, "gsea_runs")
dir.create(gsea_run_dir, recursive = T)

gmt_dir = "~/data/gene_lists/gmt"
gmt_files = list.files(gmt_dir, full.names = T)

#dest_gmt_files = dest_gmt_files[!grepl("Drug|Disease", dest_gmt_files)]
gmt_files = gmt_files[grep("c5.all.v7", gmt_files)]

gmt_pathway = gmtPathways(gmt_files)
to_exclude = names(gmt_pathway)
to_exclude = to_exclude[grep("ENDOPLAS|RIBOSOM|TRANSLATION|METABO|CATABOL|RESPIR|HYDROGEN|OXIDAT|OXIDO|MITOCH|PROTON|VIRAL|ELECTRON|ATP|RRNA|BIOSYNTHETIC|INORGANIC|INNER_MEMBRANE", to_exclude)]
gmt_pathway = gmt_pathway[!(names(gmt_pathway) %in% to_exclude)]
```

```{r}

gray_gmt_fname = "~/data/gene_lists/gray_2018_arg.xls.gmt"
gray_gmt = gmtPathways(gray_gmt_fname)
names(gray_gmt) = make.names(names(gray_gmt))
gmt_gray_pathway = c(gmt_pathway, gray_gmt)
```

```{r}
positions = as.character(unique(res$position))
coefs = unique(res$coef)

plan(multisession, workers = length(positions))
options(future.globals.maxSize = 1024^2 * 5000)
```



```{r message=FALSE, warning=FALSE}

gsea_res_fname = file.path(gsea_sub_dir, "gsea_res.qs")
gsea_res_ind_fname = file.path(gsea_sub_dir, "gsea_res_ind.qs")

redo = T
if (redo) {
  positions = unique(res$position)
  
gsea_res = future_map(positions, function(p) {
  print(p)
  res_cur = res %>% filter(position == p)
  map(coefs, function(co) {
    print(p)
    res_cur1 = res_cur  %>% filter(coef==co)
    rnks = res_cur1$t
    names(rnks) = res_cur1$gene_id
    rnks = sort(rnks)
    gsea_res = fgseaMultilevel(pathways = gmt_gray_pathway,
                     stats = rnks,
                     minSize=20,
                     maxSize=200#,
                     #nperm=1000
                     )
    
    return(gsea_res)
 
  }) %>% set_names(coefs) %>% bind_rows(.id="coef")
})  %>% set_names(positions) %>% bind_rows(.id="position")

gsea_res_ind = future_map(positions, function(p) {
  print(p)
  res_cur = res %>% filter(position == p)
  gsea_res_cur = gsea_res %>% filter(position==p)
  map(coefs, function(co) {
    res_cur1 = res_cur  %>% filter(coef==co)
    gsea_res_cur1 = gsea_res_cur %>% filter(coef==co)
    rnks = res_cur1$t
    names(rnks) = res_cur1$gene_id
    rnks = sort(rnks)
    gsea_res_filt = gsea_res_cur1 %>% filter(padj<.2) %>% arrange(pval)
    collapsedPathways = collapsePathways(gsea_res_filt, pathways = gmt_gray_pathway, stats = rnks, pval.threshold = .05)
    gsea_res_ind = gsea_res_cur1 %>% filter(pathway %in% collapsedPathways$mainPathways) %>%
      arrange(-NES)
    gsea_res_ind
  }) %>% set_names(coefs) %>% bind_rows(.id="coef")
}) %>% bind_rows()

  qsave(gsea_res, gsea_res_fname)
  qsave(gsea_res_ind, gsea_res_ind_fname)
} else {
  gsea_res = qread(gsea_res_fname)
  gsea_res_ind = qread(gsea_res_ind_fname)
}
```

###### Plot individual positions
```{r}
padj_thresh = 1
positions = unique(gsea_res_ind$position)
coefs = unique(gsea_res_ind$coef)
walk(positions, function(p) {
  walk(coefs, function(co) {
    co_label = sub("position%s\\.", "", co)
    
    gsea_res_ind_filt = gsea_res %>% 
      filter(coef==co) %>% 
      filter(position==p) %>% 
      mutate(p.adjust_sign = -1 * log10(padj)) %>%
      mutate(NES_sign = sign(NES)) %>%
      filter(padj<.2) %>% 
      group_by(NES_sign) %>%
      top_n(10, abs(NES)) %>%
      ungroup()
    
    gsea_res_filt = gsea_res_ind %>% filter(padj<padj_thresh) %>% 
      filter(position==p, coef==co) %>%
      mutate(padj_log = -1 * log10(padj),
             p_log = -1 * log10(pval)) %>%
    #  filter(NES>0) %>% 
      mutate(size1 = size) %>%
      mutate(to_label = ifelse(pathway %in% gsea_res_ind_filt$pathway, pathway, NA)) %>%
      mutate(leadingEdge_len = map_int(leadingEdge, length)) %>%
      mutate(leading_frac = leadingEdge_len / size) %>% 
      
      mutate(pathway1 = sub("^GO", "", to_label ),
             pathway1 = gsub("_", " ", pathway1),
             pathway1 = str_to_title(pathway1))
    
    if (nrow(gsea_res_filt) < 10) 
      return()
    gg = ggplot(gsea_res_filt, aes(NES, p_log, size=leading_frac)) + 
      geom_point(position=position_jitter()) + 
      
      theme_classic() 
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_no_text.pdf", padj_thresh, p, co_label)), gg, base_width=6, base_height=6)
    
    gg = gg + theme(legend.position="none")
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_no_text_no_legend.pdf", padj_thresh, p, co_label)), gg, base_width=6, base_height=6)
    
    gg = gg +  
      geom_text_repel(aes(NES, p_log, label = pathway1), inherit.aes = F, max.overlaps=20) 
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_text.pdf", padj_thresh,  p, co_label)), gg, base_width=6, base_height=6)
  })
})
```

###### Heatmap
```{r}
# coefs_to_use = c("kl_mean_2_hearing", "kl_mean_2_deaf", "kl_mean_3", "num_songs_euth_date", "num_songs_pre", "inter")
# walk(coefs_to_use, function(co) {
#   print(co)
#   co_label = sub("position%s.", "", co)
#  gsea_res_ind_filt = gsea_res %>% 
#     filter(coef==co) %>% 
#   #  distinct(position, .keep_all=T) %>%
#     mutate(p.adjust_sign = -1 * log10(padj)) %>%
#     mutate(NES_sign = sign(NES)) %>%
#     filter(padj<.2) %>% 
#     group_by(position, NES_sign) %>%
#     top_n(10, abs(NES)) %>%
#    ungroup()
#   #w
#   # gsea_res_ind_filt = gsea_res %>% 
#   #   filter(coef==co) %>% 
#   #   filter(pathway %in% gsea_res_ind_select$pathway)
#   
#   mat = gsea_res_ind_filt %>%
#     filter(coef==co) %>%
#     select(position, pathway, NES) %>%
#     pivot_wider(names_from = "position", values_from = "NES", values_fill=list(NES=0)) %>%
#     column_to_rownames("pathway") %>% 
#     as.matrix() 
#   
#   mat = mat[,intersect(position_levels, colnames(mat))]
#   
#   
#   height = 4
#   width = height * ncol(mat) / nrow(mat)
#   hm = Heatmap(mat, cluster_columns=F, 
#                width = unit(width, "in"),
#                height = unit(height, "in"))
#   print(hm)
#   
#   pdf(file.path(gsea_sub_dir, sprintf("heatmap_%s.pdf", co_label)), width=width + 8, height=height + 2)
#   print(hm)
#   dev.off()
# })
```

```{r}
coefs = c("kl_mean_2_hearing", "kl_mean_2_deaf", "kl_mean_3", "num_songs_euth_date", "num_songs_pre", "inter")
walk(coefs, function(co) {
  print(co)
co_label = sub("position%s.", "", co)
gsea_res_ind_filt = gsea_res_ind %>% 
  filter(padj<.1) %>%
  filter(coef==co) #%>%
 # group_by(position)# %>%
 # top_n(10, abs(NES))
  # gsea_res_ind_top = gsea_res_ind %>% filter(padj<.2) %>%
  #   top_n(10, NES)
##gsea_res_filt = gsea_res %>% filter(padj<.2)
gsea_res_filt = gsea_res %>% 
  #filter(padj<.2) %>%
  mutate(padj_signed = sign(NES) *  -1 * log10(padj)) %>% 
  filter(pathway %in% gsea_res_ind_filt$pathway)# %>%
 # filter(position %in% c("ra", "hvc", "lman", "x"))

mat = gsea_res_filt %>% filter(coef==co) %>%
  select(position, pathway, padj_signed) %>%
  pivot_wider(names_from = "position", values_from = "padj_signed", values_fill=list(NES=0)) %>%
  column_to_rownames("pathway") %>% 
  as.matrix() 

mat = mat[,intersect(position_levels, colnames(mat))]

max_val = 1.5
 cols_cur = colorRamp2(breaks=c(-1*max_val, 0, max_val), colors = c(muted("purple"), "white", muted("orange")))
 
width = .5
#height = width * nrow(mat) / ncol(mat)
height = 2.5
hm = Heatmap(mat, 
             cluster_columns=F, col = cols_cur,
             clustering_method_rows = "ward.D",
             width = unit(width, "in"),
             height = unit(height, "in"))
hm

pdf(file.path(gsea_sub_dir, sprintf("heatmap_pval_%s.pdf", co_label)), width=width + 8, height=height + 2)
print(hm)
dev.off()

width = 1
height = width * nrow(mat) / ncol(mat)
#height = 4
hm = Heatmap(mat, 
             cluster_columns=F, col = cols_cur,
             clustering_method_rows = "ward.D",
             width = unit(width, "in"),
             height = unit(height, "in"))
hm

pdf(file.path(gsea_sub_dir, sprintf("heatmap_pval_%s_big.pdf", co_label)), width=width + 8, height=height + 2)
print(hm)
dev.off()


mat_cor = cor(mat)
mat_cor[mat_cor>.99] = NA
max_val = .5
 cols_cur = colorRamp2(breaks=c(-1*max_val, 0, max_val), colors = c(muted("blue"), "white", muted("red")))
 
height = width = 1
#height = width * nrow(mat) / ncol(mat)

hm = Heatmap(mat_cor, 
             cluster_columns=F,
             col = cols_cur,
             cluster_rows = F,
             width = unit(width, "in"),
             height = unit(height, "in"))
print(hm)


pdf(file.path(gsea_sub_dir, sprintf("heatmap_pval_%s_cor.pdf", co_label)), width=width + 8, height=height + 2)
print(hm)
dev.off()


})
```

```{r}
coef_to_use = "kl_mean_3"
coefs_to_use = list(c("kl_mean_2", "kl_mean_3"), c("kl_mean_2"), c("kl_mean_3"))
walk(coefs_to_use, function(coef_to_use) { 
#walk(positions, function(p) {
  # modules_cur = node_infos[[p]] %>% select(contains(scale_to_use)) %>% unlist() %>% as.character() %>% c() %>% unique()
  # modules_cur = modules_cur[!is.na(modules_cur)]
  
  # res2_cur =  %>%
  #   filter(coef == coef_to_use) %>% 
  #   mutate(sig = case_when((logFC_mean < 0) & (logFC_mean < logFC_rand_q01) ~ "YES",
  #                          (logFC_mean > 0) & (logFC_mean > logFC_rand_q99) ~ "YES",
  #                          TRUE ~ "NO")) 
  # res2_cur_filt = res2_cur %>%
  #   filter(sig=="YES") %>%
  #   filter(module %in% modules_cur)
  
  gsea_res_ind_select = gsea_res_ind %>% 
    filter(coef %in% coef_to_use) %>% 
  #  distinct(position, .keep_all=T) %>%
    mutate(p.adjust_sign = -1 * log10(padj)) %>%
    mutate(NES_sign = sign(NES)) %>%
    filter(padj<.1) %>% 
    group_by(position, NES_sign) %>%
    top_n(5, abs(NES))
  
  gsea_res_ind_filt = gsea_res_ind %>% 
    filter(coef %in% coef_to_use) %>% 
    filter(pathway %in% gsea_res_ind_select$pathway) %>%
    distinct(position, pathway, .keep_all=T) %>%
     mutate(pathway = sub("^GO", "", pathway ),
             pathway = gsub("_", " ", pathway)) %>% 
    mutate(pathway = case_when(grepl("NONSENSE MEDIATED DECAY",pathway) ~ "NONSENSE MEDIATED DECAY",
                               grepl("SERINE THREONINE KINASE", pathway) ~ "SERINE THREONINE KINASE SIGNALING",
                               TRUE ~ pathway))
  #modules_to_plot = unique(res2_cur_filt$module) 
  
  go_res_mat = gsea_res_ind_filt %>% select(position, pathway, NES) %>%
    pivot_wider(names_from = "position", values_from = "NES", values_fill = list(NES= 0)) %>%
    column_to_rownames("pathway") %>% as.matrix()
  
  hc_row = hclust(dist(go_res_mat), method="complete")
  hc_row_order = hc_row$labels[hc_row$order]
  
  
  gsea_res_ind_filt = gsea_res_ind_filt %>% 
    mutate(leadingEdge_len = map_int(leadingEdge, length))
  
  gsea_res_ind_filt = gsea_res_ind_filt %>%
    mutate(pathway = factor(pathway, levels=hc_row_order)) %>%
    mutate(position = factor(position, levels=c("ra", "arco", "hvc", "ncl", "lman", "nido", "x", "stri"))) %>%
    complete(pathway, position, fill=list(NES=NA, size=NA, leadingEdge_len = NA))

  size_limits = seq(round(min(gsea_res_ind_filt$size, na.rm=T), -1), round(max(gsea_res_ind_filt$size, na.rm=T), -1), by=50)
    size_limits = seq(round(min(gsea_res_ind_filt$leadingEdge_len, na.rm=T), -1), round(max(gsea_res_ind_filt$leadingEdge_len, na.rm=T), -1), by=20)
  #size_limits = seq(-50, 250, 50)
  #print(size_limits)
  gg = ggplot(gsea_res_ind_filt, aes(position, pathway)) + 
    geom_point(aes(size=leadingEdge_len, color=NES)) + 
    #scale_color_viridis() + 
    scale_color_gradient2(low=muted("blue"), high=muted("red")) +
    scale_size_continuous(range=c(.5, 3), breaks=size_limits) + 
    labs(x = "", y ="") + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle=90, vjust=.5, hjust=1, size=6),
          axis.text.y = element_text(hjust=1, size=6),
          axis.ticks = element_line(size=.25),
          panel.border = element_blank()) + 
      coord_fixed()
  print(gg)
  ncols = ncol(go_res_mat)
  nrows = nrow(go_res_mat)
  
  width = ncols / 10
  height = nrows / 10
  gg = set_panel_size(gg, width = unit(width, "in"), height = unit(height, "in"))
  save_plot(file.path(gsea_sub_dir, sprintf("go_terms_by_module_%s.pdf", paste(coef_to_use, collapse="-"))), gg, base_height = (nrows / 2) + 4, base_width = (ncols / 2) + 4)
})
```

###### Plot leading edges
```{r}
pathways_to_use = c("GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY", "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE")
coef_to_use = c("kl_mean_3")
p = "ra"
positions_to_use = c("ra", "hvc")
walk(positions_to_use, function(p) {
  print(p)
  walk(pathways_to_use, function(pathway_to_use) {
    print(pathway_to_use)
    gsea_res_ind1 = gsea_res_ind %>% 
      filter(pathway==pathway_to_use) %>%
      filter(coef==coef_to_use) %>%
      filter(position==p) 
    genes = unique(gsea_res_ind1$leadingEdge[[1]])
    
    res_gsea = res %>% 
      
      filter(coef==coef_to_use)  %>%
      filter(position==p) %>%
      filter(gene_id %in% genes) %>%
      filter(logFC>0) %>% 
      arrange(logFC) %>%
      distinct(gene_id, .keep_all=T) %>% 
      mutate(gene_id = factor(gene_id, levels=gene_id))
    
    gg = ggplot(res_gsea, aes(gene_id, logFC)) + 
      geom_point() + 
      coord_flip() + 
      theme(axis.text=element_text(size=5)
      )
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("go_leadingEdge_%s_%s_%s.pdf", p, coef_to_use, pathway_to_use)), gg, base_height=4, base_width=1)
  })
})
```

Plot leading edge split by position
```{r}

p = "ra"
pathway_to_use = "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY"
coef_to_use = c("kl_mean_3")

gsea_res_ind1 = gsea_res_ind %>% 
  filter(pathway==pathway_to_use) %>%
  filter(coef==coef_to_use) %>%
  filter(position==p) 
genes = unique(gsea_res_ind1$leadingEdge[[1]])[19:24]
genes = c("JUN", "KLF15", "IRF2BP2", "PBX1", "LHX2", "NOTCH1", "PLAGL2")

df = FetchData(se1, c(genes), slot="data") %>%
  as.data.frame() %>%
  rownames_to_column(var="id") %>%
  pivot_longer(cols=-id, names_to="gene_id", values_to="value") %>%
  mutate(id = as.numeric(id)) %>% 
  left_join(info2 %>% select(id, position, procedure2)) %>%
  mutate(position = factor(position, levels=position_levels),
         gene_id = factor(gene_id, levels=genes)) #%>%

df_id = df %>% distinct(id, position) %>%
  arrange(position) 
id_levels = df_id$id

df = df %>%
#  filter(position == p) %>% 
  mutate(id = factor(id, levels=id_levels))
df_sum = df  %>%
  #filter(gene_id %in% genes) %>% 
  group_by(gene_id, position) %>%
  summarize(value_mean = mean(value))
gg = ggplot(df, aes(position, value, color=position)) + 
  geom_point(size = .5, alpha=.1, shape=20, position=position_jitter(width=.1)) + 
  geom_point(data=df_sum, color="black", size=.5, aes(position, value_mean)) + 
  scale_color_manual(values=position_colors) + 
  facet_wrap(~gene_id) + 
  labs(x = "", y = "log expression") + 
  theme_bw() + 
  theme(
    legend.position="none",
    #strip.text = element_blank(),
    strip.background = element_blank(),
    axis.title.y = element_text(size=6),
    axis.title.x = element_blank(),
    axis.text.y = element_text(size=5),
    axis.text.x = element_blank(),
    axis.line = element_blank(),
    axis.ticks.y = element_line(size=.25),
    axis.ticks.x = element_blank(),
    panel.grid=element_blank())
gg
#save_plot(file.path(out_dir, "points_by_position_crh.pdf"), gg, base_height=.75, base_width=1, ncol=length(to_plot_manual), nrow=1)
```

Plot leading edge across kl_mean, breakout
```{r}
# 
# p = "x"
# pathway_to_use = "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY"
# pathways_to_use = c(
# #  "GO_GLIAL_CELL_DIFFENTIATION",
#   "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY",
#   "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE",
#   #"GO_RESPIRASOME",
#   #"GO_STRUCTURAL_CONSTITUENT_OF_RIBOSOME"
#   )
# coef_to_use = c("kl_mean_3")
# 
# nrows = 5
# walk(positions, function(p) {
#   walk(pathways_to_use, function(pathway_to_use) {
#     print(pathway_to_use)
#     gsea_res_ind1 = gsea_res %>% 
#       filter(pathway==pathway_to_use) %>%
#       filter(coef==coef_to_use) %>%
#       filter(position==p) 
#     genes = unique(gsea_res_ind1$leadingEdge[[1]])[1:20]
#     #genes = c("JUN", "KLF15", "IRF2BP2", "PBX1", "LHX2", "NOTCH1", "PLAGL2")
#     
#     line_color = position_colors[p]
#     df = FetchData(se1, c(genes), slot="data") %>%
#       as.data.frame() %>%
#       rownames_to_column(var="id") %>%
#       pivot_longer(cols=-id, names_to="gene_id", values_to="value") %>%
#       mutate(id = as.numeric(id)) %>% 
#       left_join(info2 %>% select(id, position, procedure2)) %>%
#       mutate(position = factor(position, levels=position_levels),
#              gene_id = factor(gene_id, levels=genes)) #%>%
#     
#     df_id = df %>% 
#       distinct(id, position) %>%
#       arrange(position) 
#     id_levels = df_id$id
#     
#     df = df %>%
#       filter(position == p) %>% 
#       #mutate(id = factor(id, levels=id_levels)) %>%
#       left_join(md)
#     df_sum = df  %>%
#       group_by(gene_id, kl_mean_log_scale_cut2_proc2) %>%
#       summarize(value_mean = mean(value))
#     gg = ggplot(df, aes(kl_mean_log_scale_cut2_proc2, value)) + 
#       geom_point(size = .5, alpha=.5, shape=20, position=position_jitter(width=.1)) + 
#       geom_point(data=df_sum, color=line_color, size=.5, aes(kl_mean_log_scale_cut2_proc2, value_mean)) + 
#       geom_line(data=df_sum, color=line_color, aes(kl_mean_log_scale_cut2_proc, value_mean,  group=gene_id)) +
#       #scale_color_manual(values=position_colors) + 
#       facet_wrap(~gene_id, scales="free", nrow=nrows) + 
#       labs(x = "", y = "log expression") + 
#       theme_bw() + 
#       theme(
#         legend.position="none",
#         strip.text = element_text(size=6, face="italic", hjust=0, vjust=1),
#         strip.background = element_blank(),
#         axis.title.y = element_text(size=6),
#         axis.title.x = element_blank(),
#         axis.text.y = element_text(size=5),
#         axis.text.x = element_blank(),
#         axis.line = element_blank(),
#         axis.ticks.y = element_line(size=.25),
#         axis.ticks.x = element_blank(),
#         #panel.grid=element_blank()
#       )
#     print(gg)
#     save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2_%s_%s.pdf", p, pathway_to_use)), gg, base_height=.75, base_width=.75, ncol=length(genes)/nrows, nrow=nrows)
#   })
# })

```

Plot leading edge across kl_mean, grouped
```{r}

p = "x"
pathway_to_use = "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY"
pathways_to_use = c(
  "GO_GLIAL_CELL_DIFFERENTIATION",
  "GO_NEURON_SPINE",
  "GO_PRIMARY_ACTIVE_TRANSMEMBRANE_TRANSPORTER_ACTIVITY",
  "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY",
  "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE"#,
 # "GO_RESPIRASOME",
#  "GO_STRUCTURAL_CONSTITUENT_OF_RIBOSOME"
  )
coef_to_use = c("kl_mean_3")

nrows = 5
positions = as.character(positions)
walk(positions, function(p) {
  walk(pathways_to_use, function(pathway_to_use) {
    print(pathway_to_use)
    gsea_res_ind1 = gsea_res %>% 
      filter(pathway==pathway_to_use) %>%
      filter(coef==coef_to_use) %>%
      filter(position==p) 
    genes = unique(gsea_res_ind1$leadingEdge[[1]])[1:20]
    
    line_color = position_colors[p]
    df = FetchData(se1, c(genes), slot="data") %>%
      as.data.frame() %>%
      rownames_to_column(var="id") %>%
      pivot_longer(cols=-id, names_to="gene_id", values_to="value") %>%
      mutate(id = as.numeric(id)) %>% 
      left_join(md %>% select(id, position, procedure2)) %>%
      mutate(position = factor(position, levels=position_levels),
             gene_id = factor(gene_id, levels=genes)) #%>%
    
    df_id = df %>% 
      distinct(id, position) %>%
      arrange(position) 
    id_levels = df_id$id
    
    df = df %>%
      filter(position == p) %>% 
      left_join(md %>% select(id, position, kl_mean_log_scale_cut2_proc2))
    
    df = df %>%
      group_by(gene_id) %>% 
      mutate(
        baseline = kl_mean_log_scale_cut2_proc2=="intact-1",
        value_rel = value - mean(value[baseline]),
        value_z = (value - mean(value[baseline])) / sd(value[baseline]))
    
    df_sum = df %>% 
      group_by(gene_id, kl_mean_log_scale_cut2_proc2) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean))
    
    df_group_sum = df  %>%
      group_by(kl_mean_log_scale_cut2_proc2) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean)) %>% 
      ungroup()
    
    gg = ggplot() + 
      #geom_point(size = .5, alpha=.5, shape=20, position=position_jitter(width=.1)) + 
      #geom_point(data=df_sum, color=line_color, size=.5, aes(kl_mean_log_scale_cut2_proc, value_mean)) + 
            geom_hline(yintercept=0, linetype=1, size=.25) + 
      geom_line(data=df_sum, color="grey", alpha=.5, aes(kl_mean_log_scale_cut2_proc2, value_z_mean,  group=gene_id), size=.25) +
      geom_line(data=df_group_sum, color=line_color, aes(kl_mean_log_scale_cut2_proc2, value_z_mean, group=1)) +

      #scale_color_manual(values=position_colors) + 
      #facet_wrap(~gene_id, scales="free", nrow=nrows) + 
      labs(x = "", y = "expression\nz-scored rel. low Song DKL") + 
      theme_bw() + 
      theme(
        legend.position="none",
        strip.text = element_text(size=6, face="italic", hjust=0, vjust=1),
        strip.background = element_blank(),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.text.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid=element_blank()
      )
    print(gg)
    width = .5
    height = .6
    gg = set_panel_size(gg, width = unit(width, "in"), height = unit(height, "in"))
    save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2_%s_%s_grouped.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
    
    gg = gg + geom_text_repel(data=df_sum %>% filter(kl_mean_log_scale_cut2_proc2=="deaf-3"), aes(x = 3, value_z_mean, label=gene_id), size=4)
    
        save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2_%s_%s_grouped_with-labels.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
  })
})

```

Plot leading edge across kl_mean, grouped, just intact-1 and deaf-3
```{r}

p = "x"
pathway_to_use = "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY"
pathways_to_use = c(
  "GO_GLIAL_CELL_DIFFERENTIATION",
  "GO_NEURON_SPINE",
  "GO_HORMONE_ACTIVITY",
  "GO_POSITIVE_REGULATION_OF_NEURON_DEATH",
  "GO_PRIMARY_ACTIVE_TRANSMEMBRANE_TRANSPORTER_ACTIVITY",
  "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY",
  "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE"#,
 # "GO_RESPIRASOME",
#  "GO_STRUCTURAL_CONSTITUENT_OF_RIBOSOME"
  )
coef_to_use = c("kl_mean_3")

nrows = 5
positions = as.character(positions)
walk(positions, function(p) {
  walk(pathways_to_use, function(pathway_to_use) {
    print(pathway_to_use)
    gsea_res_ind1 = gsea_res %>% 
      filter(pathway==pathway_to_use) %>%
      filter(coef==coef_to_use) %>%
      filter(position==p) 
    genes = unique(gsea_res_ind1$leadingEdge[[1]])[1:20]
    
    line_color = position_colors[p]
    df = FetchData(se1, c(genes), slot="data") %>%
      as.data.frame() %>%
      rownames_to_column(var="id") %>%
      pivot_longer(cols=-id, names_to="gene_id", values_to="value") %>%
      mutate(id = as.numeric(id)) %>% 
      left_join(md %>% select(id, position, procedure2)) %>%
      mutate(position = factor(position, levels=position_levels),
             gene_id = factor(gene_id, levels=genes))
    

    
    df_id = df %>% 
      distinct(id, position) %>%
      arrange(position) 
    id_levels = df_id$id
    
    df = df %>%
      filter(position == p) %>% 
      left_join(md %>% select(id, position, kl_mean_log_scale_cut2_proc2))
    
    df = df %>% filter(kl_mean_log_scale_cut2_proc2 %in% c("intact-1", "deaf-3"))
    
    df = df %>%
      group_by(gene_id) %>% 
      mutate(
        baseline = kl_mean_log_scale_cut2_proc2=="intact-1",
        value_rel = value - mean(value[baseline]),
        value_z = (value - mean(value[baseline])) / sd(value[baseline]))
    
    df_sum = df %>% 
      group_by(gene_id, kl_mean_log_scale_cut2_proc2) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean))
    
    df_group_sum = df  %>%
      group_by(kl_mean_log_scale_cut2_proc2) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean)) %>% 
      ungroup()
    
    gg = ggplot() + 
      #geom_point(size = .5, alpha=.5, shape=20, position=position_jitter(width=.1)) + 
      #geom_point(data=df_sum, color=line_color, size=.5, aes(kl_mean_log_scale_cut2_proc, value_mean)) + 
            geom_hline(yintercept=0, linetype=1, size=.25) + 
      geom_line(data=df_sum, color="grey", alpha=.5, aes(kl_mean_log_scale_cut2_proc2, value_z_mean,  group=gene_id), size=.25) +
      geom_line(data=df_group_sum, color=line_color, aes(kl_mean_log_scale_cut2_proc2, value_z_mean, group=1)) +

      #scale_color_manual(values=position_colors) + 
      #facet_wrap(~gene_id, scales="free", nrow=nrows) + 
      labs(x = "", y = "expression\nz-scored rel. low Song DKL") + 
      theme_bw() + 
      theme(
        legend.position="none",
        strip.text = element_text(size=6, face="italic", hjust=0, vjust=1),
        strip.background = element_blank(),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.text.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid=element_blank()
      )
    print(gg)
    width = .5
    height = .6
    gg = set_panel_size(gg, width = unit(width, "in"), height = unit(height, "in"))
    save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2-low-hi_%s_%s_grouped.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
    
    gg = gg + geom_text_repel(data=df_sum %>% filter(kl_mean_log_scale_cut2_proc2=="deaf-3"), aes(x = 3, value_z_mean, label=gene_id), size=4)
    
        save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2-low-hi_%s_%s_grouped_with-labels.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
  })
})

```

Plot leading edge across kl_mean, logFC, grouped, just intact-1 and deaf-3
```{r}

p = "x"
pathway_to_use = "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY"
pathways_to_use = c(
  "GO_GLIAL_CELL_DIFFERENTIATION",
  "GO_NEURON_SPINE",
  "GO_HORMONE_ACTIVITY",
  "GO_POSITIVE_REGULATION_OF_NEURON_DEATH",
  "GO_PRIMARY_ACTIVE_TRANSMEMBRANE_TRANSPORTER_ACTIVITY",
  "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY",
  "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE"#,
 # "GO_RESPIRASOME",
#  "GO_STRUCTURAL_CONSTITUENT_OF_RIBOSOME"
  )
coef_to_use = c("kl_mean_3")

nrows = 5
positions = as.character(positions)
walk(positions, function(p) {
  walk(pathways_to_use, function(pathway_to_use) {
    print(pathway_to_use)
    gsea_res_ind1 = gsea_res %>% 
      filter(pathway==pathway_to_use) %>%
      filter(coef==coef_to_use) %>%
      filter(position==p) 
    genes = unique(gsea_res_ind1$leadingEdge[[1]])[1:20]
    
    line_color = position_colors[p]
    
    res_select = res %>% 
      filter(coef == coef_to_use)  %>%
      filter(position == p) %>%
      filter(gene_id %in% genes) %>%
      mutate(coef = factor(coef, levels=c("baseline", coef_to_use))) %>%
      complete(coef, gene_id, fill=list(logFC=0, std_error=0))
      
    # df = FetchData(se1, c(genes), slot="data") %>%
    #   as.data.frame() %>%
    #   rownames_to_column(var="id") %>%
    #   pivot_longer(cols=-id, names_to="gene_id", values_to="value") %>%
    #   mutate(id = as.numeric(id)) %>% 
    #   left_join(md %>% select(id, position, procedure2)) %>%
    #   mutate(position = factor(position, levels=position_levels),
    #          gene_id = factor(gene_id, levels=genes))
    # 
    # 
    # 
    # df_id = df %>% 
    #   distinct(id, position) %>%
    #   arrange(position) 
    # id_levels = df_id$id
    # 
    # df = df %>%
    #   filter(position == p) %>% 
    #   left_join(md %>% select(id, position, kl_mean_log_scale_cut2_proc2))
    # 
    # df = df %>% filter(kl_mean_log_scale_cut2_proc2 %in% c("intact-1", "deaf-3"))
    # 
    # df = df %>%
    #   group_by(gene_id) %>% 
    #   mutate(
    #     baseline = kl_mean_log_scale_cut2_proc2=="intact-1",
    #     value_rel = value - mean(value[baseline]),
    #     value_z = (value - mean(value[baseline])) / sd(value[baseline]))
    # 
    # df_sum = df %>% 
    #   group_by(gene_id, kl_mean_log_scale_cut2_proc2) %>%
    #   summarize_at(vars(value_rel, value_z), list(mean = mean))
    # 
    df_group_sum = res_select  %>%
      group_by(coef) %>%
      summarize_at(vars(logFC,std_error), list(mean = mean, sem = sem, sd = sd)) %>%
      ungroup()
    
    gg = ggplot() + 
      #geom_point(size = .5, alpha=.5, shape=20, position=position_jitter(width=.1)) + 
      #geom_point(data=df_sum, color=line_color, size=.5, aes(kl_mean_log_scale_cut2_proc, value_mean)) + 
            geom_hline(yintercept=0, linetype=1, size=.25) + 
      geom_line(data=res_select, color="grey", alpha=.5, aes(coef, logFC,  group=gene_id), size=.25) +
     # geom_pointrange(data=df_group_sum, color=line_color, aes(coef, logFC_mean, ymin = (logFC_mean-logFC_sd), ymax=(logFC_mean+logFC_sd)), size=.2)  +
      geom_line(data=df_group_sum, color=line_color, aes(coef, logFC_mean, group=1)) +

      #scale_color_manual(values=position_colors) + 
      #facet_wrap(~gene_id, scales="free", nrow=nrows) + 
      labs(x = "", y = "logFC") + 
      theme_bw() + 
      theme(
        legend.position="none",
        strip.text = element_text(size=6, face="italic", hjust=0, vjust=1),
        strip.background = element_blank(),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.text.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid=element_blank()
      )
    print(gg)
    width = .5
    height = .6
    gg = set_panel_size(gg, width = unit(width, "in"), height = unit(height, "in"))
    save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2-low-hi_logFC_%s_%s_grouped.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
    # 
    # gg = gg + geom_text_repel(data=df_sum %>% filter(kl_mean_log_scale_cut2_proc2=="deaf-3"), aes(x = 3, value_z_mean, label=gene_id), size=4)
    # 
    #     save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2-low-hi_%s_%s_grouped_with-labels.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
  })
})

```


Plot leading edge across kl_mean, continuous, just intact-1 and deaf-3
```{r}

p = "x"
pathway_to_use = "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY"
pathways_to_use = c(
  "GO_GLIAL_CELL_DIFFERENTIATION",
  "GO_NEURON_SPINE",
  "GO_HORMONE_ACTIVITY",
  "GO_POSITIVE_REGULATION_OF_NEURON_DEATH",
  "GO_PRIMARY_ACTIVE_TRANSMEMBRANE_TRANSPORTER_ACTIVITY",
  "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY",
  "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE"#,
 # "GO_RESPIRASOME",
#  "GO_STRUCTURAL_CONSTITUENT_OF_RIBOSOME"
  )
coef_to_use = c("kl_mean_3")

nrows = 5
positions = as.character(positions)
walk(positions, function(p) {
  walk(pathways_to_use, function(pathway_to_use) {
    print(pathway_to_use)
    gsea_res_ind1 = gsea_res %>% 
      filter(pathway==pathway_to_use) %>%
      filter(coef==coef_to_use) %>%
      filter(position==p) 
    genes = unique(gsea_res_ind1$leadingEdge[[1]])[1:20]
    
    line_color = position_colors[p]
    df = FetchData(se1, c(genes), slot="data") %>%
      as.data.frame() %>%
      rownames_to_column(var="id") %>%
      pivot_longer(cols=-id, names_to="gene_id", values_to="value") %>%
      mutate(id = as.numeric(id)) %>% 
      left_join(md %>% select(id, position, procedure2, kl_mean_log_scale)) %>%
      mutate(position = factor(position, levels=position_levels),
             gene_id = factor(gene_id, levels=genes))
    

    
    df_id = df %>% 
      distinct(id, position) %>%
      arrange(position) 
    id_levels = df_id$id
    
    df = df %>%
      filter(position == p) %>% 
      left_join(md %>% select(id, position, kl_mean_log_scale_cut2_proc2, kl_mean_log_scale, tags))
    
    df = df %>% filter(kl_mean_log_scale_cut2_proc2 %in% c("intact-1", "deaf-3"))
    
    df = df %>%
      group_by(gene_id) %>% 
      mutate(
        baseline = kl_mean_log_scale_cut2_proc2=="intact-1",
        value_rel = value - mean(value[baseline]),
        value_z = (value - mean(value[baseline])) / sd(value[baseline]))
    
    df_sum = df %>% 
      group_by(gene_id, tags, kl_mean_log_scale) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean, sem = sem))
    
    df_group_sum = df  %>%
      group_by(kl_mean_log_scale_cut2_proc2) %>%
      summarize_at(vars(value_rel, value_z), list(mean = mean)) %>% 
      ungroup()
    
    gg = ggplot() + 
      #geom_point(size = .5, alpha=.5, shape=20, position=position_jitter(width=.1)) + 
      geom_line(data=df_sum, color=line_color, size=.5, aes(kl_mean_log_scale, value_z_mean, group=gene_id, size=.25)) + 
            geom_hline(yintercept=0, linetype=1, size=.25) + 
      #geom_line(data=df_sum, color="grey", alpha=.5, aes(kl_mean_log_scale_cut2_proc2, value_z_mean,  group=gene_id), size=.25) #+
      #geom_line(data=df_group_sum, color=line_color, aes(kl_mean_log_scale_cut2_proc2, value_z_mean, group=1)) +

      #scale_color_manual(values=position_colors) + 
      #facet_wrap(~gene_id, scales="free", nrow=nrows) + 
      labs(x = "", y = "expression\nz-scored rel. low Song DKL") + 
      theme_bw() + 
      theme(
        legend.position="none",
        strip.text = element_text(size=6, face="italic", hjust=0, vjust=1),
        strip.background = element_blank(),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.text.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid=element_blank()
      )
    print(gg)
    width = .5
    height = .6
    gg = set_panel_size(gg, width = unit(width, "in"), height = unit(height, "in"))
    save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2-low-hi_%s_%s_grouped.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
    
    gg = gg + geom_text_repel(data=df_sum %>% filter(kl_mean_log_scale_cut2_proc2=="deaf-3"), aes(x = 3, value_z_mean, label=gene_id), size=4)
    
        save_plot(file.path(gsea_sub_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2-low-hi_%s_%s_grouped_with-labels.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
  })
})

```

###### Plot leading edges, heatmap across positions
```{r}
co = "kl_mean_3"

pathways_to_use = c(
  "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY",
  "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE",
  "GO_RESPIRASOME",
  "GO_STRUCTURAL_CONSTITUENT_OF_RIBOSOME"
  )
coef_to_use = c("kl_mean_3")

nrows = 5

positions_to_use = c("ra", "hvc", "lman", "x")
walk(pathways_to_use, function(pathway_to_use) {
  print(pathway_to_use)
  gsea_res_ind1 = gsea_res %>% 
    filter(pathway==pathway_to_use) %>%
    filter(coef==coef_to_use) 
  
  genes = map(positions_to_use, function(p) {
    gsea_res_ind2 = gsea_res_ind1 %>% filter(position==p)
    unique(gsea_res_ind2$leadingEdge[[1]])[1:20]
  }) %>% unlist() %>% unique()
  
  res_sig = res %>%
    filter(coef==co) %>% 
    filter(position %in% positions_to_use) %>%
    filter(gene_id %in% genes)
  

  value_var = "logFC"
  res_cur = res %>% 
    filter(gene_id %in% res_sig$gene_id) %>% 
    filter(coef==co) %>%
    group_by(position, gene_id) %>%
    summarize(logFC = mean(logFC),
              logFC_rescale = mean(logFC_rescale))
  mat = acast(res_cur, gene_id~position, value.var=value_var)
  mat = mat[,c("ra", "hvc", "lman", "x")]
  colnames(mat) = toupper(colnames(mat))
  
  mat_var = apply(mat, 1, var)
  #mat = mat[mat_var>quantile(mat_var, .25),]
  width = .25
  height = width * nrow(mat) / ncol(mat)
  
  max_val = max(c(abs(min(mat)), max(mat)))
  max_val = 1.5
  #length_out = 9
 # cols_cur = colorRamp2(breaks=seq(-1*max_val, max_val, length.out=length_out), colors = rev(brewer_pal(palette=cols[[co]])(length_out)))
  cols_cur = colorRamp2(breaks=c(-1*max_val, 0, max_val), colors = c(muted("purple"), "white", muted("orange")))
  hm = Heatmap(mat, col=cols_cur,
               clustering_method_rows = "ward.D", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=5), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=5)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(gsea_sub_dir, sprintf("heatmap_%s_song-only.pdf", pathway_to_use)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  #hm
  
  hm_t = Heatmap(t(mat), col=cols_cur,
               clustering_method_columns  = "average", 
               cluster_rows = F,
               height = unit(width, "in"),
               width =unit( height, "in"),
               column_names_gp = gpar(fontsize=5), column_names_side = "top",
               row_names_gp = gpar(fontsize=5)
               #  clustering_method_columns = "ward.D"
  )
  
    pdf(file.path(gsea_sub_dir, sprintf("heatmap_%s_song-only_horiz.pdf", pathway_to_use)), width=height+2, height=width+2)
  print(hm_t)
  dev.off()
  #hm
})
```

###### Plot top GO terms, points
```{r}

walk(positions, function(p) {
  print(p)
  walk(coefs, function(co) {
    print(co)
    co_label = sub("position%s.", "", co)
    gsea_res_ind_filt = gsea_res_ind %>% 
      filter(position==p) %>%
      filter(coef==co) %>%
      filter(padj<.2)
    
    gsea_res_ind_top = gsea_res_ind %>%
      filter(position==p) %>%
      filter(padj<.2) %>%
      filter(coef==co) %>% 
      # group_by(position) %>% 
      mutate(NES_sign = sign(NES)) %>% 
      group_by(NES_sign) %>% 
      top_n(10, abs(NES)) %>%
      ungroup() %>%
      distinct(pathway)
    
    gsea_res_filt = gsea_res %>% 
      filter(position==p) %>%
      filter(coef==co) %>% 
      filter(pathway %in% gsea_res_ind_top$pathway)# %>%
    #  filter(padj<.2)
    
    if (nrow(gsea_res_filt) < 2)
      return(NULL)
    tmp = gsea_res_filt %>% 
      
      mutate(class = case_when(grepl("HORMONE|NEUROTR|PEPTI", pathway) ~ "modulation",
                               grepl("OXID|NADH|MITO|RESPIR|PROTON|REDOX", pathway) ~ "respiration",
                               grepl("RIBO|PROTEIN_LOCALIZ|TRANSLATION|RRNA", pathway) ~ "translation",
                               TRUE ~ "none"),
      ) %>%
      mutate(pathway = sub("GO_", "", pathway),
             pathway = case_when(grepl("NONSENSE", pathway) ~ "NONSENSE_MEDIATED_DECAY",
                                 #  grepl("ESTABLISHMENT_OF_PROTEIN", pathway) ~ "PROTEIN_LOCALIZATION_TO_ER",
                                 TRUE ~ pathway),
             pathway = gsub("_", " ", pathway),
             pathway = gsub("\\.", " ", pathway), 
             pathway = str_to_title(pathway),
      )  %>%
      arrange(NES) %>%
      mutate(pathway = factor(pathway, levels=pathway)) 
    
    #cols =c("black",  pal_aaas()(3))
    #names(cols) = c("none", "respiration",  "modulation", "translation")
    gg = ggplot(tmp, aes(NES, pathway)) + 
      
      geom_segment(aes(x=min(NES),xend = NES, y=pathway, yend=pathway), size=.25, linetype=3) + 
      geom_segment(aes(x = 0, xend = NES, y=pathway, yend=pathway), size=.25) +
      geom_point() + 
      theme_light() + 
      theme(axis.text = element_text(size=5),
            axis.title = element_text(size=6),
            axis.line.x = element_line(size=.25),
            axis.line.y = element_blank(),
            axis.ticks.x = element_line(size=.25),
            axis.ticks.y = element_blank(),
            legend.position = "bottom",
            legend.direction="horizontal",
            legend.text = element_text(size=5)) + 
      labs(y="") + 
      scale_color_manual(values=cols, name="")
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("point_%s_%s.pdf", co_label, p)), gg, base_height=2.5, base_width=5)
  })
})
```

###### MSIGDBR

```{r}
library(msigdbr)
msigdbr_df = msigdbr() %>%
  filter(gs_subcat %in% c("CP:BIOCARTA", "GO:MF", "GO:BP", "GO:CC", "CP:KEGG", "CP", "CP:REACTOME", "CP:WIKIPATHWAYS", "TFT:GTRD"))
msigdbr_lists = split(msigdbr_df$gene_symbol, msigdbr_df$gs_name)
msigdbr_lists = c(msigdbr_lists, gray_gmt)
```

```{r}
# genes_avail = bitr(unique(res$gene_id), fromType = "SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db") %>%
#   mutate(gene_id = SYMBOL, entrezid = ENTREZID)
# res_cur = res %>% left_join(genes_avail)
```

```{r message=FALSE, warning=FALSE}

gsea_res_fname = file.path(gsea_sub_dir, "gsea_msigdbr_res.qs")
gsea_res_ind_fname = file.path(gsea_sub_dir, "gsea_msigdbr_res_ind.qs")

redo = F
if (redo) {
  
gsea_res = future_map(positions, function(p) {
  print(p)
  res_cur = res %>% filter(position == p)
  map(coefs, function(co) {
    res_cur1 = res_cur %>% filter(coef==co)
    rnks = res_cur1$t
    names(rnks) = res_cur1$gene_id
    rnks = sort(rnks)
    gsea_res = fgsea(pathways = msigdbr_lists,
                     stats = rnks,
                     minSize=20,
                     maxSize=200,
                     nperm=1000)
    
    return(gsea_res)
 
  }) %>% set_names(coefs) %>% bind_rows(.id="coef")
})  %>% set_names(positions) %>% bind_rows(.id="position")

gsea_res_ind = future_map(positions, function(p) {
  print(p)
  res_cur = res %>% filter(position == p)
  gsea_res_cur = gsea_res %>% filter(position==p)
  map(coefs, function(co) {
    res_cur1 = res_cur  %>% filter(coef==co)
    gsea_res_cur1 = gsea_res_cur %>% filter(coef==co)
    rnks = res_cur1$t
    names(rnks) = res_cur1$gene_id
    rnks = sort(rnks)
    gsea_res_filt = gsea_res_cur1 %>% filter(padj<.2) %>% arrange(pval)
    collapsedPathways = collapsePathways(gsea_res_filt, pathways = msigdbr_lists, stats = rnks, pval.threshold = .001)
    gsea_res_ind = gsea_res_cur1 %>% filter(pathway %in% collapsedPathways$mainPathways) %>%
      arrange(-NES)
    gsea_res_ind
  }) %>% set_names(coefs) %>% bind_rows(.id="coef")
}) %>% bind_rows()

  qsave(gsea_res, gsea_res_fname)
  qsave(gsea_res_ind, gsea_res_ind_fname)
} else {
  gsea_res = qread(gsea_res_fname)
  gsea_res_ind = qread(gsea_res_ind_fname)
}
```

```{r}
padj_thresh = 0.2
positions = unique(gsea_res_ind$position)
coefs = unique(gsea_res_ind$coef)
walk(positions, function(p) {
  walk(coefs, function(co) {
    co_label = sub("position%s\\.", "", co)
    gsea_res_filt = gsea_res_ind %>% filter(padj<padj_thresh) %>% 
      filter(position==p, coef==co) %>%
      mutate(padj_log = -1 * log10(padj),
             p_log = -1 * log10(pval)) %>%
    #  filter(NES>0) %>% 
      mutate(size1 = size) %>%
      # mutate(to_label = ifelse(pathway %in% gsea_res_ind_filt$pathway, pathway, NA)) %>%
      mutate(leadingEdge_len = map_int(leadingEdge, length)) %>%
      mutate(leading_frac = leadingEdge_len / size) %>% 
      mutate(pathway1 = sub("^GO", "", pathway ),
             pathway1 = gsub("_", " ", pathway1),
             pathway1 = str_to_title(pathway1))
    
    if (nrow(gsea_res_filt) < 10) 
      return()
    gg = ggplot(gsea_res_filt, aes(NES, p_log, size=leading_frac)) + 
      geom_point() + 
      
      theme_classic() 
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_no_text.pdf", padj_thresh, p, co_label)), gg, base_width=6, base_height=6)
    
    gg = gg + theme(legend.position="none")
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_no_text_no_legend.pdf", padj_thresh, p, co_label)), gg, base_width=6, base_height=6)
    
    gg = gg +  
      geom_text(aes(NES, p_log, label = pathway1), inherit.aes = F) 
    save_plot(file.path(gsea_sub_dir, sprintf("gsea_gray_nes_padj_%s_%s_%s_text.pdf", padj_thresh,  p, co_label)), gg, base_width=6, base_height=6)
  })
})
```

```{r}
walk(coefs, function(co) {
co_label = sub("position%s.", "", co)
gsea_res_ind_filt = gsea_res_ind %>% 
  filter(padj<.2) %>%
  filter(coef==co) %>%
  group_by(position) %>%
  top_n(10, abs(NES))
  # gsea_res_ind_top = gsea_res_ind %>% filter(padj<.2) %>%
  #   top_n(10, NES)
##gsea_res_filt = gsea_res %>% filter(padj<.2)
gsea_res_filt = gsea_res %>% 
  #filter(padj<.2) %>%
  filter(pathway %in% gsea_res_ind_filt$pathway) %>%
  filter(position %in% c("ra", "hvc", "lman", "x"))

mat = gsea_res_filt %>% filter(coef==co) %>%
  select(position, pathway, NES) %>%
  pivot_wider(names_from = "position", values_from = "NES", values_fill=list(NES=0)) %>%
  column_to_rownames("pathway") %>% 
  as.matrix() 

mat = mat[,intersect(position_levels, colnames(mat))]

width = 1
height = 8
hm = Heatmap(mat, cluster_columns=F, width = unit(width, "in"),
             height = unit(height, "in"))
hm

pdf(file.path(gsea_sub_dir, sprintf("heatmap_%s.pdf", co_label)), width=width + 8, height=height + 2)
print(hm)
dev.off()
})
```

```{r}

walk(positions, function(p) {
  print(p)
  walk(coefs, function(co) {
    print(co)
    co_label = sub("position%s.", "", co)
    gsea_res_ind_filt = gsea_res_ind %>% 
      filter(position==p) %>%
      filter(coef==co) %>%
      filter(padj<.2)
    gsea_res_ind_top = gsea_res_ind %>%
      filter(position==p) %>%
      filter(padj<.2) %>%
      filter(coef==co) %>% 
      # group_by(position) %>% 
    # top_n(20, abs(NES)) %>%
      ungroup() %>%
      distinct(pathway)
    
    gsea_res_filt = gsea_res %>% 
      filter(position==p) %>%
      filter(coef==co) %>% 
      filter(pathway %in% gsea_res_ind_top$pathway)# %>%
    #  filter(padj<.2)
    
    if (nrow(gsea_res_filt) < 2)
      return(NULL)
    tmp = gsea_res_filt %>% 
      
      mutate(class = case_when(grepl("HORMONE|NEUROTR|PEPTI", pathway) ~ "modulation",
                               grepl("OXID|NADH|MITO|RESPIR|PROTON|REDOX", pathway) ~ "respiration",
                               grepl("RIBO|PROTEIN_LOCALIZ|TRANSLATION|RRNA", pathway) ~ "translation",
                               TRUE ~ "none"),
      ) %>%
      mutate(pathway = sub("GO_", "", pathway),
             pathway = case_when(grepl("NONSENSE", pathway) ~ "NONSENSE_MEDIATED_DECAY",
                                 #  grepl("ESTABLISHMENT_OF_PROTEIN", pathway) ~ "PROTEIN_LOCALIZATION_TO_ER",
                                 TRUE ~ pathway),
             pathway = gsub("_", " ", pathway),
             pathway = gsub("\\.", " ", pathway), 
             pathway = str_to_title(pathway),
      )  %>%
      arrange(NES) %>%
      mutate(pathway = factor(pathway, levels=pathway)) 
    
    cols =c("black",  pal_aaas()(3))
    names(cols) = c("none", "respiration",  "modulation", "translation")
    gg = ggplot(tmp, aes(NES, pathway)) + 
      
      geom_segment(aes(x=min(NES),xend = NES, y=pathway, yend=pathway), size=.25, linetype=3) + 
      geom_segment(aes(x = 0, xend = NES, y=pathway, yend=pathway), size=.25) +
      geom_point() + 
      theme_light() + 
      theme(axis.text = element_text(size=5),
            axis.title = element_text(size=6),
            axis.line.x = element_line(size=.25),
            axis.line.y = element_blank(),
            axis.ticks.x = element_line(size=.25),
            axis.ticks.y = element_blank(),
            legend.position = "bottom",
            legend.direction="horizontal",
            legend.text = element_text(size=5)) + 
      labs(y="") + 
      scale_color_manual(values=cols, name="")
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("point_%s_%s.pdf", co_label, p)), gg, base_height=2.5, base_width=3.5)
  })
})
```

Bar
```{r}

walk(positions, function(p) {
  print(p)
  walk(coefs, function(co) {
    print(co)
    co_label = sub("position%s.", "", co)
    gsea_res_ind_filt = gsea_res_ind %>% 
      filter(position==p) %>%
      filter(coef==co) %>%
      filter(padj<.2)
    gsea_res_ind_top = gsea_res_ind %>%
      filter(position==p) %>%
      filter(padj<.2) %>%
      filter(coef==co) %>% 
      # group_by(position) %>% 
      top_n(20, abs(NES)) %>%
      ungroup() %>%
      distinct(pathway)
    
    gsea_res_filt = gsea_res %>% 
      filter(position==p) %>%
      filter(coef==co) %>% 
      filter(pathway %in% gsea_res_ind_top$pathway)# %>%
    #  filter(padj<.2)
    
    if (nrow(gsea_res_filt) < 2)
      return(NULL)
    tmp = gsea_res_filt %>% 
      
      mutate(class = case_when(grepl("HORMONE|NEUROTR|PEPTI", pathway) ~ "modulation",
                               grepl("OXID|NADH|MITO|RESPIR|PROTON|REDOX", pathway) ~ "respiration",
                               grepl("RIBO|PROTEIN_LOCALIZ|TRANSLATION|RRNA", pathway) ~ "translation",
                               TRUE ~ "none"),
      ) %>%
      mutate(pathway = sub("GO_", "", pathway),
             pathway = case_when(grepl("NONSENSE", pathway) ~ "NONSENSE_MEDIATED_DECAY",
                                 #  grepl("ESTABLISHMENT_OF_PROTEIN", pathway) ~ "PROTEIN_LOCALIZATION_TO_ER",
                                 TRUE ~ pathway),
             pathway = gsub("_", " ", pathway),
             pathway = gsub("\\.", " ", pathway), 
             pathway = str_to_title(pathway),
      )  %>%
      distinct(pathway, .keep_all=T) %>% 
      arrange(NES) %>%
      mutate(pathway = factor(pathway, levels=pathway)) 
    
    cols =c("black",  pal_aaas()(3))
    names(cols) = c("none", "respiration",  "modulation", "translation")
    gg = ggplot(tmp, aes(NES, pathway)) + 
      geom_bar(stat="identity") + 
      #geom_segment(aes(x=min(NES),xend = NES, y=pathway, yend=pathway), size=.25, linetype=3) + 
      #geom_segment(aes(x = 0, xend = NES, y=pathway, yend=pathway), size=.25) +
      #geom_point() + 
      theme_light() + 
      theme(axis.text = element_text(size=5),
            axis.title = element_text(size=6),
            axis.line.x = element_line(size=.25),
            axis.line.y = element_blank(),
            axis.ticks.x = element_line(size=.25),
            axis.ticks.y = element_blank(),
            legend.position = "bottom",
            legend.direction="horizontal",
            legend.text = element_text(size=5)) + 
      labs(y="") + 
      scale_color_manual(values=cols, name="")
    gg
    save_plot(file.path(gsea_sub_dir, sprintf("bar_%s_%s.pdf", co_label, p)), gg, base_height=2.5, base_width=10)
  })
})
```
##### ARGs

```{r}
arg_dir = file.path(outdir2, "ARGs")
dir.create(arg_dir)
```

```{r}
library(fgsea)
gray_gmt_fname = "~/data/gene_lists/gray_2018_arg.xls.gmt"
gray_gmt = gmtPathways(gray_gmt_fname)
```

```{r}
res_args = map(gray_gmt, function(x) {
  res_cur = res %>% filter(gene_id %in% x)
  res_cur
})
res_args = c(res_args, list(res))
names(res_args) = c(names(gray_gmt), "all")
co = "position%s.num_songs_on_euth_date_log_scale"
p = "ra"

res_arg = res_args %>% bind_rows(.id="set")

res_arg_cur = res_arg %>%
  filter(coef==co) %>%
    filter(position==p) %>%
  filter(set!="ARGs")

set_colors = c(pal_locuszoom()(3), "grey50")
names(set_colors) = c("rapid PRGs", "delayed PRGs", "SRGs", "all")
gg = ggplot(res_arg_cur, aes(logFC)) + 
  geom_density(aes(color=set)) + 
  scale_color_manual(values=set_colors)
gg

```

```{r}
positions = unique(res_arg$position)
coefs = unique(res_arg$coef)

walk(positions, function(p) {
  walk(coefs, function(co) {
    co_label = sub("position%s\\.", "", co)
    walk(c("rapid PRGs", "delayed PRGs", "SRGs"), function(set_to_use) {
      res_arg_cur1 = res_arg %>%
          filter(coef==co) %>%
    filter(position==p) %>%
        filter(set %in% c(set_to_use, "all"))
      gg = ggplot(res_arg_cur1, aes(logFC)) + 
        geom_density(aes(fill=set), alpha=.5, size=.5) + 
        scale_fill_manual(values=set_colors, guide=F) + 
        theme_bw() + 
        theme(axis.text = element_text(size=5),
              axis.title = element_text(size=6),
              axis.line.x = element_blank(),
              axis.line.y = element_blank(),
              axis.ticks.x = element_line(size=.25),
              axis.ticks.y = element_line(size=.25))  + 
        labs(title=sprintf("%s - %s - %s", p, co_label, set_to_use))
    #  print(gg)
      save_plot(file.path(arg_dir, sprintf("logFC_density_%s_vs_all_%s_%s.pdf", p, co_label, make.names(set_to_use))), gg, base_height=2, base_asp=1)
    })
  })
})
```

##### Combine with cell type expression

```{r}
ct_dir = file.path(outdir2, "celltype_expression")
dir.create(ct_dir)
```

```{r}

scrna_dir = "~/hugin_home/data2/rstudio/birds/scRNA/devin_combined/songbird_cells/trees/celltypes_hclust_all_int_sub2_sct_regress_position2_zf_ortho"
scrna_fname = "average_expr.qs"
sc_avg = qread(file.path(scrna_dir, scrna_fname))
cluster_int_sub2_levels = c("HVC_Glut-1", "HVC_Glut-2", "HVC_Glut-3", "HVC_Glut-4", "HVC_Glut-5", "RA_Glut-1", "RA_Glut-2", "RA_Glut-3", 
                            "GABA-1-1",  "GABA-1-2", "GABA-2", "GABA-3", "GABA-4", "GABA-5-1", "GABA-5-2", "GABA-5-3", "GABA-6", "GABA-7", "GABA-8", "Pre-1", "Pre-2", "Pre-3", "Pre-4", "GABA-Pre",
                            "Astro", "Oligo", "OPC",
                            "Micro", "Endo", "Epen", "Mural", "VLMC")
sc_avg = sc_avg[,cluster_int_sub2_levels]
```

```{r}

dist_method = "cor"
hc_method = "ward.D"

# if (dist_method=="cor") {
#   mat_dist = as.dist(1-cor(t(coefs2_cor_un_mat)))
# } else {
#   mat_dist = dist(coefs2_cor_un_mat)
# }
# hc = hclust(mat_dist,method=hc_method)
# plot(hc)

# k = 8
# hc_cut = cutree(hc, k=k)
positions_to_use = c("ra", "hvc")
coefs_to_plot = c("kl_mean_3", "num_song_euth_date", "kl_mean_aov", "inter", "kl_mean_2" )
walk(positions_to_use, function(p) {
  print(p)
  walk(coefs_to_plot, function(co) {
    print(co)
    #p = "ra"
    #co = "position%s.kl_mean_log_scale.num_songs_on_euth_date_log_scale"
    co_label = sub("position%s\\.", "", co)
    res_sig = res %>%
      filter(position == p) %>%
      filter(coef==co) %>% 
      mutate(sign = sign(logFC)) %>% 
      group_by(sign) %>%
      filter(adj.P.Val<.1) %>%
      #  top_n(50, abs(logFC))
      top_n(-25, adj.P.Val)
    
    
    if (nrow(res_sig) < 3) {
      return()
    }
    value_var = case_when(co %in%  c("position%s.num_songs_on_euth_date_log_scale", "position%s.kl_mean_log_scale") ~ "logFC_rescale",
                          co == "position%s.kl_mean_log_scale.num_songs_on_euth_date_log_scale" ~ "logFC")
    value_var = "logFC"
    
    
    walk(c(1, -1), function(si) {
      res_sig_cur = res_sig %>% filter(sign==si)
      if(nrow(res_sig_cur) < 6) return()
      mat = acast(res %>% filter(gene_id %in% res_sig_cur$gene_id) %>% filter(coef==co), gene_id~position, value.var=value_var, fill = 0)
      colnames(mat) = toupper(colnames(mat))
      sc_avg_clust = as.matrix(sc_avg[intersect(rownames(sc_avg), rownames(mat)),])
      
      if (p == "ra") {
        clusters_to_keep = colnames(sc_avg_clust)[-grep("HVC|Pre", colnames(sc_avg_clust))]
      } else {
        clusters_to_keep = colnames(sc_avg_clust)[-grep("RA", colnames(sc_avg_clust))]
      }
      
      sc_avg_clust = sc_avg_clust[,clusters_to_keep]
      sc_avg_clust_spec = t(apply(sc_avg_clust, 1, calc_gene_specificity))
      sc_avg_clust_spec[is.na(sc_avg_clust_spec)] = 0
      
      colnames(sc_avg_clust_spec) = clusters_to_keep
      sc_avg_clust_spec_max = apply(sc_avg_clust_spec, 1, max)
      #plot(density(sc_avg_clust_spec_max))
      #spec_thresh = 0.1
      
      #sc_avg_clust_spec = sc_avg_clust_spec[sc_avg_clust_spec_max>=spec_thresh,]
      sc_avg_clust_spec = t(apply(sc_avg_clust_spec, 1, scale_01))
      
      mat_cur= mat[intersect(rownames(sc_avg_clust_spec), rownames(mat)),]
      mat_cur = mat_cur[,toupper(p)]
      
      
      #cols = colorRamp2(breaks=c(0, max(sc_avg_clust_spec)), colors=c("white", "darkblue"))
      cols = colorRamp2(breaks=seq(0, max(sc_avg_clust_spec), length.out=50), colors = viridis_pal()(50))
      
      width = 1.5
      height = width * nrow(sc_avg_clust_spec) / ncol(sc_avg_clust_spec)
      
      hc_spec = hclust(dist(sc_avg_clust_spec, method="euclidean"),method="ward")
      
      #hc_spec_cut = cutree(hc_spec, k = 7)
      
      mat_cur = mat_cur[hc_spec$labels[hc_spec$order]]
      
      ha_breaks_max = if_else(si<0, min(mat_cur), max(mat_cur))
      ha_breaks_max_color = if_else(si<0, "darkblue", "darkred")
      ha_cols = colorRamp2(breaks=c(0, ha_breaks_max), colors=c("white", ha_breaks_max_color))
      ha =rowAnnotation(logFC=mat_cur, col=list(logFC=ha_cols),
                        simple_anno_size = unit(width / ncol(sc_avg_clust_spec), "in"))
      
      hm = Heatmap(sc_avg_clust_spec, 
                   cluster_columns=F, 
                   cluster_rows=hc_spec, 
                   col=cols,
                   row_names_gp = gpar(fontsize=5),
                   column_names_gp = gpar(fontsize=5),
                   show_row_dend = T,
                   #split=7, 
                   height = unit(height, "in"),
                   width = unit(width, "in"),
                   left_annotation=ha)
      hm
      
      pdf(file.path(ct_dir, sprintf("celltype_%s_%s_sign%s_specificity.pdf", p, co_label, si)), width=width+10, height=height+2)
      print(hm)
      dev.off()
    })
  })
})
```

#### Relative to surround

```{r}

outdir2 = file.path(outdir1, "surround_relative")
dir.create(outdir2)
results_fname = file.path(outdir2, "model_results.rds")
redo = F

if (redo) {

design = fit$design

pairs_list = list(ra= c("ra", "arco"),
                  hvc = c("hvc", "ncl"),
                  lman = c("lman", "nido"),
                  x = c("x", "stri"))
coefs_to_test = c("num_songs_on_euth_date_log_scale", "kl_mean_log_scale")
cons_list = map(coefs_to_test, function(co) {
  tmp = map(pairs_list, function(ps) {
    str_interp("position${p1}.${co} - position${p2}.${co}", list(p1=ps[1], p2=ps[2], co=co))
    })
  names(tmp) = paste(co, names(pairs_list), sep=":")
  tmp
}) %>% unlist()


cons_list_all = map(coefs_to_test, function(co) {
  tmp = map(pairs_list, function(ps) {
    str_interp("position${p1}.${co} - (positionarco.${co} + positionncl.${co} + positionnido.${co} + positionx.${co})", list(p1=ps[1], co=co))
    })
  names(tmp) = paste(co, names(pairs_list), sep=":")
  tmp
}) %>% unlist()
names(cons_list_all) = paste(names(cons_list_all), "all", sep="_")

cons_list = c(cons_list, cons_list_all)
cmat = makeContrasts(contrasts=cons_list,
                     levels=colnames(design))
colnames(cmat) = names(cons_list)
fit2 = contrasts.fit(fit, cmat)   
fit2 = eBayes(fit2)

res_cons = map(names(cons_list), function(x) {
  tt = topTable(fit2, p.value=1, number=nrow(se1), confint=T, coef = x)
  tt = as.data.frame(tt) %>% rownames_to_column(var="gene_id")
  tt
}) %>% set_names(names(cons_list)) %>% bind_rows(.id="contrast")
qsave(res_cons, results_fname)
} else {
  res_cons = qread(results_fname)
}
```

```{r}
res_cons %>% filter(adj.P.Val<.1) %>% select(contrast) %>% table()
```

##### Heatmap

```{r}
heat_dir = file.path(outdir2, "heatmaps")
dir.create(heat_dir)
```

```{r}
genes_to_plot = map(coefs_to_test, function(co) {
  res_cons %>%
  filter(!grepl("all", contrast)) %>% 
  filter(grepl(co, contrast)) %>% 
  filter(adj.P.Val<0.1) %>%
  distinct(gene_id) %>% unlist()
})
```

```{r}
walk(unique(res_$coef), function(co) {

  res_sig = res %>%
    filter(coef==co) %>% 
    group_by(position) %>%
    filter(adj.P.Val<.1) #%>%
    #top_n(10, abs(logFC))

  mat = acast(res %>% filter(gene_id %in% res_sig$gene_id) %>% filter(coef==co), gene_id~position, value.var="logFC")
  colnames(mat) = toupper(colnames(mat))
  
  width = .75
  height = width * nrow(mat) / ncol(mat)
  
  
  hm = Heatmap(mat, clustering_method_rows = "ward.D", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("%s_top_fc_heatmap.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  hm
})
```

```{r}
walk(unique(res$coef), function(co) {

  res_sig = res %>%
    filter(position %in% c("ra", "hvc", "lman", "x")) %>% 
    filter(coef==co) %>% 
    group_by(position) %>%
    filter(adj.P.Val<.1) %>%
    top_n(10, abs(logFC))

  res_cur = res %>% filter(gene_id %in% res_sig$gene_id) %>% filter(coef==co) %>%
     filter(position %in% c("ra", "hvc", "lman", "x"))
  mat = acast(res_cur, gene_id~position, value.var="logFC")
  colnames(mat) = toupper(colnames(mat))
  
  width = .75
  height = width * nrow(mat) / ncol(mat)
  
  
  hm = Heatmap(mat, clustering_method_rows = "ward.D", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("%s_top_fc_songsystem_heatmap.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  hm
})
```

```{r}
#mat = scale(mat, scale = T)
km = kmeans(mat, centers = 10)
res = res %>% mutate(k = km$cluster[match(gene_id, names(km$cluster))])
res_sum = res %>% 
  group_by(k, position) %>% 
  summarize(logFC_mean = mean(logFC)) %>%
  mutate(position = factor(position, levels=position_levels))
gg = ggplot(res_sum, aes(position, logFC_mean)) + 
  geom_point()+ 
  theme_cowplot() + 
  facet_wrap(~k)
gg

gg = ggplot(res %>% filter(gene_id %in% c("CRHBP", "PMP2", "TRAP1")), aes(position, logFC)) + 
  geom_point()+ 
  theme_cowplot() + 
  facet_wrap(~gene_id)
gg
```

#### Across song

```{r}

outdir2 = file.path(outdir1, "songsystem")
dir.create(outdir2)
results_fname = file.path(outdir2, "model_results.rds")
redo = F

if (redo) {
  
  design = fit$design
  
  coefs_list = list(
       pallial_aov = c("positionra.kl_mean_log_scale_cut2_proc3", "positionhvc.kl_mean_log_scale_cut2_proc3", "positionlman.kl_mean_log_scale_cut2_proc3" )
  )
  cons_list = list(

    songsystem= "positionra.kl_mean_log_scale_cut2_proc3 + positionhvc.kl_mean_log_scale_cut2_proc3 + positionlman.kl_mean_log_scale_cut2_proc3", 
                   kl_mean_smp_specific_pallial = "(positionra.kl_mean_log_scale_cut2_proc3 + positionra.kl_mean_log_scale_cut2_proc2  +
                   positionhvc.kl_mean_log_scale_cut2_proc3 + positionhvc.kl_mean_log_scale_cut2_proc2) / 4 - 
                 (positionarco.kl_mean_log_scale_cut2_proc3 + positionarco.kl_mean_log_scale_cut2_proc2 + 
                 positionncl.kl_mean_log_scale_cut2_proc3 + positionncl.kl_mean_log_scale_cut2_proc2) / 4",
     kl_mean_3_ra_lman_specific_pallial = "positionra.kl_mean_log_scale_cut2_proc3 + positionlman.kl_mean_log_scale_cut2_proc3 - 
                 (positionarco.kl_mean_log_scale_cut2_proc3 + positionnido.kl_mean_log_scale_cut2_proc3)",
      kl_mean_3_ra_specific_pallial = "positionra.kl_mean_log_scale_cut2_proc3 - 
                 (positionarco.kl_mean_log_scale_cut2_proc3)",
                 kl_mean_3_smp_specific_pallial = "positionra.kl_mean_log_scale_cut2_proc3 + positionhvc.kl_mean_log_scale_cut2_proc3 - 
                 (positionarco.kl_mean_log_scale_cut2_proc3 + positionncl.kl_mean_log_scale_cut2_proc3)",
                 kl_mean_songsystem_specific_pallial = "(positionra.kl_mean_log_scale_cut2_proc3 +  positionra.kl_mean_log_scale_cut2_proc2  + positionhvc.kl_mean_log_scale_cut2_proc3 + positionhvc.kl_mean_log_scale_cut2_proc2 + positionlman.kl_mean_log_scale_cut2_proc3 + positionlman.kl_mean_log_scale_cut2_proc2) - 
                 (positionarco.kl_mean_log_scale_cut2_proc3 + positionarco.kl_mean_log_scale_cut2_proc2 + positionncl.kl_mean_log_scale_cut2_proc3 + positionncl.kl_mean_log_scale_cut2_proc2 + 
                  positionnido.kl_mean_log_scale_cut2_proc3 + positionnido.kl_mean_log_scale_cut2_proc2)",
                 kl_mean_3_songsystem_specific_pallial = "(positionra.kl_mean_log_scale_cut2_proc3 + positionhvc.kl_mean_log_scale_cut2_proc3 + positionlman.kl_mean_log_scale_cut2_proc3) / 3 - 
                 (positionarco.kl_mean_log_scale_cut2_proc3 + positionncl.kl_mean_log_scale_cut2_proc3 + positionnido.kl_mean_log_scale_cut2_proc3) / 3",
                 kl_mean_3_songsystem_specific = "(positionra.kl_mean_log_scale_cut2_proc3 + positionhvc.kl_mean_log_scale_cut2_proc3 + positionlman.kl_mean_log_scale_cut2_proc3 +  positionx.kl_mean_log_scale_cut2_proc3)  - (positionarco.kl_mean_log_scale_cut2_proc3 + positionncl.kl_mean_log_scale_cut2_proc3 + positionnido.kl_mean_log_scale_cut2_proc3 +  positionstri.kl_mean_log_scale_cut2_proc3)"
  )
  cmat = makeContrasts(contrasts=cons_list,
                       levels=colnames(design))
  colnames(cmat) = names(cons_list)
  fit2 = contrasts.fit(fit, cmat)   
  fit2 = eBayes(fit2)

res_cons = map(names(cons_list), function(x) {
  tt = topTable(fit2, p.value=1, number=nrow(se1), confint=T, coef = x)
  tt = as.data.frame(tt) %>% rownames_to_column(var="gene_id")
  tt
}) %>% set_names(names(cons_list)) %>% bind_rows(.id="contrast")

res_co  = map(coefs_list, function(co) {
  print(co)
  #co_cur = sprintf(co, p)
  fit2 = contrasts.fit(fit, coefficients = co)   
  fit2 = eBayes(fit2)
  tt = topTable(fit2, p.value=1, number=nrow(se1), confint=T)
  tt = as.data.frame(tt) %>% rownames_to_column(var="gene_id") %>%
    select(gene_id, AveExpr, P.Value, adj.P.Val, one_of("F", "t", "logFC"))
  tt
}) %>% set_names(names(coefs_list)) %>% bind_rows(.id="coef")

qsave(res_cons, results_fname)
} else {
  res_cons = qread(results_fname)
}
```

```{r}
res_cons %>% filter(adj.P.Val<.1) %>% select(contrast) %>% table()
res_co %>% filter(adj.P.Val<.1) %>% select(coef) %>% table()
```

##### Volcano

```{r}
volcano_dir = file.path(outdir1, "volcano")
dir.create(volcano_dir)
```

```{r}
res = res %>% mutate(padjust_signed = -1 * log10(adj.P.Val))
res = res %>% mutate(logFC_2 = 100* logFC)
res = res %>% filter(gene_id != "RLOC_00005451g")
```

```{r}
pos_pos2_all = info2 %>% distinct(position, position2, position_pretty, songsystem)
positions = levels(droplevels(res)$position)
ggs = map(positions, function(p) {
  pos_pos2_all = pos_pos2_all %>% mutate(col = position_colors[match(position, names(position_colors))])
  pos_pos2_all_cur = pos_pos2_all %>% filter(position==p)
  #pos_pos2_all_cur = pos_pos2_all_cur %>% mutate(color_label = if_else(songsystem, "white", "black"))
  #nss = pos_pos2_all_cur %>% filter(!songsystem) %>% select(position) %>% unlist() %>% as.character()
  #ss = pos_pos2_all_cur %>% filter(songsystem) %>% select(position) %>% unlist() %>% as.character()
  tmp = res %>% filter(position==p) %>%
    mutate(sig = adj.P.Val < .1) %>% 
    mutate( class = "none",
                    class = if_else(logFC< 0 & sig, "down" , class),
                    class = if_else(logFC > 0 & sig, "up", class))
  
    #mutate( class = "none",
    #                 class = if_else(estimate < 0 & sig, nss , class),
    #                 class = if_else(estimate > 0 & sig, ss, class))
  #colors_cur = pos_pos2_all_cur$col %>% set_names(pos_pos2_all_cur$position)
  colors_cur = c(none="black", up = "firebrick3", down="dodgerblue4")
  colors_fill = c(none="black", up = "firebrick1", down="dodgerblue2")
  #colors_label_cur = pos_pos2_all_cur$color_label %>% set_names(pos_pos2_all_cur$position)
  #colors_label_cur = c(none="black", colors_label_cur)
  sig_label = tmp %>% 
    filter(sig) %>%
    top_n(-15, wt = adj.P.Val)
  tmp = tmp %>%
    mutate(gene_id_sig = ifelse(gene_id %in% sig_label$gene_id, gene_id, NA))
  if (nrow(tmp) == 0)
    next
  gg = ggplot(tmp, aes(logFC_2, padjust_signed, color=class))
  gg = gg + geom_point()
  gg = gg + geom_label_repel(aes(label=gene_id_sig, fill=class), color="black", point.padding = .5, alpha=1)
  gg = gg + scale_color_manual("", values=colors_cur)
  #gg = gg + scale_color_manual("", values=colors_label_cur)
  gg = gg + scale_fill_manual("", values=colors_fill)
  gg = gg + labs(x="log2 fold-change per hundred songs",
                y="-log10(adjusted p-value)",
                title=as.character(pos_pos2_all_cur$position_pretty))
  gg = gg + theme(legend.position="none")
  save_plot(file.path(volcano_dir, sprintf("%s.pdf", p)), gg, base_height=6, base_aspect_ratio = 1.1)
  save_plot(file.path(volcano_dir, sprintf("%s.png", p)), gg, base_height=6, base_aspect_ratio = 1.1)
  return(gg)
})
#ggs = flatten(ggs)
pg = plot_grid(plotlist=ggs, ncol=2, nrow=4, align="hv")
save_plot(file.path(volcano_dir, "combined.pdf"), pg, ncol=2, nrow=4, base_height=5, base_aspect_ratio = 1.1)
save_plot(file.path(volcano_dir, "combined.png"), pg, ncol=2, nrow=4, base_height=5, base_aspect_ratio = 1.1)
pg
#pg = plot_grid(plotlist=ggs, ncol=2, nrow=4, align="hv")
#pg
#save_plot(file.path(volcano_dir, "combined_grid.pdf"), pg, ncol=2, nrow=4, base_height=5, base_aspect_ratio = 1.1)
#save_plot(file.path(volcano_dir, "combined_grid.png"), pg, ncol=2, nrow=4, base_height=5, base_aspect_ratio = 1.1)
```

```{r}
res_mat = res %>% select(position, logFC, gene_id) %>% spread(position, logFC) %>%
  mutate(to_label = if_else(gene_id=="CRHBP", gene_id, ""))
gg = ggplot(res_mat, aes(ra, arco)) + 
  geom_text_repel(aes(label=to_label)) + 
  geom_point()
gg
```

```{r}
res_mat_cor = res_mat %>% select(-gene_id, -to_label) %>% as.matrix() %>% cor()
```

##### Heatmap

```{r}

walk(unique(res_cons$contrast), function(con) {
  print(con)
ntop = 50
res_sig = res_cons %>%
  filter(contrast==con) %>%
  top_n(ntop, abs(logFC))


mat = acast(res %>% filter(gene_id %in% res_sig$gene_id), gene_id~position, value.var="logFC")
colnames(mat) = toupper(colnames(mat))

width = .75
height = width * nrow(mat) / ncol(mat)


hm = Heatmap(mat, clustering_method_rows = "ward.D2", 
             cluster_columns = F,
             width = unit(width, "in"),
             height =unit( height, "in"),
             column_names_gp = gpar(fontsize=6), column_names_side = "top",
             
            row_names_gp = gpar(fontsize=6)
           #  clustering_method_columns = "ward.D"
             )

pdf(file.path(outdir2, sprintf("%s_heatmap_top_fc.pdf", con)), width=width+2, height=height+2)
print(hm)
dev.off()
hm

res_sig = res_cons %>%
  filter(contrast==con) %>%
  filter(adj.P.Val<.1) #%>%
  #top_n(ntop, abs(logFC))


mat = acast(res %>% filter(gene_id %in% res_sig$gene_id), gene_id~position, value.var="logFC")
colnames(mat) = toupper(colnames(mat))

width = .75
height = width * nrow(mat) / ncol(mat)


hm = Heatmap(mat, clustering_method_rows = "ward.D2", 
             cluster_columns = F,
             width = unit(width, "in"),
             height =unit( height, "in"),
             column_names_gp = gpar(fontsize=6), column_names_side = "top",
             
            row_names_gp = gpar(fontsize=6)
           #  clustering_method_columns = "ward.D"
             )

pdf(file.path(outdir2, sprintf("%s_heatmap_sig.pdf", con)), width=width+2, height=height+2)
print(hm)
dev.off()
hm
})
```

```{r}

ntop = 50
con = "songsystem_specific"
res_sig = res_cons %>%
  filter(contrast==con) %>%
  top_n(ntop, abs(logFC))
  #filter(adj.P.Val<.1)

res = res %>% 
  mutate(logFC_100 = 100 * logFC)
mat = acast(res %>% filter(gene_id %in% res_sig$gene_id), gene_id~position, value.var="logFC_100")
colnames(mat) = toupper(colnames(mat))

width = .75
height = width * nrow(mat) / ncol(mat)


hm = Heatmap(mat, clustering_method_rows = "ward.D2", 
             cluster_columns = F,
             width = unit(width, "in"),
             height =unit( height, "in"),
             column_names_gp = gpar(fontsize=6), column_names_side = "top",
             
            row_names_gp = gpar(fontsize=6)
           #  clustering_method_columns = "ward.D"
             )

pdf(file.path(outdir2, sprintf("top_fc_%s_heatmap.pdf", con)), width=width+2, height=height+2)

hm
dev.off()
hm

#ntop = 50
res_sig = res_cons %>%
  filter(contrast=="songsystem_specific") %>%
  #top_n(ntop, abs(logFC))
  filter(adj.P.Val<.1)

res = res %>% 
  mutate(logFC_100 = 100 * logFC)
mat = acast(res %>% filter(gene_id %in% res_sig$gene_id), gene_id~position, value.var="logFC_100")
colnames(mat) = toupper(colnames(mat))

width = .75
height = width * nrow(mat) / ncol(mat)


hm = Heatmap(mat, clustering_method_rows = "ward.D2", 
             cluster_columns = F,
             width = unit(width, "in"),
             height =unit( height, "in"),
             column_names_gp = gpar(fontsize=6), column_names_side = "top",
             
            row_names_gp = gpar(fontsize=6)
           #  clustering_method_columns = "ward.D"
             )

pdf(file.path(outdir2, sprintf("sig_%s_heatmap.pdf", con)), width=width+2, height=height+2)

hm
dev.off()
hm


```

```{r}
#mat = scale(mat, scale = T)
km = kmeans(mat, centers = 10)
res = res %>% mutate(k = km$cluster[match(gene_id, names(km$cluster))])
res_sum = res %>% 
  group_by(k, position) %>% 
  summarize(logFC_mean = mean(logFC)) %>%
  mutate(position = factor(position, levels=position_levels))
gg = ggplot(res_sum, aes(position, logFC_mean)) + 
  geom_point()+ 
  theme_cowplot() + 
  facet_wrap(~k)
gg

gg = ggplot(res %>% filter(gene_id %in% c("CRHBP", "PMP2", "TRAP1")), aes(position, logFC)) + 
  geom_point()+ 
  theme_cowplot() + 
  facet_wrap(~gene_id)
gg
```

##### GSEA

```{r}
gsea_dir = file.path(outdir2, "gsea")
dir.create(gsea_dir)
```

```{r}
gsea_sub_dir = gsea_dir
gsea_run_dir = file.path(gsea_sub_dir, "gsea_runs")
dir.create(gsea_run_dir, recursive = T)

gmt_dir = "~/data/gene_lists/gmt"
gmt_files = list.files(gmt_dir, full.names = T)

#dest_gmt_files = dest_gmt_files[!grepl("Drug|Disease", dest_gmt_files)]
gmt_files = gmt_files[grep("c5.all.v7", gmt_files)]

gmt_pathway = gmtPathways(gmt_files)
#to_exclude = names(gmt_pathway)
#to_exclude = to_exclude[grep("ENDOPLAS|RIBOSOM|TRANSLATION|METABO|CATABOL|RESPIR|HYDROGEN|OXIDAT|OXIDO|MITOCH|PROTON|VIRAL|ELECTRON|ATP|RRNA|BIOSYNTHETIC|INORGANIC|INNER_MEMBRANE", to_exclude)]
#gmt_pathway = gmt_pathway[!(names(gmt_pathway) %in% to_exclude)]
```

```{r}
walk(c("songsystem", "songsystem_specific", "songsystem_specific_pallial", "songsystem_specific_procedure2", "songsystem_specific_pallial_procedure2"), function(con) {
  print(con)
  res_cons_cur = res_cons %>% filter(contrast == con)
  rnks = res_cons_cur$t
  names(rnks) = res_cons_cur$gene_id
  rnks = sort(rnks)
  gsea_res = fgsea(pathways = gmt_pathway,
                   stats = rnks,
                   minSize=15,
                   maxSize=500,
                   nperm=10000)
  
  #positions = as.character(unique(se1$position))
  
  #rnks = res_cons_cur$t
  #names(rnks) = markers_cur$gene_id
  #rnks = sort(rnks)
  gsea_res_filt = gsea_res %>% filter(padj<.2) %>% arrange(pval)
  collapsedPathways = collapsePathways(gsea_res_filt, pathways = gmt_pathway, stats = rnks)
  gsea_res_ind = gsea_res %>% filter(pathway %in% collapsedPathways$mainPathways) %>%
    arrange(-NES)
  
  gsea_res_ind_filt = gsea_res_ind %>% filter(padj<.2)
  gsea_res_ind_top = gsea_res_ind %>% filter(padj<.2) %>%
    top_n(10, NES)
  gsea_res_filt = gsea_res %>% filter(padj<.2)
  
  tmp = gsea_res_ind_filt %>% 
    
    mutate(class = case_when(grepl("HORMONE|NEUROTR|PEPTI", pathway) ~ "modulation",
                             grepl("OXID|NADH|MITO|RESPIR|PROTON|REDOX", pathway) ~ "respiration",
                             grepl("RIBO|PROTEIN_LOCALIZ|TRANSLATION|RRNA", pathway) ~ "translation",
                             TRUE ~ "none"),
    ) %>%
    mutate(pathway = sub("GO_", "", pathway),
           pathway = case_when(grepl("NONSENSE", pathway) ~ "NONSENSE_MEDIATED_DECAY",
                               grepl("ESTABLISHMENT_OF_PROTEIN", pathway) ~ "PROTEIN_LOCALIZATION_TO_ER",
                               TRUE ~ pathway),
           pathway = gsub("_", " ", pathway),
    )  %>%
    arrange(NES) %>%
    mutate(pathway = factor(pathway, levels=pathway)) 
  
  cols =c("black",  pal_aaas()(3))
  names(cols) = c("none", "respiration",  "modulation", "translation")
  gg = ggplot(tmp, aes(NES, pathway)) + 
    
    geom_segment(aes(x=min(NES),xend = NES, y=pathway, yend=pathway), size=.25, linetype=3) + 
    geom_segment(aes(x = 0, xend = NES, y=pathway, yend=pathway), size=.25) +
    
    geom_point(aes(color=class)) + 
    theme_light() + 
    theme(axis.text = element_text(size=5),
          axis.title = element_text(size=6),
          axis.line.x = element_line(size=.25),
          axis.line.y = element_blank(),
          axis.ticks.x = element_line(size=.25),
          axis.ticks.y = element_blank(),
          legend.position = "bottom",
          legend.direction="horizontal",
          legend.text = element_text(size=5)) + 
    labs(y="") + 
    scale_color_manual(values=cols, name="")
  gg
  save_plot(file.path(gsea_dir, sprintf("%s_point.pdf", con)), gg, base_height=3, base_width=4)
})
```
